---
title: "Codacore Ustek and Vedo"
author: "Iwan"
date: "2023-12-04"
output: html_document
---

#### code adapted from : 
"https://github.com/GRONINGEN-MICROBIOME-CENTRE/Fecal_Metabolites_IBD/blob/main/IBD-prediction-models/predictions.Rmd". 
which is adapted from : 
"https://egr95.github.io/R-codacore/inst/misc/guide.html".

```{r setup, include=FALSE}
library (codacore)
library (tensorflow)
library (tidyverse)
library (plyr)
```



```{r ranko function}
#### Ranko's geweldige filter functie -----
#================================================================================
  # function for filtering microbiome (metaphlan) results
  # - takes dataframe (any)
  # - replaces NAs with 0.0
  # - filters OUT taxa present in less then presPerc samples
  # - filters OUT taxa with mean releative abundance < minMRelAb
  # - filters OUT taxa with median relative abundance < minMedRelAb
  # - filters OUT taxonomic levels not in keepLevels ()
  #   -> keepLevels should be vector of following: T = strain, S = 
species, G = Genera, F = families
  #                                                O = Orders, C = 
classes, P = phyla, K = Kingdoms
  #   -> example: keepLevels=c('S','G') keeps only species and genera
  # - filters OUT domains not in keepDomains: 
#   -> input is vector of following: Eukaryota, Bacteria, Viruses, Archaea
#   -> example: keepDomains=c('B') keeps only bacteria, removes viruses, 
archaea and eukarya
# - if rescaleTaxa = True, rescales relative abundancies after filtering
# returns modified dataframe
# NOTES:
# - assumes metaphlan encoding (k__<kingdom>.o__<order> ... ); it will 
mess stuff
# up if non-metagenome rows are encoded like this!
# DEFAULTS: 
# - removes non-bacteria
# - keeps all except kingdoms
filterMetaGenomeDF <- function(inDF,presPerc = 0.1,minMRelAb = 
0.01,minMedRelAb=0.0,
                               rescaleTaxa=T,verbose=T,
                               keepDomains=c('Bacteria','Archaea'),
                               keepLevels=c('S','G','F','O','C','P'),
                               keepUnknown=F) {
  
  # -- drop unknown & rescale? --
  if (!keepUnknown) {
    inDF$UNKNOWN <- NULL
  }
  tCols = grep('k__',colnames(inDF)) # colums with microbiome
  tColsNMG = grep('k__',colnames(inDF),invert = T) # colums with 
microbiome
  # replaces NAs in microbiome with 0s
  for (c in tCols) {
    inDF[,c][is.na(inDF[,c])] <- 0.0
  }
  
  # filter for presence
  # -----------------
  nrRemoved = 0
  toRemove = c()
  for (c in tCols) {
    nrnZ = as.numeric(sum(inDF[,c]!=0.0))
    if ( (nrnZ/as.numeric(nrow(inDF))) < presPerc) {
      # if (verbose) {
      #   print (paste('col',c,': ',colnames(inDF)[c],'; nr non 
Zero:',nrnZ,'=',nrnZ/as.numeric(nrow(inDF)),'>> Column removed!'))
      # }
      nrRemoved = nrRemoved + 1
      toRemove <- c(toRemove,c)
    }
  }
  if (length(toRemove) > 0) {
    inDF <- inDF[,-toRemove]
  }
  tCols = grep('^[dgtspcfko]__',colnames(inDF)) # colums with microbiome
  if (verbose) {print (paste(' > presence filter: 
Removed',nrRemoved,'taxa!, ',length(tCols),'taxa left!')); }
  
  # filter for abundance (mean)
  # ---------------------------
  nrRemoved = 0
  toRemove = c()
  for (c in tCols) {
    mn = mean(inDF[,c])
    if ( mn < minMRelAb) {
      #if (verbose) {
      #  print (paste('col',c,': ',colnames(inDF)[c],'; mean rel 
abundance:',mn,' >> Column removed!'))
      #}
      nrRemoved = nrRemoved + 1
      toRemove <- c(toRemove,c)
    }
  }
  if (length(toRemove) > 0) {
    inDF <- inDF[,-toRemove]
  }
  tCols = grep('k__',colnames(inDF)) # colums with microbiome
  if (verbose) {print (paste(' > mean abundance filter: 
Removed',nrRemoved,'taxa!, ',length(tCols),'taxa left!')); }
  
  # filter for abundance (median)
  # -----------------------------
  nrRemoved = 0
  toRemove = c()
  for (c in tCols) {
    mn = median(inDF[,c])
    if ( mn < minMedRelAb) {
      #if (verbose) {
      #  print (paste('col',c,': ',colnames(inDF)[c],'; median rel 
abundance:',mn,' >> Column removed!'))
      #}
      nrRemoved = nrRemoved + 1
      toRemove <- c(toRemove,c)
    }
  }
  if (length(toRemove) > 0) {
    inDF <- inDF[,-toRemove]
  }
  if (verbose) {print (paste(' > median abundance filter: 
Removed',nrRemoved,'taxa!, ',length(tCols),'taxa left!')); }  
  
  # keep domains
  # -----------------------------
  toKeep <- NULL
  if (length(keepDomains) == 1) {
    if (keepDomains == 'All' | keepDomains == "") {
      toKeep = grep('k__',colnames(inDF),invert = T)
    } else {
      toKeep = grep(paste('k__',d,sep=''),colnames(inDF))
    }
  } else {
    for (d in keepDomains) {
      #d print(d)
      toKeep = c(toKeep,grep(paste('k__',d,sep=''),colnames(inDF)))
    }
  }
  inDF <- inDF[,toKeep]
  
  # remove taxonomic levels
  # -----------------------------
  inDFnonTaxa <- as.data.frame(inDF[,grep('k__',colnames(inDF),invert=T)])
  colnames(inDFnonTaxa) <- 
colnames(inDF)[grep('k__',colnames(inDF),invert=T)]
  
  inDF2 <- as.data.frame(inDF[,grep('k__',colnames(inDF),invert=F)])
  # pick strains (T)
  taxaTCols <- grep('t__',colnames(inDF2))
  taxaT <- as.data.frame(inDF2[,taxaTCols])
  colnames(taxaT) <- colnames(inDF2)[grep('t__',colnames(inDF2))]
  if (length(taxaTCols) > 0) {inDF2 <- inDF2[,-taxaTCols]}
  # pick species (S)
  taxaSCols <- grep('s__',colnames(inDF2))
  if (length(taxaSCols) > 0) {
    taxaS <- as.data.frame(inDF2[,taxaSCols])
    colnames(taxaS) <- colnames(inDF2)[grep('s__',colnames(inDF2))]
  }
  if (length(taxaSCols) > 0) {inDF2 <- inDF2[,-taxaSCols]}
  # pick genera (G)
  taxaGCols <- grep('g__',colnames(inDF2))
  if (length(taxaGCols) > 0) {
    taxaG <- as.data.frame(inDF2[,taxaGCols])
    colnames(taxaG) <- colnames(inDF2)[grep('g__',colnames(inDF2))]
  }
  if (length(taxaGCols) > 0) {inDF2 <- inDF2[,-taxaGCols]}
  # pick families (F)
  taxaFCols <- grep('f__',colnames(inDF2))
  if (length(taxaFCols) > 0) {    
    taxaF <- as.data.frame(inDF2[,taxaFCols])
    colnames(taxaF) <- colnames(inDF2)[grep('f__',colnames(inDF2))]
  }
  if (length(taxaFCols) > 0) {inDF2 <- inDF2[,-taxaFCols]}
  # pick orders (O)
  taxaOCols <- grep('o__',colnames(inDF2))
  if (length(taxaOCols) > 0) {
    taxaO <- as.data.frame(inDF2[,taxaOCols])
    colnames(taxaO) <- colnames(inDF2)[grep('o__',colnames(inDF2))]
  }
  if (length(taxaOCols) > 0) {inDF2 <- inDF2[,-taxaOCols]}
  # pick classes (C)
  taxaCCols <- grep('c__',colnames(inDF2))
  if (length(taxaCCols) > 0) {
    taxaC <- as.data.frame(inDF2[,taxaCCols])
    colnames(taxaC) <- colnames(inDF2)[grep('c__',colnames(inDF2))]
  }
  if (length(taxaCCols) > 0) {
    if( length(colnames(inDF2)) - length(taxaCCols) == 1) {
      ccn <- colnames(inDF2)[grep('c__',colnames(inDF2))]
      tempN <- colnames(inDF2)[!(colnames(inDF2) %in% ccn)]
      inDF2 <- as.data.frame(inDF2[,-taxaCCols])
      colnames(inDF2) <- tempN
    } else {
      inDF2 <- inDF2[,-taxaCCols]
    }
  }
  # pick phyla (P)
  taxaPColsKeepN <- NULL
  taxaPCols <- grep('p__',colnames(inDF2))
  if (length(taxaPCols) > 0) {
    taxaPColsN <- colnames(inDF2)[grep('p__',colnames(inDF2))]
    taxaPColsKeep <- grep('p__',colnames(inDF2),invert = T)
    taxaPColsKeepN <- colnames(inDF2)[grep('p__',colnames(inDF2),invert = 
T)]
    taxaP <- as.data.frame(inDF2[,taxaPCols])
    colnames(taxaP) <- taxaPColsN
  }
  # pick kingdoms (K)
  if (length(taxaPCols) > 0) {inDF2 <- 
as.data.frame(inDF2[,taxaPColsKeep])}
  if (!is.null(taxaPColsKeepN)){
    colnames(inDF2) <- taxaPColsKeepN
  }
  taxaK <- inDF2
  # pick 
  oDF <- inDFnonTaxa
  if (verbose) {print ('Keeping following taxonomic levels:'); 
print(keepLevels)}
  if (ncol(taxaK) > 0) {  
    if ('K' %in% keepLevels) {
      if (verbose){print(paste(' -> kept',ncol(taxaK),'Kingdoms'))}
      if (rescaleTaxa) {taxaK <- taxaK/rowSums(taxaK)}
      oDF <- cbind(oDF,taxaK)}
  }
  if (ncol(taxaP) > 0) {
    if ('P' %in% keepLevels) {
      if (verbose){print(paste(' -> kept',ncol(taxaP),'Phyla'))} 
      if (rescaleTaxa) {taxaP <- taxaP/rowSums(taxaP)}
      oDF <- cbind(oDF,taxaP)}
  }
  if (ncol(taxaC) > 0) {
    if ('C' %in% keepLevels) {
      if (verbose) {print(paste(' -> kept',ncol(taxaC),'Classes'))} 
      if (rescaleTaxa) {taxaC <- taxaC/rowSums(taxaC)}
      oDF <- cbind(oDF,taxaC)}
  }
  if (ncol(taxaO) > 0) {
    if ('O' %in% keepLevels) {
      if (verbose){print(paste(' -> kept',ncol(taxaO),'Orders'))}
      if (rescaleTaxa) {taxaO <- taxaO/rowSums(taxaO)}
      oDF <- cbind(oDF,taxaO)}
  }
  if (ncol(taxaF) > 0) {
    if ('F' %in% keepLevels) {
      if (verbose){print(paste(' -> kept',ncol(taxaF),'Families'))}
      if (rescaleTaxa) {taxaF <- taxaF/rowSums(taxaF)}
      oDF <- cbind(oDF,taxaF)}
  }
  if (ncol(taxaG) > 0) { 
    if ('G' %in% keepLevels) {if (verbose){ print(paste(' -> 
kept',ncol(taxaG),'Genera'))}
      if (rescaleTaxa) {taxaG <- taxaG/rowSums(taxaG)}
      oDF <- cbind(oDF,taxaG)}
  }
  if (ncol(taxaS) > 0) {
    if ('S' %in% keepLevels) {if (verbose){print(paste(' -> 
kept',ncol(taxaS),'Species'))}
      if (rescaleTaxa) {taxaS <- taxaS/rowSums(taxaS)}
      oDF <- cbind(oDF,taxaS)}
  }
  if (ncol(taxaT) > 0) {
    if ('T' %in% keepLevels) {if (verbose){print(paste(' -> 
kept',ncol(taxaT),'Strains'))}
      if (rescaleTaxa) {taxaT <- taxaT/rowSums(taxaT)}
      oDF <- cbind(oDF,taxaT)}
  }
  if (verbose) {print ('data processing done, returning Dataframe')}
  oDF
}
```



#### vedo data prep
```{r vedo data prep}
#Prepare data#

taxa_df <- 
as.data.frame(t(read.csv("/Users/iwan/Research/drug_response/metadata/lastest_metadata/complete_metaphlan_merged.txt", 
sep='\t',row.names = 1)))

taxa_df$UNCLASSIFIED <- NULL

taxa_df <- as.data.frame(t(taxa_df))
taxa_df$NCBI_tax_id <- NULL

#Vedo_taxa_df <- taxa_df[, grep("Vedop1", colnames(taxa_df))]
Vedo_metabolites <- 
read.csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/metabolites_vedo_clin.txt')
vedo_t1_samples <- c()
columnsnames_vedo <- c()
for (i in Vedo_metabolites$Fecal_sample_ID_1){
  if (i != ""){
    columnsnames_vedo <- c(columnsnames_vedo, i)
    vedo_t1_samples <- c(vedo_t1_samples, colnames(taxa_df)[grep(i, 
colnames(taxa_df))])
  }
  
}

Vedo_taxa_df <- taxa_df[,vedo_t1_samples]
colnames(Vedo_taxa_df) <-  columnsnames_vedo
saved_rownames <- colnames(Vedo_taxa_df)

Vedo_taxa_df_input <- as.data.frame(t(Vedo_taxa_df))
Vedo_taxa_df_input<- data.frame(lapply(Vedo_taxa_df_input, as.numeric))



Taxa_DR_species <- filterMetaGenomeDF(Vedo_taxa_df_input, presPerc = 
0.15,minMRelAb = 0.1 ,minMedRelAb=0.0,
                                      rescaleTaxa=T,verbose=T,
                                      keepDomains=c('Bacteria','Archaea'),
                                      keepLevels=c('S'),
                                      keepUnknown=F) 


#updated_rownames <- str_split_i(saved_rownames, '_', 1)
row.names(Taxa_DR_species) <- saved_rownames
Taxa_DR_species <- na.omit(Taxa_DR_species)

#Taxa_DR_species
Taxa_DR_species <- na.omit(Taxa_DR_species)


Vedo_metabolites$X <- NULL



merged_taxa_species_clinical_vedo <- merge(Vedo_metabolites, 
Taxa_DR_species, by.x = "Fecal_sample_ID_1", by.y = 0)


vedo_clinical_data <- merged_taxa_species_clinical_vedo[, 1:689]
vedo_metabolites_only <- merged_taxa_species_clinical_vedo[,693:2097]
vedo_microbiome_only <- merged_taxa_species_clinical_vedo[,2099:2165]

Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/Vedo_read_counts.txt', 
sep ='\t', header=TRUE)

merge_samples <- c()
ordered_samples <- c()
for (i in merged_taxa_species_clinical_vedo$Fecal_sample_ID_1){
  merge_samples <- c(merge_samples, i)
  ordered_samples <- c(ordered_samples, Microbiome_counts$Sample[grep( i, 
Microbiome_counts$Sample)])
}

rownames(Microbiome_counts) <- Microbiome_counts$Sample
  
Microbiome_counts_filtered <- Microbiome_counts[ordered_samples,]


Microbiome_counts_filtered$merging_column <- merge_samples

vedo_microbiome_only$Fecal_sample_ID_1 <- 
merged_taxa_species_clinical_vedo$Fecal_sample_ID_1


vedo_microbiome_only_with_counts <- merge(vedo_microbiome_only, 
Microbiome_counts_filtered, by.x = "Fecal_sample_ID_1", by.y = 
"merging_column")


vedo_microbiome_only_with_counts$merging_column <- NULL
vedo_microbiome_only_with_counts$Fecal_sample_ID_1 <- NULL
vedo_microbiome_only_with_counts$Sample <- NULL
vedo_microbiome_only_with_counts$Raw.reads..p1. <- NULL
vedo_microbiome_only_with_counts$Raw.reads..p2. <- NULL
vedo_microbiome_only_with_counts$Final..p2. <- NULL
vedo_microbiome_only_with_counts$Fecal_sample_ID_1 <- NULL
vedo_microbiome_only_with_counts$sample_id_right <- NULL

vedo_microbiome_sub_counts <- vedo_microbiome_only_with_counts %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))

vedo_microbiome_sub_counts$Final..p1. <- NULL

vedo_microbiome_sub_counts[vedo_microbiome_sub_counts == 0] <- NA

vedo_microbiome_sub_counts[sapply(vedo_microbiome_sub_counts, is.numeric)] 
<- lapply(vedo_microbiome_sub_counts[sapply(vedo_microbiome_sub_counts, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))

response_score <- vedo_clinical_data$ResponseScoreYesNo

vedo_all_new_ID_raw_filt=vedo_metabolites_only[,((colSums(!is.na(vedo_metabolites_only)) 
/ nrow(vedo_metabolites_only)) *100 )>70]
#all_new_ID_raw[is.na(all_new_ID_raw)]=0
vedo_all_new_ID_raw_filt[sapply(vedo_all_new_ID_raw_filt, is.numeric)] <- 
lapply(vedo_all_new_ID_raw_filt[sapply(vedo_all_new_ID_raw_filt, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))

```


#### 6 months metabolites
```{r vedo 6 months metabolites}

my_phenos=vedo_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling")]

mtb_cc=merge(my_phenos,all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list_0 <- c()
num_list_0 <- c()
denum_list_0 <- c()
AUC_list_0 <- c()
test_auc_list_0 <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,7:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )
  
  possibleError <- tryCatch(
    {
    print('training')
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list_0 <- c(num_list_0, (nums))
    denum_list_0 <- c(denum_list_0, (denums))
    test_auc_list_0 <- c(test_auc_list_0, test_auc)
    AUC_list_0 <- c(AUC_list_0, AUC)
    acc_list_0 <- c(acc_list_0, acc)
    counter <- counter ++ 1
    print(counter)
    },
    error=function(cond){
      print(cond)
      
    }
  )
  if(inherits(possibleError, "error")) next
}

vedo_6_month_metabolites_values <- c(mean(AUC_list_0), sd(AUC_list_0), 
mean(test_auc_list_0), sd(test_auc_list_0))
cat(mean(AUC_list_0))
cat(sd(AUC_list_0))
cat(mean(test_auc_list_0))
cat(sd(test_auc_list_0))

vedo_nums_6_month_response <- as.data.frame(table(num_list_0))
vedo_denums_6_month_response <- as.data.frame(table(denum_list_0))


vedo_nums_6_month_response$values <- vedo_nums_6_month_response$Freq/100
vedo_denums_6_month_response$values <- 
vedo_denums_6_month_response$Freq/100

write.csv(vedo_nums_6_month_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_6_month_metabolites.csv')
write.csv(vedo_denums_6_month_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_6_month_metabolites.csv')


```



#### 2 years metabolites
```{r vedo 2 year metabolites}

my_phenos=vedo_clinical_data[,c("Response_2years","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling")]

mtb_cc=merge(my_phenos,all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list_0 <- c()
num_list_0 <- c()
denum_list_0 <- c()
AUC_list_0 <- c()
test_auc_list_0 <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (verhuisd)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (overleden)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "?", ]
predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,7:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )
  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list_0 <- c(num_list_0, (nums))
    denum_list_0 <- c(denum_list_0, (denums))
    test_auc_list_0 <- c(test_auc_list_0, test_auc)
    AUC_list_0 <- c(AUC_list_0, AUC)
    acc_list_0 <- c(acc_list_0, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
      
    }
  )
  if(inherits(possibleError, "error")) next
}
vedo_2_year_metabolites_values <- c(mean(AUC_list_0), sd(AUC_list_0), 
mean(test_auc_list_0), sd(test_auc_list_0))
cat(mean(AUC_list_0))
cat(sd(AUC_list_0))
cat(mean(test_auc_list_0))
cat(sd(test_auc_list_0))

vedo_nums_2_year_response <- as.data.frame(table(num_list_0))
vedo_denums_2_year_response <- as.data.frame(table(denum_list_0))


vedo_nums_2_year_response$values <- vedo_nums_2_year_response$Freq/100
vedo_denums_2_year_response$values <- vedo_denums_2_year_response$Freq/100

write.csv(vedo_nums_2_year_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_2_year_metabolites.csv')
write.csv(vedo_denums_2_year_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_2_year_metabolites.csv')


```





#### HBI SCCAI metabolites
```{r vedo HBI SCCAI scores metabolites}

my_phenos=vedo_clinical_data[,c("Response_HBI_SCCAI","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling")]

mtb_cc=merge(my_phenos,all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list_0 <- c()
num_list_0 <- c()
denum_list_0 <- c()
AUC_list_0 <- c()
test_auc_list_0 <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,7:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )
  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list_0 <- c(num_list_0, (nums))
    denum_list_0 <- c(denum_list_0, (denums))
    test_auc_list_0 <- c(test_auc_list_0, test_auc)
    AUC_list_0 <- c(AUC_list_0, AUC)
    acc_list_0 <- c(acc_list_0, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
      
    }
  )
  if(inherits(possibleError, "error")) next
}
vedo_HBI_metabolites_values <- c(mean(AUC_list_0), sd(AUC_list_0), 
mean(test_auc_list_0), sd(test_auc_list_0))
cat(mean(AUC_list_0))
cat(sd(AUC_list_0))
cat(mean(test_auc_list_0))
cat(sd(test_auc_list_0))

vedo_nums_HBI_response <- as.data.frame(table(num_list_0))
vedo_denums_HBI_response <- as.data.frame(table(denum_list_0))


vedo_nums_HBI_response$values <- vedo_nums_HBI_response$Freq/100
vedo_denums_HBI_response$values <- vedo_denums_HBI_response$Freq/100

write.csv(vedo_nums_HBI_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_HBI_metabolites.csv')
write.csv(vedo_denums_HBI_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_HBI_metabolites.csv')


```


#### 6 months microbes
```{r vedo 6 month microbes}

my_phenos=vedo_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling")]

mtb_cc=merge(my_phenos,vedo_microbiome_sub_counts,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list_0 <- c()
num_list_0 <- c()
denum_list_0 <- c()
AUC_list_0 <- c()
test_auc_list_0 <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,7:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )
  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list_0 <- c(num_list_0, (nums))
    denum_list_0 <- c(denum_list_0, (denums))
    test_auc_list_0 <- c(test_auc_list_0, test_auc)
    AUC_list_0 <- c(AUC_list_0, AUC)
    acc_list_0 <- c(acc_list_0, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
      
    }
  )
  if(inherits(possibleError, "error")) next
}
vedo_6_month_microbes_values <- c(mean(AUC_list_0), sd(AUC_list_0), 
mean(test_auc_list_0), sd(test_auc_list_0))
cat(mean(AUC_list_0))
cat(sd(AUC_list_0))
cat(mean(test_auc_list_0))
cat(sd(test_auc_list_0))

vedo_nums_6_month_response <- as.data.frame(table(num_list_0))
vedo_denums_6_month_response <- as.data.frame(table(denum_list_0))


vedo_nums_6_month_response$values <- vedo_nums_6_month_response$Freq/100
vedo_denums_6_month_response$values <- 
vedo_denums_6_month_response$Freq/100

write.csv(vedo_nums_6_month_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_6_month_microbes.csv')
write.csv(vedo_denums_6_month_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_6_month_microbes.csv')


```



#### 2 years microbes
```{r vedo 2 year microbes}

my_phenos=vedo_clinical_data[,c("Response_2years","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling")]

mtb_cc=merge(my_phenos,vedo_microbiome_sub_counts,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list_0 <- c()
num_list_0 <- c()
denum_list_0 <- c()
AUC_list_0 <- c()
test_auc_list_0 <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (verhuisd)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (overleden)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "?", ]

predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,7:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )
  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list_0 <- c(num_list_0, (nums))
    denum_list_0 <- c(denum_list_0, (denums))
    test_auc_list_0 <- c(test_auc_list_0, test_auc)
    AUC_list_0 <- c(AUC_list_0, AUC)
    acc_list_0 <- c(acc_list_0, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
      
    }
  )
  if(inherits(possibleError, "error")) next
}
vedo_2_year_microbes_values <- c(mean(AUC_list_0), sd(AUC_list_0), 
mean(test_auc_list_0), sd(test_auc_list_0))
cat(mean(AUC_list_0))
cat(sd(AUC_list_0))
cat(mean(test_auc_list_0))
cat(sd(test_auc_list_0))

vedo_nums_2_year_response <- as.data.frame(table(num_list_0))
vedo_denums_2_year_response <- as.data.frame(table(denum_list_0))


vedo_nums_2_year_response$values <- vedo_nums_2_year_response$Freq/100
vedo_denums_2_year_response$values <- vedo_denums_2_year_response$Freq/100

write.csv(vedo_nums_2_year_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_2_year_microbes.csv')
write.csv(vedo_denums_2_year_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_2_year_microbes.csv')


```




#### HBI SCCAI microbes
```{r vedo HBI SCCAI scores microbes}

my_phenos=vedo_clinical_data[,c("Response_HBI_SCCAI","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling")]

mtb_cc=merge(my_phenos,vedo_microbiome_sub_counts,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list_0 <- c()
num_list_0 <- c()
denum_list_0 <- c()
AUC_list_0 <- c()
test_auc_list_0 <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,7:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )
  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list_0 <- c(num_list_0, (nums))
    denum_list_0 <- c(denum_list_0, (denums))
    test_auc_list_0 <- c(test_auc_list_0, test_auc)
    AUC_list_0 <- c(AUC_list_0, AUC)
    acc_list_0 <- c(acc_list_0, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
      
    }
  )
  if(inherits(possibleError, "error")) next
}
vedo_HBI_microbes_values <- c(mean(AUC_list_0), sd(AUC_list_0), 
mean(test_auc_list_0), sd(test_auc_list_0))
cat(mean(AUC_list_0))
cat(sd(AUC_list_0))
cat(mean(test_auc_list_0))
cat(sd(test_auc_list_0))

vedo_nums_HBI_response <- as.data.frame(table(num_list_0))
vedo_denums_HBI_response <- as.data.frame(table(denum_list_0))


vedo_nums_HBI_response$values <- vedo_nums_HBI_response$Freq/100
vedo_denums_HBI_response$values <- vedo_denums_HBI_response$Freq/100

write.csv(vedo_nums_HBI_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_HBI_microbes.csv')
write.csv(vedo_denums_HBI_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_HBI_microbes.csv')


```




#### Vedo result table
```{r Vedo result table}


Vedo_predictions <- as.data.frame(t(data.frame("Vedo_6_month_metabolites" 
= vedo_6_month_metabolites_values,
                     "Vedo_2_year_metabolites" = 
vedo_2_year_metabolites_values,
                     "Vedo_HBI_metabolites" = vedo_HBI_metabolites_values,
                     "vedo_6_month_microbes" = 
vedo_6_month_microbes_values,
                     "vedo_2_year_microbes" = vedo_2_year_microbes_values,
                     "vedo_HBI_microbes" = vedo_HBI_microbes_values)))

colnames(Vedo_predictions) <- c("train_auc", "train_auc_sd", "test_auc", 
"test_auc_sd")
print(Vedo_predictions)
write.csv(Vedo_predictions, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/Vedo_predictions_table.csv')

```




#### now for ustek:
```{r Ustek data prep}
#Prepare data#

taxa_df <- 
as.data.frame(t(read.csv("/Users/iwan/Research/drug_response/metadata/lastest_metadata/complete_metaphlan_merged.txt", 
sep='\t',row.names = 1)))

taxa_df$UNCLASSIFIED <- NULL

taxa_df <- as.data.frame(t(taxa_df))
taxa_df$NCBI_tax_id <- NULL

Ust_taxa_df <- taxa_df[, grep("Ust", colnames(taxa_df))]
Ust_taxa_df_timepoint_one <-Ust_taxa_df[, grep("_1_", 
colnames(Ust_taxa_df))]
saved_rownames <- colnames(Ust_taxa_df_timepoint_one)

Ust_taxa_df_timepoint_one_input <- 
as.data.frame(t(Ust_taxa_df_timepoint_one))
Ust_taxa_df_timepoint_one_input<- 
data.frame(lapply(Ust_taxa_df_timepoint_one_input, as.numeric))



Taxa_DR_species <- filterMetaGenomeDF(Ust_taxa_df_timepoint_one_input, 
presPerc = 0.15,minMRelAb = 0.1 ,minMedRelAb=0.0,
                                      rescaleTaxa=T,verbose=T,
                                      keepDomains=c('Bacteria','Archaea'),
                                      keepLevels=c('S'),
                                      keepUnknown=F) 


updated_rownames <- str_split_i(saved_rownames, '_', 1)
row.names(Taxa_DR_species) <- updated_rownames
Taxa_DR_species <- na.omit(Taxa_DR_species)

#Taxa_DR_species
Taxa_DR_species <- na.omit(Taxa_DR_species)




# load metabolites
Ustek_metabolites <- 
read.csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/metabolites_ustek_clin.txt')
Ustek_metabolites$X <- NULL

starting_rownames <- Ustek_metabolites$sample_id_right
starting_rownames_v1 <- gsub('USTEK','Ust',starting_rownames)
starting_rownames_v2 <- gsub('t00','t',starting_rownames_v1)
starting_rownames_v3 <- gsub('t0','t',starting_rownames_v2)

rownames(Ustek_metabolites) <- starting_rownames_v3


Ustek_clincal <- Ustek_metabolites[, 1:697]



response_score <- Ustek_clincal$ResponseScoreYesNo


merged_taxa_species_clinical_ustek <- merge(Ustek_metabolites, 
Taxa_DR_species, by.x = 0, by.y = 0)

merged_taxa_species_clinical_ustek <- 
merged_taxa_species_clinical_ustek[merged_taxa_species_clinical_ustek$ResponseScoreYesNo 
!= 'not_known',]


write.csv(merged_taxa_species_clinical_ustek, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/Ustek_merged_everything.csv')


Ustek_clinical_data <- merged_taxa_species_clinical_ustek[, 1:697]
Ustek_metabolites_only <- merged_taxa_species_clinical_ustek[,704:2107]
Ustek_microbiome_only <- merged_taxa_species_clinical_ustek[,2110:2175]

Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_read_counts.tsv', 
sep ='\t', header=FALSE)

Microbiome_counts <- Microbiome_counts %>%
  filter(!grepl("_2$", V1))



Microbiome_counts$V1 <- gsub('_1','',Microbiome_counts$V1) 



merging_rownames <- merged_taxa_species_clinical_ustek$sample_id_right
merging_rownames_v1 <- gsub('USTEK','Ust',merging_rownames)
merging_rownames_v2 <- gsub('t00','t',merging_rownames_v1)
merging_rownames_v3 <- gsub('t0','t',merging_rownames_v2)


Ustek_microbiome_only$merging <- merging_rownames_v3

Ustek_microbiome_only_with_counts <- merge(Ustek_microbiome_only, 
Microbiome_counts, by.x = "merging", by.y = "V1")
Ustek_microbiome_only_with_counts$merging <- NULL

Ustek_microbiome_sub_counts <- Ustek_microbiome_only_with_counts %>%
  mutate(across(!c(V2), ~ . * V2))

Ustek_microbiome_sub_counts$V2 <- NULL
Ustek_microbiome_sub_counts[Ustek_microbiome_sub_counts == 0] <- NA

Ustek_microbiome_sub_counts[sapply(Ustek_microbiome_sub_counts, 
is.numeric)] <- 
lapply(Ustek_microbiome_sub_counts[sapply(Ustek_microbiome_sub_counts, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))



response_score <- merged_taxa_species_clinical_ustek$ResponseScoreYesNo
response_score_2year <- merged_taxa_species_clinical_ustek$Response_2years
response_score_HBI_SCCAI <- 
merged_taxa_species_clinical_ustek$Response_HBI_SCCAI
response_score_biochemical <- 
merged_taxa_species_clinical_ustek$Response_biochemical
response_score_biochemical_OR <- 
merged_taxa_species_clinical_ustek$Response_biochemical_OR
response_score_FCP <- merged_taxa_species_clinical_ustek$Response_FCP


ustek_all_new_ID_raw_filt=Ustek_metabolites_only[,((colSums(!is.na(Ustek_metabolites_only)) 
/ nrow(Ustek_metabolites_only)) *100 )>70]
#all_new_ID_raw[is.na(all_new_ID_raw)]=0
ustek_all_new_ID_raw_filt[sapply(ustek_all_new_ID_raw_filt, is.numeric)] 
<- lapply(ustek_all_new_ID_raw_filt[sapply(ustek_all_new_ID_raw_filt, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))


```






### ustek 6 month response metabolites
```{r Ustek 6 month response metabolites }

my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

mtb_cc=merge(my_phenos,all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list_0 <- c()
num_list_0 <- c()
denum_list_0 <- c()
AUC_list_0 <- c()
test_auc_list_0 <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )
  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list_0 <- c(num_list_0, (nums))
    denum_list_0 <- c(denum_list_0, (denums))
    test_auc_list_0 <- c(test_auc_list_0, test_auc)
    AUC_list_0 <- c(AUC_list_0, AUC)
    acc_list_0 <- c(acc_list_0, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
      
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_6_month_metabolites_values <- c(mean(AUC_list_0), sd(AUC_list_0), 
mean(test_auc_list_0), sd(test_auc_list_0))
cat(mean(AUC_list_0))
cat(sd(AUC_list_0))
cat(mean(test_auc_list_0))
cat(sd(test_auc_list_0))

ustek_nums_6_month_response <- as.data.frame(table(num_list_0))
ustek_denums_6_month_response <- as.data.frame(table(denum_list_0))


ustek_nums_6_month_response$values <- ustek_nums_6_month_response$Freq/100
ustek_denums_6_month_response$values <- 
ustek_denums_6_month_response$Freq/100

write.csv(ustek_nums_6_month_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_6_month_metabolites.csv')
write.csv(ustek_denums_6_month_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_6_month_metabolites.csv')


```




### ustek 2 year response metabolites
```{r Ustek 2 year response metabolites}

my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

mtb_cc=merge(my_phenos,all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]

predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]


counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  #print(table(train_IBD_pheno))
  #print(table(test_IBD_pheno))
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_metabolites_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))

ustek_nums_2_years_metabolites <- as.data.frame(table(num_list))
ustek_denums_2_years_metabolites  <- as.data.frame(table(denum_list))


ustek_nums_2_years_metabolites$values <- 
ustek_nums_2_years_metabolites$Freq/100
ustek_denums_2_years_metabolites$values <- 
ustek_denums_2_years_metabolites$Freq/100

write.csv(ustek_nums_2_years_metabolites, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_2_years_metabolites.csv')
write.csv(ustek_denums_2_years_metabolites, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_2_years_metabolites.csv')

```





### Ustek HBI SCCAI response metabolites
```{r Ustek HBI SCCAI scores response metabolites}
#response_score_2year <- 
merged_taxa_species_clinical_ustek$Response_2years
#response_score_HBI_SCCAI <- 
merged_taxa_species_clinical_ustek$Response_HBI_SCCAI
#response_score_biochemical <- 
merged_taxa_species_clinical_ustek$Response_biochemical
#response_score_biochemical_OR <- 
merged_taxa_species_clinical_ustek$Response_biochemical_OR
#response_score_FCP <- merged_taxa_species_clinical_ustek$Response_FCP



my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]
#,c("run_day_cat","Amount_sample_gram", 
"metabolon_Month_in_freezer","LC.COLUMN","host_Age","host_BMI","host_Sex","clinical_ChrA","clinical_HBD2", 
"clinical_Calprot200","clinical_BowelMovementADayDef", "IBD")

#my_phenos$IBD[my_phenos$IBD==1]="Control"
#my_phenos$IBD[my_phenos$IBD==2]="IBD"
#my_phenos$V1_ResponseScoreYesNo[my_phenos$V1_ResponseScoreYesNo=="yes"]=1
#my_phenos$V1_ResponseScoreYesNo[my_phenos$V1_ResponseScoreYesNo=="no"]=0
mtb_cc=merge(my_phenos,all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]


counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  #print(table(train_IBD_pheno))
  #print(table(test_IBD_pheno))
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_HBI_metabolites_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))

ustek_nums_HBI_SCCAI_metabolites <- as.data.frame(table(num_list))
ustek_denums_HBI_SCCAI_metabolites  <- as.data.frame(table(denum_list))


ustek_nums_HBI_SCCAI_metabolites$values <- 
ustek_nums_HBI_SCCAI_metabolites$Freq/100
ustek_denums_HBI_SCCAI_metabolites$values <- 
ustek_denums_HBI_SCCAI_metabolites$Freq/100




write.csv(ustek_nums_HBI_SCCAI_metabolites, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_HBI_SCCAI_metabolites.csv')
write.csv(ustek_denums_HBI_SCCAI_metabolites, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_HBI_SCCAI_metabolites.csv')


```


### Ustek 6 month microbes
```{r Ustek 6 month response microbes}

my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

taxa_cc=merge(my_phenos,Ustek_microbiome_sub_counts,by="row.names")
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}
ustek_6_month_microbes_values<- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_6_month_microbes<- as.data.frame(table(num_list))
ustek_denums_6_month_microbes  <- as.data.frame(table(denum_list))


ustek_nums_6_month_microbes$values <- ustek_nums_6_month_microbes$Freq/100
ustek_denums_6_month_microbes$values <- 
ustek_denums_6_month_microbes$Freq/100




write.csv(ustek_nums_6_month_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_6_month_microbes.csv')
write.csv(ustek_denums_6_month_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_6_month_microbes.csv')

```


### Ustek 2 year microbes
```{r  Ustek 2 year response microbes}


my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]



taxa_cc=merge(my_phenos,Ustek_microbiome_sub_counts,by="row.names")
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]

predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_2_year_microbes.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_2_year_microbes.csv')

```


### Ustek HBI SCCAI microbes
```{r  Ustek HBI SCCAI scores response microbes} 



my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

taxa_cc=merge(my_phenos,Ustek_microbiome_sub_counts,by="row.names")
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]


counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  #print(table(train_IBD_pheno))
  #print(table(test_IBD_pheno))
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_HBI_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_HBI_microbes<- as.data.frame(table(num_list))
ustek_denums_HBI_microbes  <- as.data.frame(table(denum_list))


ustek_nums_HBI_microbes$values <- ustek_nums_HBI_microbes$Freq/100
ustek_denums_HBI_microbes$values <- ustek_denums_HBI_microbes$Freq/100




write.csv(ustek_nums_HBI_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_HBI_microbes.csv')
write.csv(ustek_denums_HBI_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_HBI_microbes.csv')

```


#### Ustek result table
```{r Ustek result table}


Ustek_predictions <- 
as.data.frame(t(data.frame("ustek_6_month_metabolites_values" = 
ustek_6_month_metabolites_values,
                     "ustek_2_year_metabolites_values" = 
ustek_2_year_metabolites_values,
                     "ustek_HBI_metabolites_values" = 
ustek_HBI_metabolites_values,
                     "ustek_6_month_microbes_values" = 
ustek_6_month_microbes_values,
                     "ustek_2_year_microbes_values" = 
ustek_2_year_microbes_values,
                     "ustek_HBI_microbes_values" = 
ustek_HBI_microbes_values)))

colnames(Ustek_predictions) <- c("train_auc", "train_auc_sd", "test_auc", 
"test_auc_sd")
print(Ustek_predictions)
write.csv(Ustek_predictions, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/Ustek_predictions_table.csv')
```


# now we run it for the combined cohort

#### full cohort data prep
```{r}
#Prepare data#

taxa_df <- 
as.data.frame(t(read.csv("/Users/iwan/Research/drug_response/metadata/lastest_metadata/complete_metaphlan_merged.txt", 
sep='\t',row.names = 1)))

taxa_df$UNCLASSIFIED <- NULL

taxa_df <- as.data.frame(t(taxa_df))
taxa_df$NCBI_tax_id <- NULL


# ustek microbiome prep start to finish


Ust_taxa_df <- taxa_df[, grep("Ust", colnames(taxa_df))]
Ust_taxa_df_timepoint_one <-Ust_taxa_df[, grep("_1_", 
colnames(Ust_taxa_df))]
ust_saved_rownames <- colnames(Ust_taxa_df_timepoint_one)
updated_ust_rownames <- str_split_i(ust_saved_rownames, '_', 1)
Ust_taxa_df_timepoint_one_input <- 
as.data.frame(t(Ust_taxa_df_timepoint_one))
Ust_taxa_df_timepoint_one_input<- 
data.frame(lapply(Ust_taxa_df_timepoint_one_input, as.numeric))

# vedo microbiome start to finish
Vedo_metabolites <- 
read.csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/metabolites_vedo_clin.txt')
vedo_t1_samples <- c()
columnsnames_vedo <- c()
vedo_newnames <- c()
for (i in Vedo_metabolites$Fecal_sample_ID_1){
  if (i != ""){
    columnsnames_vedo <- c(columnsnames_vedo, i)
    vedo_t1_samples <- c(vedo_t1_samples, colnames(taxa_df)[grep(i, 
colnames(taxa_df))])
    vedo_newnames <-c(vedo_newnames, 
trimws(toupper(Vedo_metabolites[Vedo_metabolites$Fecal_sample_ID_1 == 
i,]$Study_ID)))
  }
  
}


#merge(Vedo_metabolites, Taxa_DR_species, by.x = "Fecal_sample_ID_1", by.y 
= 0)



Vedo_taxa_df <- taxa_df[,vedo_t1_samples]
colnames(Vedo_taxa_df) <-  columnsnames_vedo
vedo_saved_rownames <- colnames(Vedo_taxa_df)

Vedo_taxa_df_input <- as.data.frame(t(Vedo_taxa_df))
Vedo_taxa_df_input<- data.frame(lapply(Vedo_taxa_df_input, as.numeric))


Ust_vedo_microbes <- rbind(Vedo_taxa_df_input, 
Ust_taxa_df_timepoint_one_input)

#rownames(Ust_vedo_microbes) <- 

Taxa_DR_species <- filterMetaGenomeDF(Ust_vedo_microbes, presPerc = 
0.15,minMRelAb = 0.1 ,minMedRelAb=0.0,
                                      rescaleTaxa=T,verbose=T,
                                      keepDomains=c('Bacteria','Archaea'),
                                      keepLevels=c('S'),
                                      keepUnknown=F) 






combined_rownames <- c(vedo_newnames, updated_ust_rownames)




row.names(Taxa_DR_species) <- combined_rownames
Taxa_DR_species <- na.omit(Taxa_DR_species)

#Taxa_DR_species
Taxa_DR_species <- na.omit(Taxa_DR_species)







# load metabolites
Ustek_metabolites <- 
read.csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/metabolites_ustek_clin.txt')


combined_clinical_metabolites <- rbind.fill(Ustek_metabolites, 
Vedo_metabolites)


combined_clinical_metabolites$X <- NULL

starting_rownames <- combined_clinical_metabolites$sample_id_right
starting_rownames_v1 <- gsub('USTEK','Ust',starting_rownames)
starting_rownames_v2 <- gsub('t00','t',starting_rownames_v1)
starting_rownames_v3 <- gsub('t0','t',starting_rownames_v2)





rownames(combined_clinical_metabolites) <- starting_rownames_v3





All_clinical <- combined_clinical_metabolites[, 1:697]

All_metabolites_only <- combined_clinical_metabolites[,698:2102]


response_score <- combined_clinical_metabolites$ResponseScoreYesNo





merged_taxa_species_clinical_combined <- 
merge(combined_clinical_metabolites, Taxa_DR_species, by.x = 0, by.y = 0)
# merged_taxa_species_clinical_vedo <- merge(Vedo_metabolites, 
Taxa_DR_species, by.x = "Fecal_sample_ID_1", by.y = 0)

#merged_taxa_species_clinical_combined <- 
merged_taxa_species_clinical_combined[merged_taxa_species_clinical_combined$ResponseScoreYesNo 
!= 'not_known',]


write.csv(merged_taxa_species_clinical_combined, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_merged_everything.csv')


All_clinical_data <- merged_taxa_species_clinical_combined[, 1:698]
combined_metabolites_only <- 
merged_taxa_species_clinical_combined[,704:2107]
combined_microbiome_only <- 
merged_taxa_species_clinical_combined[,2121:2185]






# correct names
ustek_Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_updated_read_counts.csv', 
sep =';', header=TRUE)
ustek_Microbiome_counts <- ustek_Microbiome_counts %>%
  filter(!grepl("_2$", Sample))



ustek_Microbiome_counts$Sample <- 
gsub('_1','',ustek_Microbiome_counts$Sample) 

rownames(ustek_Microbiome_counts) <- ustek_Microbiome_counts$Sample


vedo_Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/Vedo_read_counts.txt', 
sep ='\t', header=TRUE)


vedo_count_newnames <- c()
for (i in vedo_Microbiome_counts$Sample){
  if (i != ""){
    short_string <- paste(as.vector(str_split_1(i, '_')[1:3]), collapse = 
"_" )
    vedo_count_newnames <- c(vedo_count_newnames, 
trimws(toupper(Vedo_metabolites[Vedo_metabolites$Fecal_sample_ID_1 == 
short_string,]$Study_ID)))
    if (identical(Vedo_metabolites[Vedo_metabolites$Fecal_sample_ID_1 == 
short_string,]$Study_ID, character(0))){
      vedo_count_newnames <- c(vedo_count_newnames, short_string)
    }    

  }
  
}
rownames(vedo_Microbiome_counts) <- vedo_count_newnames
vedo_Microbiome_counts$filter <- vedo_count_newnames

vedo_Microbiome_counts <- vedo_Microbiome_counts %>%
  filter(!grepl("Vedo", filter))

vedo_Microbiome_counts$filter <- NULL

all_read_counts <- rbind(ustek_Microbiome_counts, vedo_Microbiome_counts)



# working till here 
rownames(combined_microbiome_only) <- 
merged_taxa_species_clinical_combined$Row.names


combined_microbiome_only_with_counts <-  merge(combined_microbiome_only, 
all_read_counts, by.x=0, by.y=0)

combined_microbiome_only_with_counts$merging_column <- NULL
combined_microbiome_only_with_counts$Fecal_sample_ID_1 <- NULL
combined_microbiome_only_with_counts$Sample <- NULL
combined_microbiome_only_with_counts$Raw.reads..p1. <- NULL
combined_microbiome_only_with_counts$Raw.reads..p2. <- NULL
combined_microbiome_only_with_counts$Final..p2. <- NULL
combined_microbiome_only_with_counts$Fecal_sample_ID_1 <- NULL
combined_microbiome_only_with_counts$Row.names <- NULL

combined_microbiome_sub_counts <- combined_microbiome_only_with_counts %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))


combined_microbiome_sub_counts$Final..p1. <- NULL

combined_microbiome_sub_counts[combined_microbiome_sub_counts == 0] <- NA


combined_microbiome_sub_counts[sapply(combined_microbiome_sub_counts, 
is.numeric)] <- 
lapply(combined_microbiome_sub_counts[sapply(combined_microbiome_sub_counts, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))


combined_all_new_ID_raw_filt=combined_metabolites_only[,((colSums(!is.na(combined_metabolites_only)) 
/ nrow(combined_metabolites_only)) *100 )>70]
#all_new_ID_raw[is.na(all_new_ID_raw)]=0
combined_all_new_ID_raw_filt[sapply(combined_all_new_ID_raw_filt, 
is.numeric)] <- 
lapply(combined_all_new_ID_raw_filt[sapply(combined_all_new_ID_raw_filt, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))


```


#### combined 6 month response metabolites
```{r combined 6 month response metabolites}

my_phenos=All_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

mtb_cc=merge(my_phenos,combined_all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list_0 <- c()
num_list_0 <- c()
denum_list_0 <- c()
AUC_list_0 <- c()
test_auc_list_0 <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )
  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list_0 <- c(num_list_0, (nums))
    denum_list_0 <- c(denum_list_0, (denums))
    test_auc_list_0 <- c(test_auc_list_0, test_auc)
    AUC_list_0 <- c(AUC_list_0, AUC)
    acc_list_0 <- c(acc_list_0, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
      
    }
  )
  if(inherits(possibleError, "error")) next
}

all_6_month_metabolites_values <- c(mean(AUC_list_0), sd(AUC_list_0), 
mean(test_auc_list_0), sd(test_auc_list_0))
cat(mean(AUC_list_0))
cat(sd(AUC_list_0))
cat(mean(test_auc_list_0))
cat(sd(test_auc_list_0))

all_nums_6_month_response <- as.data.frame(table(num_list_0))
all_denums_6_month_response <- as.data.frame(table(denum_list_0))


all_nums_6_month_response$values <- all_nums_6_month_response$Freq/100
all_denums_6_month_response$values <- all_denums_6_month_response$Freq/100

write.csv(all_nums_6_month_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_6_month_metabolites.csv')
write.csv(all_denums_6_month_response, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_6_month_metabolites.csv')


```




#### combined 2 year response metabolites
```{r combined 2 year response metabolites}

my_phenos=All_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

mtb_cc=merge(my_phenos,combined_all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (verhuisd)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (overleden)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "?", ]

predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]



counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  #print(table(train_IBD_pheno))
  #print(table(test_IBD_pheno))
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

all_2_year_metabolites_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))

all_nums_2_years_metabolites <- as.data.frame(table(num_list))
all_denums_2_years_metabolites  <- as.data.frame(table(denum_list))


all_nums_2_years_metabolites$values <- 
all_nums_2_years_metabolites$Freq/100
all_denums_2_years_metabolites$values <- 
all_denums_2_years_metabolites$Freq/100

write.csv(all_nums_2_years_metabolites, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_2_years_metabolites.csv')
write.csv(all_denums_2_years_metabolites, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_2_years_metabolites.csv')

```





#### Combined HBI SCCAI response metabolites
```{r Combined HBI SCCAI response metabolites}

my_phenos=All_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]
#,c("run_day_cat","Amount_sample_gram", 
"metabolon_Month_in_freezer","LC.COLUMN","host_Age","host_BMI","host_Sex","clinical_ChrA","clinical_HBD2", 
"clinical_Calprot200","clinical_BowelMovementADayDef", "IBD")

#my_phenos$IBD[my_phenos$IBD==1]="Control"
#my_phenos$IBD[my_phenos$IBD==2]="IBD"
#my_phenos$V1_ResponseScoreYesNo[my_phenos$V1_ResponseScoreYesNo=="yes"]=1
#my_phenos$V1_ResponseScoreYesNo[my_phenos$V1_ResponseScoreYesNo=="no"]=0
mtb_cc=merge(my_phenos,combined_all_new_ID_raw_filt,by="row.names")
row.names(mtb_cc)=mtb_cc$Row.names
mtb_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=mtb_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]



counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  #print(table(train_IBD_pheno))
  #print(table(test_IBD_pheno))
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    
    #cat("Test set AUC =", pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, 
quiet=T)))
    ##cat("Classification accuracy on test set =", round(mean(IBD_Hat2 == 
test_IBD_pheno), 2))
    #print (IBD_model)
    #plot(IBD_model)
    #plotROC(IBD_model)
    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

all_HBI_metabolites_values <-  c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))

all_nums_HBI_SCCAI_metabolites <- as.data.frame(table(num_list))
all_denums_HBI_SCCAI_metabolites  <- as.data.frame(table(denum_list))


all_nums_HBI_SCCAI_metabolites$values <- 
all_nums_HBI_SCCAI_metabolites$Freq/100
all_denums_HBI_SCCAI_metabolites$values <- 
all_denums_HBI_SCCAI_metabolites$Freq/100




write.csv(all_nums_HBI_SCCAI_metabolites, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_HBI_SCCAI_metabolites.csv')
write.csv(all_denums_HBI_SCCAI_metabolites, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_HBI_SCCAI_metabolites.csv')


```





#### Combined 6 month microbes
```{r Combined 6 month microbes}

my_phenos=All_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

taxa_cc=merge(my_phenos,combined_microbiome_sub_counts,by="row.names")
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]

counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}
all_6_month_microbes_values <-  c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





all_nums_6_month_microbes<- as.data.frame(table(num_list))
all_denums_6_month_microbes  <- as.data.frame(table(denum_list))


all_nums_6_month_microbes$values <- all_nums_6_month_microbes$Freq/100
all_denums_6_month_microbes$values <- all_denums_6_month_microbes$Freq/100




write.csv(all_nums_6_month_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_6_month_microbes.csv')
write.csv(all_denums_6_month_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_6_month_microbes.csv')

```





#### Combined 2 year microbes
```{r Combined 2 year microbes}


my_phenos=All_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]



taxa_cc=merge(my_phenos,combined_microbiome_sub_counts,by="row.names")
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (verhuisd)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (overleden)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "?", ]

predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  #print(table(train_IBD_pheno))
  #print(table(test_IBD_pheno))
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

all_2_year_microbes_values <-  c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





all_nums_2_year_microbes<- as.data.frame(table(num_list))
all_denums_2_year_microbes  <- as.data.frame(table(denum_list))


all_nums_2_year_microbes$values <- all_nums_2_year_microbes$Freq/100
all_denums_2_year_microbes$values <- all_denums_2_year_microbes$Freq/100




write.csv(all_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_2_year_microbes.csv')
write.csv(all_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_2_year_microbes.csv')

```




#### Combined HBI SCCAI microbes
```{r Combined HBI SCCAI microbes}



my_phenos=All_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

taxa_cc=merge(my_phenos,combined_microbiome_sub_counts,by="row.names")
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,12:ncol(mtb_cc2)]



counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  #print(table(train_IBD_pheno))
  #print(table(test_IBD_pheno))
  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

all_HBI_microbes_values <-  c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





all_nums_HBI_microbes<- as.data.frame(table(num_list))
all_denums_HBI_microbes  <- as.data.frame(table(denum_list))


all_nums_HBI_microbes$values <- all_nums_HBI_microbes$Freq/100
all_denums_HBI_microbes$values <- all_denums_HBI_microbes$Freq/100




write.csv(all_nums_HBI_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_HBI_microbes.csv')
write.csv(all_denums_HBI_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_HBI_microbes.csv')

```


#### Combined output table
```{r Combined output table}


all_predictions <- 
as.data.frame(t(data.frame("all_6_month_metabolites_values" = 
all_6_month_metabolites_values,
                     "all_2_year_metabolites_values" = 
all_2_year_metabolites_values,
                     "all_HBI_metabolites_values" = 
all_HBI_metabolites_values,
                     "all_6_month_microbes_values" = 
all_6_month_microbes_values,
                     "all_2_year_microbes_values" = 
all_2_year_microbes_values,
                     "all_HBI_microbes_values" = 
all_HBI_microbes_values)))

colnames(all_predictions) <- c("train_auc", "train_auc_sd", "test_auc", 
"test_auc_sd")
print(all_predictions)
write.csv(all_predictions, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_cohort_predictions_table.csv')
```



#plot preprocessing

```{r}

library(ggplot2)

# install.packages("reshape")
library(reshape)

# Data 
#set.seed(8)
#m <- matrix(round(rnorm(200), 2), 10, 10)
#colnames(m) <- paste("Col", 1:10)
#rownames(m) <- paste("Row", 1:10)

# Transform the matrix in long format
# ustek data
bug6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_6_month_microbes.csv')
bug6_num$type <- 'Numerator'
bug6_num$feature <- 'Microbe'
bug6_num$response <- '6 month'
bug6_num$medication <- 'Ustekinumab'
bug6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_6_month_microbes.csv')
bug6_denum$type <- 'Denominator'
bug6_denum$feature <- 'Microbe'
bug6_denum$response <- '6 month'
bug6_denum$medication <- 'Ustekinumab'
bug2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_2_year_microbes.csv')
bug2_num$type <- 'Numerator'
bug2_num$feature <- 'Microbe'
bug2_num$response <- '2 year'
bug2_num$medication <- 'Ustekinumab'
bug2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_2_year_microbes.csv')
bug2_denum$type <- 'Denominator'
bug2_denum$feature <- 'Microbe'
bug2_denum$response <- '2 year'
bug2_denum$medication <- 'Ustekinumab'
bughbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_HBI_microbes.csv')
bughbi_num$type <- 'Numerator'
bughbi_num$feature <- 'Microbe'
bughbi_num$response <- 'HBI SCCAI'
bughbi_num$medication <- 'Ustekinumab'
bughbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_HBI_microbes.csv')
bughbi_denum$type <- 'Denominator'
bughbi_denum$feature <- 'Microbe'
bughbi_denum$response <- 'HBI SCCAI'
bughbi_denum$medication <- 'Ustekinumab'



# vedo data
vedo_bug6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_6_month_microbes.csv')
vedo_bug6_num$type <- 'Numerator'
vedo_bug6_num$feature <- 'Microbe'
vedo_bug6_num$response <- '6 month'
vedo_bug6_num$medication <- 'Vedolizumab'
vedo_bug6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_6_month_microbes.csv')
vedo_bug6_denum$type <- 'Denominator'
vedo_bug6_denum$feature <- 'Microbe'
vedo_bug6_denum$response <- '6 month'
vedo_bug6_denum$medication <- 'Vedolizumab'
vedo_bug2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_2_year_microbes.csv')
vedo_bug2_num$type <- 'Numerator'
vedo_bug2_num$feature <- 'Microbe'
vedo_bug2_num$response <- '2 year'
vedo_bug2_num$medication <- 'Vedolizumab'
vedo_bug2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_2_year_microbes.csv')
vedo_bug2_denum$type <- 'Denominator'
vedo_bug2_denum$feature <- 'Microbe'
vedo_bug2_denum$response <- '2 year'
vedo_bug2_denum$medication <- 'Vedolizumab'
vedo_bughbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_HBI_microbes.csv')
vedo_bughbi_num$type <- 'Numerator'
vedo_bughbi_num$feature <- 'Microbe'
vedo_bughbi_num$response <- 'HBI SCCAI'
vedo_bughbi_num$medication <- 'Vedolizumab'
vedo_bughbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_HBI_microbes.csv')
vedo_bughbi_denum$type <- 'Denominator'
vedo_bughbi_denum$feature <- 'Microbe'
vedo_bughbi_denum$response <- 'HBI SCCAI'
vedo_bughbi_denum$medication <- 'Vedolizumab'



# combined data
all_bug6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_6_month_microbes.csv')
all_bug6_num$type <- 'Numerator'
all_bug6_num$feature <- 'Microbe'
all_bug6_num$response <- '6 month'
all_bug6_num$medication <- 'Combined cohort'
all_bug6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_6_month_microbes.csv')
all_bug6_denum$type <- 'Denominator'
all_bug6_denum$feature <- 'Microbe'
all_bug6_denum$response <- '6 month'
all_bug6_denum$medication <- 'Combined cohort'
all_bug2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_2_year_microbes.csv')
all_bug2_num$type <- 'Numerator'
all_bug2_num$feature <- 'Microbe'
all_bug2_num$response <- '2 year'
all_bug2_num$medication <- 'Combined cohort'
all_bug2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_2_year_microbes.csv')
all_bug2_denum$type <- 'Denominator'
all_bug2_denum$feature <- 'Microbe'
all_bug2_denum$response <- '2 year'
all_bug2_denum$medication <- 'Combined cohort'
all_bughbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_HBI_microbes.csv')
all_bughbi_num$type <- 'Numerator'
all_bughbi_num$feature <- 'Microbe'
all_bughbi_num$response <- 'HBI SCCAI'
all_bughbi_num$medication <- 'Combined cohort'
all_bughbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_HBI_microbes.csv')
all_bughbi_denum$type <- 'Denominator'
all_bughbi_denum$feature <- 'Microbe'
all_bughbi_denum$response <- 'HBI SCCAI'
all_bughbi_denum$medication <- 'Combined cohort'






# ustek again
met6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_6_month_metabolites.csv')
met6_num$type <- 'Numerator'
met6_num$feature <- 'Metabolite'
met6_num$response <- '6 month'
met6_num$medication <- 'Ustekinumab'
met6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_6_month_metabolites.csv')
met6_denum$type <- 'Denominator'
met6_denum$feature <- 'Metabolite'
met6_denum$response <- '6 month'
met6_denum$medication <- 'Ustekinumab'
met2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_2_years_metabolites.csv')
met2_num$type <- 'Numerator'
met2_num$feature <- 'Metabolite'
met2_num$response <- '2 year'
met2_num$medication <- 'Ustekinumab'
met2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_2_years_metabolites.csv')
met2_denum$type <- 'Denominator'
met2_denum$feature <- 'Metabolite'
met2_denum$response <- '2 year'
met2_denum$medication <- 'Ustekinumab'
methbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_HBI_SCCAI_metabolites.csv')
methbi_num$type <- 'Numerator'
methbi_num$feature <- 'Metabolite'
methbi_num$response <- 'HBI SCCAI'
methbi_num$medication <- 'Ustekinumab'
methbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_HBI_SCCAI_metabolites.csv')
methbi_denum$type <- 'Denominator'
methbi_denum$feature <- 'Metabolite'
methbi_denum$response <- 'HBI SCCAI'
methbi_denum$medication <- 'Ustekinumab'



# vedo again 
vedo_met6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_6_month_metabolites.csv')
vedo_met6_num$type <- 'Numerator'
vedo_met6_num$feature <- 'Metabolite'
vedo_met6_num$response <- '6 month'
vedo_met6_num$medication <- 'Vedolizumab'
vedo_met6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_6_month_metabolites.csv')
vedo_met6_denum$type <- 'Denominator'
vedo_met6_denum$feature <- 'Metabolite'
vedo_met6_denum$response <- '6 month'
vedo_met6_denum$medication <- 'Vedolizumab'
vedo_met2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_2_year_metabolites.csv')
vedo_met2_num$type <- 'Numerator'
vedo_met2_num$feature <- 'Metabolite'
vedo_met2_num$response <- '2 year'
vedo_met2_num$medication <- 'Vedolizumab'
vedo_met2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_2_year_metabolites.csv')
vedo_met2_denum$type <- 'Denominator'
vedo_met2_denum$feature <- 'Metabolite'
vedo_met2_denum$response <- '2 year'
vedo_met2_denum$medication <- 'Vedolizumab'
vedo_methbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_HBI_metabolites.csv')
vedo_methbi_num$type <- 'Numerator'
vedo_methbi_num$feature <- 'Metabolite'
vedo_methbi_num$response <- 'HBI SCCAI'
vedo_methbi_num$medication <- 'Vedolizumab'
vedo_methbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_HBI_metabolites.csv')
vedo_methbi_denum$type <- 'Denominator'
vedo_methbi_denum$feature <- 'Metabolite'
vedo_methbi_denum$response <- 'HBI SCCAI'
vedo_methbi_denum$medication <- 'Vedolizumab'


# combined again 
all_met6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_6_month_metabolites.csv')
all_met6_num$type <- 'Numerator'
all_met6_num$feature <- 'Metabolite'
all_met6_num$response <- '6 month'
all_met6_num$medication <- 'Combined cohort'
all_met6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_6_month_metabolites.csv')
all_met6_denum$type <- 'Denominator'
all_met6_denum$feature <- 'Metabolite'
all_met6_denum$response <- '6 month'
all_met6_denum$medication <- 'Combined cohort'
all_met2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_2_years_metabolites.csv')
all_met2_num$type <- 'Numerator'
all_met2_num$feature <- 'Metabolite'
all_met2_num$response <- '2 year'
all_met2_num$medication <- 'Combined cohort'
all_met2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_2_years_metabolites.csv')
all_met2_denum$type <- 'Denominator'
all_met2_denum$feature <- 'Metabolite'
all_met2_denum$response <- '2 year'
all_met2_denum$medication <- 'Combined cohort'
all_methbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_nums_HBI_SCCAI_metabolites.csv')
all_methbi_num$type <- 'Numerator'
all_methbi_num$feature <- 'Metabolite'
all_methbi_num$response <- 'HBI SCCAI'
all_methbi_num$medication <- 'Combined cohort'
all_methbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/all_denums_HBI_SCCAI_metabolites.csv')
all_methbi_denum$type <- 'Denominator'
all_methbi_denum$feature <- 'Metabolite'
all_methbi_denum$response <- 'HBI SCCAI'
all_methbi_denum$medication <- 'Combined cohort'






#read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_6_month_microbes.csv')
colnames(bug6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(bug6_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(bug2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(bug2_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(bughbi_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(bughbi_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(met6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(met6_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(met2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(met2_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(methbi_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(methbi_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")




colnames(vedo_bug6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_bug6_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_bug2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_bug2_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_bughbi_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_bughbi_denum) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")
colnames(vedo_met6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_met6_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_met2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_met2_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_methbi_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_methbi_denum) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")





colnames(all_bug6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_bug6_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_bug2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_bug2_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_bughbi_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_bughbi_denum) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")
colnames(all_met6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_met6_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_met2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_met2_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_methbi_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_methbi_denum) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")



stacked_df_bugs <- do.call(rbind, list(bug6_num, bug6_denum, bug2_num, 
bug2_denum, bughbi_num, bughbi_denum, 
                                       vedo_bug6_num, vedo_bug6_denum, 
vedo_bug2_num, vedo_bug2_denum, vedo_bughbi_num, vedo_bughbi_denum,
                                       all_bug6_num, all_bug6_denum, 
all_bug2_num, all_bug2_denum, all_bughbi_num, all_bughbi_denum))

stacked_df_metas <- do.call(rbind, list(met6_num, met6_denum, 
met2_num,met2_denum,methbi_num, methbi_denum,
                                        vedo_met6_num, vedo_met6_denum, 
vedo_met2_num,vedo_met2_denum,vedo_methbi_num, vedo_methbi_denum,
                                        all_met6_num, all_met6_denum, 
all_met2_num,all_met2_denum,all_methbi_num, all_methbi_denum))






metabolite_names_df <- 
read.csv("/Users/iwan/Research/drug_response/metadata/lastest_metadata/metabolites_naming_key.txt", 
sep='\t', header = TRUE)
metabolite_names <- c()
for (i in stacked_df_metas$predict){
  metabolite_names <- c(metabolite_names, 
metabolite_names_df$CHEMICAL_NAME[metabolite_names_df$XCHEM_ID == i])
}


shortnames <- c()
for (i in stacked_df_bugs$predict){
  shortnames <- c(shortnames, str_split_i(i, '__',8))
}

stacked_df_bugs$new_predict <- shortnames
stacked_df_metas$new_predict <- metabolite_names

stacked_df <- rbind(stacked_df_bugs, stacked_df_metas)

stacked_df_filtered <- stacked_df[stacked_df$values > 0.3,]

```


# plotting
```{r}

# install.packages("ggplot2")
library(ggplot2)
library(ggh4x)

#color = "white",
#l#wd = 1.5,
#linetype = 1
pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/all_predictive_features.pdf", 
width=10)

bp <- ggplot(stacked_df_filtered, aes(x = response, y = new_predict, fill 
= values)) +
  geom_tile() + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_grid(( factor(feature, levels=c('Microbe', 'Metabolite'))~  
Medication +factor(type, levels=c('Numerator', 'Denominator')) ), scales = 
"free", space = "free") +
  ggtitle("Frequency of features use in CoDaCoRe predictions") +
  xlab("Response definition")+
  ylab("Features") #, title = "Features used in Ustek response 
definition") 


bp
dev.off()
stacked_df_filtered_combined_only <- 
stacked_df_filtered[stacked_df_filtered$Medication == 'Combined cohort',]
stacked_df_filtered_combined_only_6_month <-  
stacked_df_filtered_combined_only[stacked_df_filtered_combined_only$response 
== '6 month',]





bp <- ggplot(stacked_df_filtered_combined_only_6_month, aes(x = response, 
y = new_predict, fill = values)) +
  geom_tile() + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  facet_nested(( factor(feature, levels=c('Microbe', 'Metabolite'))~  
factor(Medication, levels =c('Vedolizumab', 'Ustekinumab', 'Combined 
cohort')) +factor(type, levels=c('Numerator', 'Denominator')) ), scales = 
"free", space = "free") +
  ggtitle("Frequency of features use in CoDaCoRe predictions") +
  xlab("Response definition")+
  ylab("Features") #, title = "Features used in Ustek response 
definition") 


bp
stacked_df_bugs_combined_only <- 
stacked_df_bugs[stacked_df_bugs$Medication == 'Combined cohort',]
stacked_df_bugs_combined_only_6_month <-  
stacked_df_bugs_combined_only[stacked_df_bugs_combined_only$response == '6 
month',]


bp <- 
ggplot(stacked_df_bugs_combined_only_6_month[stacked_df_bugs_combined_only_6_month$values 
> 0.3,], aes(x = response, y = reorder(new_predict, values), fill = 
values)) +
  geom_tile() + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  facet_nested(( ~  factor(Medication, levels =c('Vedolizumab', 
'Ustekinumab', 'Combined cohort')) +factor(type, levels=c('Numerator', 
'Denominator')) ), scales = "free", space = "free") +
  ggtitle("Frequency of microbes used in CoDaCoRe predictions") +
  xlab("Response definition")+
  labs(fill="Frequency")+
  ylab("Features") #, title = "Features used in Ustek response 
definition") 


bp

stacked_meta_combined_only <- stacked_df_metas[stacked_df_metas$Medication 
== 'Combined cohort',]
stacked_meta_combined_only_6_month <-  
stacked_meta_combined_only[stacked_meta_combined_only$response == '6 
month',]
#x=reorder(name,depth)
bp <- 
ggplot(stacked_meta_combined_only_6_month[stacked_meta_combined_only_6_month$values 
> 0.3,], aes(x = response, y = reorder(new_predict, values), fill = 
values)) +
  geom_tile() + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  facet_nested(( ~  factor(Medication, levels =c('Vedolizumab', 
'Ustekinumab', 'Combined cohort')) +factor(type, levels=c('Numerator', 
'Denominator')) ), scales = "free", space = "free") +
  ggtitle("Frequency of metabolites used in CoDaCoRe predictions") +
  xlab("Response definition")+
  labs(fill="Frequency")+
  ylab("Features") #, title = "Features used in Ustek response 
definition") 


bp

```






# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]

stacked_df_bugs_ustek <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
stacked_df_bugs_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]






stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',] 

vedo_responders_microbes_denum <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 6 month response to 
Vedolizumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```


# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]


stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]




vedo_responders_microbes_denum <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 2 year response to Vedolizumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```



# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]

stacked_df_bugs_ustek <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
stacked_df_bugs_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]






stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]


vedo_responders_microbes_denum <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview HBI SCCAI response to 
Vedolizumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```

# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]





stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',] 


vedo_responders_microbes_denum <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 6 month response to 
Ustekinumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```


# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]


stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',] 
vedo_responders_microbes_denum <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 2 year response to Ustekinumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```



# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]

stacked_df_bugs_ustek <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
stacked_df_bugs_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]



stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',] 

vedo_responders_microbes_denum <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview HBI SCCAI response to 
Ustekinumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```











# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]


stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_num <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',]





vedo_responders_microbes_denum <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)
pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/6_month_microbiome.pdf")
# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 6 month response to the combined 
cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)

dev.off()

responder_ratio_microbes <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio_microbes  <- rowSums(vedo_non_responders_microbes_num) 
/ rowSums(vedo_non_responders_microbes_denum)




```


# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]


stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',] 




vedo_responders_microbes_denum <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 2 year response to the combined 
cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```



# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]

stacked_df_bugs_ustek <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
stacked_df_bugs_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]



stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',]



vedo_responders_microbes_denum <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)
library(ggsignif)
# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview HBI SCCAI response to the 
combined cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

print(p)




```






# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]

stacked_df_bugs_ustek <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
stacked_df_bugs_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]





stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_num <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]





vedo_responders_microbes_denum <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Metabolite ratio overview 6 month response to 
Vedolizumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```


# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]


stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',] 


vedo_responders_microbes_denum <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Metabolite ratio overview 2 year response to Vedolizumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```



# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]

stacked_df_bugs_ustek <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
stacked_df_bugs_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]






stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]

vedo_responders_microbes_num <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',] 



vedo_responders_microbes_denum <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][vedo_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Metabolite ratio overview HBI SCCAI response to 
Vedolizumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```

# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]





stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]



vedo_responders_microbes_num <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',] 

vedo_responders_microbes_denum <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title ="Metabolite ratio overview 6 month response to Ustekinumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```


# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]


stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]



vedo_responders_microbes_num <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',] 



vedo_responders_microbes_denum <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Metabolite ratio overview 2 year response to Ustekinumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```



# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]

stacked_df_bugs_ustek <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
stacked_df_bugs_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]



stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]



vedo_responders_microbes_num <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',] 

vedo_responders_microbes_denum <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][Ustek_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Metabolite ratio overview HBI SCCAI response to 
Ustekinumab",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```











# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]


stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]


vedo_responders_microbes_num <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',] 



vedo_responders_microbes_denum <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)
pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/6_month_metabolite.pdf")
# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Metabolite ratio overview 6 month response in the combined 
cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)

dev.off()

responder_ratio_metabolites <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio_metabolites  <- 
rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


```


# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]


stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]


vedo_responders_microbes_num <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',] 




vedo_responders_microbes_denum <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Metabolite ratio overview 2 year response in the combined 
cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))
# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)




```



# plot the box/violin of the individual bugs and metabolites, then 
combined
```{r}
stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]

stacked_df_bugs_ustek <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
stacked_df_bugs_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]



stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]


vedo_responders_microbes_num <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',] 

vedo_non_responders_microbes_num <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',] 



vedo_responders_microbes_denum <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'yes',]

vedo_non_responders_microbes_denum <- 
combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict][All_clinical_data$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)
library(ggsignif)
# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Metabolite ratio overview HBI SCCAI response in the 
combined cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))


print(p)




```

### now we test the ratios in the utrecht cohort: 
``` {r }


utrecht_df <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_cohort_full_df.txt')


clinical_response <- 'Clinical_responder (HBI ≤4 points\nor a reduction 
in the HBI by ≥3 points)'

new_response <- c()
for (row in 1:length(rownames(utrecht_df))){
  if (utrecht_df[row, 'V1_HarveyBradshawScore'] - utrecht_df[row, 
'V3_HarveyBradshawScore'] > 3){
    new_response <- c(new_response, 'Yes')
  } else {
    new_response <- c(new_response, 'No')
  }
    
}




utrecht_df$our_clinical_response <- new_response
month_6_response <-  'Responder_study'

stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]

stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]


vedo_responders_microbes_num <- 
utrecht_df[,stacked_df_bugs_vedo_6month_num$predict][utrecht_df$our_clinical_response 
== 'Yes',] 

vedo_non_responders_microbes_num <- 
utrecht_df[,stacked_df_bugs_vedo_6month_num$predict][utrecht_df$our_clinical_response 
== 'No',] 



vedo_responders_microbes_denum <- 
utrecht_df[,stacked_df_bugs_vedo_6month_denum$predict][utrecht_df$our_clinical_response 
== 'Yes',]

vedo_non_responders_microbes_denum <- 
utrecht_df[,stacked_df_bugs_vedo_6month_denum$predict][utrecht_df$our_clinical_response 
== 'No',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview clinical score response to 
Vedolizumab in Utrecht cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)





```



``` {r }


utrecht_df <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_cohort_full_df.txt')


clinical_response <- 'Clinical_responder (HBI ≤4 points\nor a reduction 
in the HBI by ≥3 points)'
month_6_response <-  'Responder_study'

stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]

stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]


vedo_responders_microbes_num <- 
utrecht_df[,stacked_df_bugs_vedo_6month_num$predict][utrecht_df$Responder_study 
== 'Yes',] 

vedo_non_responders_microbes_num <- 
utrecht_df[,stacked_df_bugs_vedo_6month_num$predict][utrecht_df$Responder_study 
== 'No',] 



vedo_responders_microbes_denum <- 
utrecht_df[,stacked_df_bugs_vedo_6month_denum$predict][utrecht_df$Responder_study 
== 'Yes',]

vedo_non_responders_microbes_denum <- 
utrecht_df[,stacked_df_bugs_vedo_6month_denum$predict][utrecht_df$Responder_study 
== 'No',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 6 month response to Vedolizumab 
in Utrecht cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)





```


``` {r }


utrecht_df <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_cohort_full_df.txt')

utrecht_df <- utrecht_df[utrecht_df$V1_CurrentIBDMedication == 
'Vedolizumab',]
clinical_response <- 'Clinical_responder (HBI ≤4 points\nor a reduction 
in the HBI by ≥3 points)'
month_6_response <-  'Responder_study'


new_response <- c()
for (row in 1:length(rownames(utrecht_df))){
  if (utrecht_df[row, 'V1_HarveyBradshawScore'] - utrecht_df[row, 
'V3_HarveyBradshawScore'] > 3){
    new_response <- c(new_response, 'Yes')
  } else {
    new_response <- c(new_response, 'No')
  }
    
}




utrecht_df$our_clinical_response <- new_response
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]

stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]


vedo_responders_microbes_num <- 
utrecht_df[,stacked_df_bugs_vedo_6month_num$predict][utrecht_df$our_clinical_response 
== 'Yes',] 

vedo_non_responders_microbes_num <- 
utrecht_df[,stacked_df_bugs_vedo_6month_num$predict][utrecht_df$our_clinical_response 
== 'No',] 



vedo_responders_microbes_denum <- 
utrecht_df[,stacked_df_bugs_vedo_6month_denum$predict][utrecht_df$our_clinical_response 
== 'Yes',]

vedo_non_responders_microbes_denum <- 
utrecht_df[,stacked_df_bugs_vedo_6month_denum$predict][utrecht_df$our_clinical_response 
== 'No',]


responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview clinical score response to 
Vedolizumab in Utrecht only Vedolizumab patients",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)





```


# utrecht 6 months combined plot
``` {r }
# add counts for utrecht data 

utrecht_df <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_cohort_full_df.txt')



disease_duration_df <- 
read_delim('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_disease_activity.txt', 
delim = '\t') 

###########

Utrecht_Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/Utrecht_read_counts.txt', 
sep ='\t', header=TRUE)

merge_samples <- c()
ordered_samples <- c()
for (i in utrecht_df$...1){
  merge_samples <- c(merge_samples, i)
  ordered_samples <- c(ordered_samples, 
Utrecht_Microbiome_counts$Sample[grep( i, 
Utrecht_Microbiome_counts$Sample)])
}

rownames(Utrecht_Microbiome_counts) <- Utrecht_Microbiome_counts$Sample
  
Microbiome_counts_filtered <- Utrecht_Microbiome_counts[ordered_samples,]


Microbiome_counts_filtered$merging_column <- merge_samples

#vedo_microbiome_only$Fecal_sample_ID_1 <- 
merged_taxa_species_clinical_vedo$Fecal_sample_ID_1


Utrecht_Microbiome_with_counts <- merge(utrecht_df, 
Microbiome_counts_filtered, by.x = "...1", by.y = "merging_column")

Utrecht_Microbiome_with_counts <- merge(Utrecht_Microbiome_with_counts, 
disease_duration_df, by.x ="...1", by.y = "Fecal_sample_ID")

Utrecht_Microbiome_with_counts$merging_column <- NULL
Utrecht_Microbiome_with_counts$Fecal_sample_ID_1 <- NULL
Utrecht_Microbiome_with_counts$Sample <- NULL
Utrecht_Microbiome_with_counts$Raw.reads..p1. <- NULL
Utrecht_Microbiome_with_counts$Raw.reads..p2. <- NULL
Utrecht_Microbiome_with_counts$Final..p2. <- NULL
Utrecht_Microbiome_with_counts$Fecal_sample_ID_1 <- NULL
Utrecht_Microbiome_with_counts$sample_id_right <- NULL
Utrecht_Microbiome_with_counts_vedo <- 
Utrecht_Microbiome_with_counts[Utrecht_Microbiome_with_counts$V1_CurrentIBDMedication 
== 'Vedolizumab',]
#vedo_microbiome_sub_counts <- Utrecht_Microbiome_with_counts %>%
#  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
#
#vedo_microbiome_sub_counts$Final..p1. <- NULL

#vedo_microbiome_sub_counts[vedo_microbiome_sub_counts == 0] <- NA

#vedo_microbiome_sub_counts[sapply(vedo_microbiome_sub_counts, 
is.numeric)] <- 
lapply(vedo_microbiome_sub_counts[sapply(vedo_microbiome_sub_counts, 
is.numeric)], #function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))



##########










#utrecht_df <- utrecht_df[utrecht_df$V1_CurrentIBDMedication == 
'Vedolizumab',]
clinical_response <- 'Clinical_responder (HBI ≤4 points\nor a reduction 
in the HBI by ≥3 points)'
month_6_response <-  'Responder_study'

stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]

stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]


vedo_responders_microbes_num <- 
Utrecht_Microbiome_with_counts_vedo[,c(stacked_df_bugs_vedo_6month_num$predict, 
'Final..p1.')][Utrecht_Microbiome_with_counts_vedo$Responder_study == 
'Yes',] 
vedo_responders_microbes_num <- vedo_responders_microbes_num %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
vedo_responders_microbes_num$Final..p1. <- NULL
#vedo_responders_microbes_num<- vedo_responders_microbes_num * 
Utrecht_Microbiome_with_counts$Final..p1.
vedo_responders_microbes_num[vedo_responders_microbes_num == 0] <- NA
vedo_responders_microbes_num <- 
vedo_responders_microbes_num[,colSums(is.na(vedo_responders_microbes_num))<nrow(vedo_responders_microbes_num)]
vedo_responders_microbes_num[sapply(vedo_responders_microbes_num, 
is.numeric)] <- 
lapply(vedo_responders_microbes_num[sapply(vedo_responders_microbes_num, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))


vedo_non_responders_microbes_num <- 
Utrecht_Microbiome_with_counts_vedo[,c(stacked_df_bugs_vedo_6month_num$predict, 
'Final..p1.')][Utrecht_Microbiome_with_counts_vedo$Responder_study == 
'No',] 
vedo_non_responders_microbes_num <- vedo_non_responders_microbes_num %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
vedo_non_responders_microbes_num$Final..p1. <- NULL
vedo_non_responders_microbes_num[vedo_non_responders_microbes_num == 0] <- 
NA
vedo_non_responders_microbes_num <- 
vedo_non_responders_microbes_num[,colSums(is.na(vedo_non_responders_microbes_num))<nrow(vedo_non_responders_microbes_num)]
vedo_non_responders_microbes_num[sapply(vedo_non_responders_microbes_num, 
is.numeric)] <- 
lapply(vedo_non_responders_microbes_num[sapply(vedo_non_responders_microbes_num, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))



vedo_responders_microbes_denum <- 
Utrecht_Microbiome_with_counts_vedo[,c(stacked_df_bugs_vedo_6month_denum$predict, 
'Final..p1.')][Utrecht_Microbiome_with_counts_vedo$Responder_study == 
'Yes',]
vedo_responders_microbes_denum <- vedo_responders_microbes_denum %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
vedo_responders_microbes_denum$Final..p1. <- NULL

vedo_responders_microbes_denum[vedo_responders_microbes_denum == 0] <- NA
vedo_responders_microbes_denum <- 
vedo_responders_microbes_denum[,colSums(is.na(vedo_responders_microbes_denum))<nrow(vedo_responders_microbes_denum)]
vedo_responders_microbes_denum[sapply(vedo_responders_microbes_denum, 
is.numeric)] <- 
lapply(vedo_responders_microbes_denum[sapply(vedo_responders_microbes_denum, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))



vedo_non_responders_microbes_denum <- 
Utrecht_Microbiome_with_counts_vedo[,c(stacked_df_bugs_vedo_6month_denum$predict, 
'Final..p1.')][Utrecht_Microbiome_with_counts_vedo$Responder_study == 
'No',]
vedo_non_responders_microbes_denum <- vedo_non_responders_microbes_denum 
%>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
vedo_non_responders_microbes_denum$Final..p1. <- NULL
vedo_non_responders_microbes_denum[vedo_non_responders_microbes_denum == 
0] <- NA
vedo_non_responders_microbes_denum <- 
vedo_non_responders_microbes_denum[,colSums(is.na(vedo_non_responders_microbes_denum))<nrow(vedo_non_responders_microbes_denum)]
vedo_non_responders_microbes_denum[sapply(vedo_non_responders_microbes_denum, 
is.numeric)] <- 
lapply(vedo_non_responders_microbes_denum[sapply(vedo_non_responders_microbes_denum, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))





responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)
pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/utrecht_vedo.pdf")
# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 6 month response in Vedo in 
Utrecht Vedo patients",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)

dev.off()




```

# utrecht 6 months combined plot
``` {r }
# add counts for utrecht data 

utrecht_df <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_cohort_full_df.txt')



disease_duration_df <- 
read_delim('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_disease_activity.txt', 
delim = '\t') 

###########

Utrecht_Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/Utrecht_read_counts.txt', 
sep ='\t', header=TRUE)

merge_samples <- c()
ordered_samples <- c()
for (i in utrecht_df$...1){
  merge_samples <- c(merge_samples, i)
  ordered_samples <- c(ordered_samples, 
Utrecht_Microbiome_counts$Sample[grep( i, 
Utrecht_Microbiome_counts$Sample)])
}

rownames(Utrecht_Microbiome_counts) <- Utrecht_Microbiome_counts$Sample
  
Microbiome_counts_filtered <- Utrecht_Microbiome_counts[ordered_samples,]


Microbiome_counts_filtered$merging_column <- merge_samples

#vedo_microbiome_only$Fecal_sample_ID_1 <- 
merged_taxa_species_clinical_vedo$Fecal_sample_ID_1


Utrecht_Microbiome_with_counts <- merge(utrecht_df, 
Microbiome_counts_filtered, by.x = "...1", by.y = "merging_column")

Utrecht_Microbiome_with_counts <- merge(Utrecht_Microbiome_with_counts, 
disease_duration_df, by.x ="...1", by.y = "Fecal_sample_ID")

Utrecht_Microbiome_with_counts$merging_column <- NULL
Utrecht_Microbiome_with_counts$Fecal_sample_ID_1 <- NULL
Utrecht_Microbiome_with_counts$Sample <- NULL
Utrecht_Microbiome_with_counts$Raw.reads..p1. <- NULL
Utrecht_Microbiome_with_counts$Raw.reads..p2. <- NULL
Utrecht_Microbiome_with_counts$Final..p2. <- NULL
Utrecht_Microbiome_with_counts$Fecal_sample_ID_1 <- NULL
Utrecht_Microbiome_with_counts$sample_id_right <- NULL
Utrecht_Microbiome_with_counts_vedo <- 
Utrecht_Microbiome_with_counts[Utrecht_Microbiome_with_counts$V1_CurrentIBDMedication 
!= 'Vedolizumab',]
#vedo_microbiome_sub_counts <- Utrecht_Microbiome_with_counts %>%
#  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
#
#vedo_microbiome_sub_counts$Final..p1. <- NULL

#vedo_microbiome_sub_counts[vedo_microbiome_sub_counts == 0] <- NA

#vedo_microbiome_sub_counts[sapply(vedo_microbiome_sub_counts, 
is.numeric)] <- 
lapply(vedo_microbiome_sub_counts[sapply(vedo_microbiome_sub_counts, 
is.numeric)], #function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))



##########










#utrecht_df <- utrecht_df[utrecht_df$V1_CurrentIBDMedication == 
'Vedolizumab',]
clinical_response <- 'Clinical_responder (HBI ≤4 points\nor a reduction 
in the HBI by ≥3 points)'
month_6_response <-  'Responder_study'

stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]

stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]

stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]


vedo_responders_microbes_num <- 
Utrecht_Microbiome_with_counts_vedo[,c(stacked_df_bugs_vedo_6month_num$predict, 
'Final..p1.')][Utrecht_Microbiome_with_counts_vedo$Responder_study == 
'Yes',] 
vedo_responders_microbes_num <- vedo_responders_microbes_num %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
vedo_responders_microbes_num$Final..p1. <- NULL
#vedo_responders_microbes_num<- vedo_responders_microbes_num * 
Utrecht_Microbiome_with_counts$Final..p1.
vedo_responders_microbes_num[vedo_responders_microbes_num == 0] <- NA
vedo_responders_microbes_num <- 
vedo_responders_microbes_num[,colSums(is.na(vedo_responders_microbes_num))<nrow(vedo_responders_microbes_num)]
vedo_responders_microbes_num[sapply(vedo_responders_microbes_num, 
is.numeric)] <- 
lapply(vedo_responders_microbes_num[sapply(vedo_responders_microbes_num, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))


vedo_non_responders_microbes_num <- 
Utrecht_Microbiome_with_counts_vedo[,c(stacked_df_bugs_vedo_6month_num$predict, 
'Final..p1.')][Utrecht_Microbiome_with_counts_vedo$Responder_study == 
'No',] 
vedo_non_responders_microbes_num <- vedo_non_responders_microbes_num %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
vedo_non_responders_microbes_num$Final..p1. <- NULL
vedo_non_responders_microbes_num[vedo_non_responders_microbes_num == 0] <- 
NA
vedo_non_responders_microbes_num <- 
vedo_non_responders_microbes_num[,colSums(is.na(vedo_non_responders_microbes_num))<nrow(vedo_non_responders_microbes_num)]
vedo_non_responders_microbes_num[sapply(vedo_non_responders_microbes_num, 
is.numeric)] <- 
lapply(vedo_non_responders_microbes_num[sapply(vedo_non_responders_microbes_num, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))



vedo_responders_microbes_denum <- 
Utrecht_Microbiome_with_counts_vedo[,c(stacked_df_bugs_vedo_6month_denum$predict, 
'Final..p1.')][Utrecht_Microbiome_with_counts_vedo$Responder_study == 
'Yes',]
vedo_responders_microbes_denum <- vedo_responders_microbes_denum %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
vedo_responders_microbes_denum$Final..p1. <- NULL

vedo_responders_microbes_denum[vedo_responders_microbes_denum == 0] <- NA
vedo_responders_microbes_denum <- 
vedo_responders_microbes_denum[,colSums(is.na(vedo_responders_microbes_denum))<nrow(vedo_responders_microbes_denum)]
vedo_responders_microbes_denum[sapply(vedo_responders_microbes_denum, 
is.numeric)] <- 
lapply(vedo_responders_microbes_denum[sapply(vedo_responders_microbes_denum, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))



vedo_non_responders_microbes_denum <- 
Utrecht_Microbiome_with_counts_vedo[,c(stacked_df_bugs_vedo_6month_denum$predict, 
'Final..p1.')][Utrecht_Microbiome_with_counts_vedo$Responder_study == 
'No',]
vedo_non_responders_microbes_denum <- vedo_non_responders_microbes_denum 
%>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
vedo_non_responders_microbes_denum$Final..p1. <- NULL
vedo_non_responders_microbes_denum[vedo_non_responders_microbes_denum == 
0] <- NA
vedo_non_responders_microbes_denum <- 
vedo_non_responders_microbes_denum[,colSums(is.na(vedo_non_responders_microbes_denum))<nrow(vedo_non_responders_microbes_denum)]
vedo_non_responders_microbes_denum[sapply(vedo_non_responders_microbes_denum, 
is.numeric)] <- 
lapply(vedo_non_responders_microbes_denum[sapply(vedo_non_responders_microbes_denum, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))





responder_ratio <- rowSums(vedo_responders_microbes_num) / 
rowSums(vedo_responders_microbes_denum)

non_responder_ratio <- rowSums(vedo_non_responders_microbes_num) / 
rowSums(vedo_non_responders_microbes_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)
pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/utrecht_antitnf.pdf")
# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Microbiome ratio overview 6 month response in Combined 
cohort in Utrecht anti-TNF patients",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)

dev.off()



```

# Utrecht model validation
```{r } 
utrecht_df <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_cohort_full_df.txt')
utrecht_df <- utrecht_df[utrecht_df$V1_CurrentIBDMedication == 
'Vedolizumab',]



Utrecht_Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/Utrecht_read_counts.txt', 
sep ='\t', header=TRUE)

merge_samples <- c()
ordered_samples <- c()
for (i in utrecht_df$...1){
  merge_samples <- c(merge_samples, i)
  ordered_samples <- c(ordered_samples, 
Utrecht_Microbiome_counts$Sample[grep( i, 
Utrecht_Microbiome_counts$Sample)])
}

rownames(Utrecht_Microbiome_counts) <- Utrecht_Microbiome_counts$Sample
  
Microbiome_counts_filtered <- Utrecht_Microbiome_counts[ordered_samples,]


Microbiome_counts_filtered$merging_column <- merge_samples

#vedo_microbiome_only$Fecal_sample_ID_1 <- 
merged_taxa_species_clinical_vedo$Fecal_sample_ID_1


Utrecht_Microbiome_with_counts <- merge(utrecht_df, 
Microbiome_counts_filtered, by.x = "...1", by.y = "merging_column")

Utrecht_Microbiome_with_counts <- merge(Utrecht_Microbiome_with_counts, 
disease_duration_df, by.x ="...1", by.y = "Fecal_sample_ID")

Utrecht_Microbiome_with_counts$merging_column <- NULL
Utrecht_Microbiome_with_counts$Fecal_sample_ID_1 <- NULL
Utrecht_Microbiome_with_counts$Sample <- NULL
Utrecht_Microbiome_with_counts$Raw.reads..p1. <- NULL
Utrecht_Microbiome_with_counts$Raw.reads..p2. <- NULL
Utrecht_Microbiome_with_counts$Final..p2. <- NULL
Utrecht_Microbiome_with_counts$Fecal_sample_ID_1 <- NULL
Utrecht_Microbiome_with_counts$sample_id_right <- NULL
Utrecht_Microbiome_with_counts_vedo <- 
Utrecht_Microbiome_with_counts[Utrecht_Microbiome_with_counts$V1_CurrentIBDMedication 
== 'Vedolizumab',]


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

my_phenos_no_na$ratio <- NULL
trainIndex <- sample(1:nrow(my_phenos_no_na), 1 * nrow(my_phenos_no_na))
        
        
train_correct_IBD=my_phenos_no_na[trainIndex,]
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}

        #Regress out age, sex, bmi, calprotectin to IBD
partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
#partial_IBD_AUC <- pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], 
predict(partial_IBD, my_phenos_no_na[-trainIndex,]), quiet=T))

        
        # calculate ratios for vedo samples
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]
        
stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
vedo_6_month_microbe_ratio <- 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
my_phenos_no_na <- drop_na(my_phenos)
        
        
train_correct_IBD=my_phenos_no_na[trainIndex,]
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')

# here insert utrecht data

Utrecht_Microbiome_with_counts$V1_CurrentIBDMedication 
utrecht_nums <- 
Utrecht_Microbiome_with_counts[,c(stacked_df_bugs_vedo_6month_num$predict,  
'Final..p1.')]
utrecht_nums <- utrecht_nums %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
utrecht_nums$Final..p1. <- NULL
utrecht_nums[utrecht_nums == 0] <- NA
utrecht_nums <- 
utrecht_nums[,colSums(is.na(utrecht_nums))<nrow(utrecht_nums)]
utrecht_nums[sapply(utrecht_nums, is.numeric)] <- 
lapply(utrecht_nums[sapply(utrecht_nums, is.numeric)], function(a) 
ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))




utrecht_denums <- 
Utrecht_Microbiome_with_counts[,c(stacked_df_bugs_vedo_6month_denum$predict,  
'Final..p1.')]
utrecht_denums <- utrecht_denums %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
utrecht_denums$Final..p1. <- NULL
utrecht_denums[utrecht_denums == 0] <- NA
utrecht_denums <- 
utrecht_denums[,colSums(is.na(utrecht_denums))<nrow(utrecht_denums)]
utrecht_denums[sapply(utrecht_denums, is.numeric)] <- 
lapply(utrecht_denums[sapply(utrecht_denums, is.numeric)], function(a) 
ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))

utrecht_df$ratio <- rowSums(utrecht_nums) / rowSums(utrecht_denums)


#Utrecht_Microbiome_with_counts <- merge(utrecht_df, 
Microbiome_counts_filtered, by.x = "...1", by.y = "merging_column")

utrecht_df <- merge(utrecht_df, disease_duration_df, by.x ="...1", by.y = 
"Fecal_sample_ID")

utrecht_df
 
utrecht_df$V1_FecalCalprotectine <- 
utrecht_df$V1_FecalCalprotectine...5819
clinical_pred = utrecht_df[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP", "V1_PreviousantiTNF", 
"V2_disease_activity")]

ratio_pred = utrecht_df[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP", "V1_PreviousantiTNF", 
"V2_disease_activity", "ratio")]



utrecht_df$Responder_study[utrecht_df$Responder_study=="Yes"]=1
utrecht_df$Responder_study[utrecht_df$Responder_study=="No"]=0
utrecht_df$Responder_study=as.numeric(as.character(utrecht_df$Responder_study))
y_pred = utrecht_df$Responder_study


clinical_pred$V2_disease_activity[clinical_pred$V2_disease_activity=='No 
activity']=0
clinical_pred$V2_disease_activity[clinical_pred$V2_disease_activity=='Mild']=1
clinical_pred$V2_disease_activity[clinical_pred$V2_disease_activity=='Moderate']=2
clinical_pred$V2_disease_activity[clinical_pred$V2_disease_activity=='Severe']=3
clinical_pred$V2_disease_activity=as.numeric(as.character(clinical_pred$V2_disease_activity))

clinical_pred$V1_Sex[clinical_pred$V1_Sex=='female']='Female'
clinical_pred$V1_Sex[clinical_pred$V1_Sex=='Female ']='Female'
clinical_pred$V1_Sex[clinical_pred$V1_Sex=='male']='Male'
clinical_pred$V1_Sex[clinical_pred$V1_Sex=='F']='Female'
clinical_pred$V1_Sex[clinical_pred$V1_Sex=='M']='Male'
clinical_pred$V1_PreviousantiTNF[clinical_pred$V1_PreviousantiTNF=='Yes']=1
clinical_pred$V1_PreviousantiTNF[clinical_pred$V1_PreviousantiTNF=='No']=0
clinical_pred$V1_PreviousantiTNF=as.numeric(as.character(clinical_pred$V1_PreviousantiTNF))
clinical_pred$V1_SerumCRP[clinical_pred$V1_SerumCRP=='<0.3'] = 0
clinical_pred$V1_SerumCRP=as.numeric(as.character(clinical_pred$V1_SerumCRP))

clinical_pred$V1_SerumCRP[clinical_pred$V1_SerumCRP=='<0.3'] = 0
#my_phenos_no_na <- drop_na(clinical_pred)


ratio_pred$V2_disease_activity[ratio_pred$V2_disease_activity=='No 
activity']=0
ratio_pred$V2_disease_activity[ratio_pred$V2_disease_activity=='Mild']=1
ratio_pred$V2_disease_activity[ratio_pred$V2_disease_activity=='Moderate']=2
ratio_pred$V2_disease_activity[ratio_pred$V2_disease_activity=='Severe']=3
ratio_pred$V2_disease_activity=as.numeric(as.character(ratio_pred$V2_disease_activity))

ratio_pred$V1_Sex[ratio_pred$V1_Sex=='female']='Female'
ratio_pred$V1_Sex[ratio_pred$V1_Sex=='Female ']='Female'
ratio_pred$V1_Sex[ratio_pred$V1_Sex=='male']='Male'
ratio_pred$V1_Sex[ratio_pred$V1_Sex=='F']='Female'
ratio_pred$V1_Sex[ratio_pred$V1_Sex=='M']='Male'
ratio_pred$V1_PreviousantiTNF[ratio_pred$V1_PreviousantiTNF=='Yes']=1
ratio_pred$V1_PreviousantiTNF[ratio_pred$V1_PreviousantiTNF=='No']=0
ratio_pred$V1_PreviousantiTNF=as.numeric(as.character(ratio_pred$V1_PreviousantiTNF))

ratio_pred$V1_SerumCRP[ratio_pred$V1_SerumCRP=='<0.3'] = 0
ratio_pred$V1_SerumCRP=as.numeric(as.character(ratio_pred$V1_SerumCRP))

ratio_pred$V1_SerumCRP[ratio_pred$V1_SerumCRP=='<0.3'] = 0
#my_phenos_no_na <- drop_na(ratio_pred)


# predict utrecht based on clinical phenotypes
partial_IBD_AUC <- pROC::auc(pROC::roc(y_pred, predict(partial_IBD, 
clinical_pred), quiet=T))


# predict utrecht based on ratio
ratio_model_AUC <- pROC::auc(pROC::roc(y_pred, predict(ratio_model, 
ratio_pred), quiet=T))



#pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/AUC_utrecht_vedo.pdf")
obj.rpart1=pROC::roc(y_pred, predict(partial_IBD, clinical_pred), quiet=T, 
smooth = T)
obj.rpart2=pROC::roc(y_pred, predict(ratio_model, ratio_pred), quiet=T, 
smooth = T)
pROC::plot.roc(obj.rpart1, xlim=c(1,0), ylim=c(0,1), print.auc=TRUE, 
print.auc.x = 0.6, print.auc.y = 0.18, print.auc.pattern='Clinical model 
AUC=0.8')
pROC::plot.roc(obj.rpart2, xlim=c(1,0), ylim=c(0,1), print.auc=TRUE, 
add=TRUE, print.auc.x = 0.6, print.auc.y = 0.1, print.auc.pattern='Ratio 
model AUC=0.77')


#dev.off()
lrtest(partial_IBD, ratio_model)


```
# Utrecht model validation combined model in anti tnf
```{r } 
utrecht_df <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/utrecht_cohort_full_df.txt')
utrecht_df <- utrecht_df[utrecht_df$V1_CurrentIBDMedication != 
'Vedolizumab',]

Utrecht_Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/Utrecht_read_counts.txt', 
sep ='\t', header=TRUE)

merge_samples <- c()
ordered_samples <- c()
for (i in utrecht_df$...1){
  merge_samples <- c(merge_samples, i)
  ordered_samples <- c(ordered_samples, 
Utrecht_Microbiome_counts$Sample[grep( i, 
Utrecht_Microbiome_counts$Sample)])
}

rownames(Utrecht_Microbiome_counts) <- Utrecht_Microbiome_counts$Sample
  
Microbiome_counts_filtered <- Utrecht_Microbiome_counts[ordered_samples,]


Microbiome_counts_filtered$merging_column <- merge_samples

#vedo_microbiome_only$Fecal_sample_ID_1 <- 
merged_taxa_species_clinical_vedo$Fecal_sample_ID_1


Utrecht_Microbiome_with_counts <- merge(utrecht_df, 
Microbiome_counts_filtered, by.x = "...1", by.y = "merging_column")

Utrecht_Microbiome_with_counts <- merge(Utrecht_Microbiome_with_counts, 
disease_duration_df, by.x ="...1", by.y = "Fecal_sample_ID")

Utrecht_Microbiome_with_counts$merging_column <- NULL
Utrecht_Microbiome_with_counts$Fecal_sample_ID_1 <- NULL
Utrecht_Microbiome_with_counts$Sample <- NULL
Utrecht_Microbiome_with_counts$Raw.reads..p1. <- NULL
Utrecht_Microbiome_with_counts$Raw.reads..p2. <- NULL
Utrecht_Microbiome_with_counts$Final..p2. <- NULL
Utrecht_Microbiome_with_counts$Fecal_sample_ID_1 <- NULL
Utrecht_Microbiome_with_counts$sample_id_right <- NULL
Utrecht_Microbiome_with_counts_vedo <- 
Utrecht_Microbiome_with_counts[Utrecht_Microbiome_with_counts$V1_CurrentIBDMedication 
!= 'Vedolizumab',]






my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

my_phenos_no_na$ratio <- NULL
trainIndex <- sample(1:nrow(my_phenos_no_na), 1 * nrow(my_phenos_no_na))
        
        
train_correct_IBD=my_phenos_no_na[trainIndex,]
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}

        #Regress out age, sex, bmi, calprotectin to IBD
partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
#partial_IBD_AUC <- pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], 
predict(partial_IBD, my_phenos_no_na[-trainIndex,]), quiet=T))

        
        # calculate ratios for vedo samples
stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]
        
stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
my_phenos_no_na <- drop_na(my_phenos)
        
        
train_correct_IBD=my_phenos_no_na[trainIndex,]
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')

# here insert utrecht data

utrecht_nums <- 
Utrecht_Microbiome_with_counts[,c(stacked_df_bugs_vedo_6month_num$predict,  
'Final..p1.')]
utrecht_nums <- utrecht_nums %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
utrecht_nums$Final..p1. <- NULL
utrecht_nums[utrecht_nums == 0] <- NA
utrecht_nums <- 
utrecht_nums[,colSums(is.na(utrecht_nums))<nrow(utrecht_nums)]
utrecht_nums[sapply(utrecht_nums, is.numeric)] <- 
lapply(utrecht_nums[sapply(utrecht_nums, is.numeric)], function(a) 
ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))




utrecht_denums <- 
Utrecht_Microbiome_with_counts[,c(stacked_df_bugs_vedo_6month_denum$predict,  
'Final..p1.')]
utrecht_denums <- utrecht_denums %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))
utrecht_denums$Final..p1. <- NULL
utrecht_denums[utrecht_denums == 0] <- NA
utrecht_denums <- 
utrecht_denums[,colSums(is.na(utrecht_denums))<nrow(utrecht_denums)]
utrecht_denums[sapply(utrecht_denums, is.numeric)] <- 
lapply(utrecht_denums[sapply(utrecht_denums, is.numeric)], function(a) 
ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))

utrecht_df$ratio <- rowSums(utrecht_nums) / rowSums(utrecht_denums)


#Utrecht_Microbiome_with_counts <- merge(utrecht_df, 
Microbiome_counts_filtered, by.x = "...1", by.y = "merging_column")

utrecht_df <- merge(utrecht_df, disease_duration_df, by.x ="...1", by.y = 
"Fecal_sample_ID")

utrecht_df
 
utrecht_df$V1_FecalCalprotectine <- 
utrecht_df$V1_FecalCalprotectine...5819
clinical_pred = utrecht_df[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP", "V1_PreviousantiTNF", 
"V2_disease_activity")]

ratio_pred = utrecht_df[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP", "V1_PreviousantiTNF", 
"V2_disease_activity", "ratio")]



utrecht_df$Responder_study[utrecht_df$Responder_study=="Yes"]=1
utrecht_df$Responder_study[utrecht_df$Responder_study=="No"]=0
utrecht_df$Responder_study=as.numeric(as.character(utrecht_df$Responder_study))
y_pred = utrecht_df$Responder_study


clinical_pred$V2_disease_activity[clinical_pred$V2_disease_activity=='No 
activity']=0
clinical_pred$V2_disease_activity[clinical_pred$V2_disease_activity=='Mild']=1
clinical_pred$V2_disease_activity[clinical_pred$V2_disease_activity=='Moderate']=2
clinical_pred$V2_disease_activity[clinical_pred$V2_disease_activity=='Severe']=3
clinical_pred$V2_disease_activity=as.numeric(as.character(clinical_pred$V2_disease_activity))

clinical_pred$V1_Sex[clinical_pred$V1_Sex=='female']='Female'
clinical_pred$V1_Sex[clinical_pred$V1_Sex=='Female ']='Female'
clinical_pred$V1_Sex[clinical_pred$V1_Sex=='male']='Male'
clinical_pred$V1_Sex[clinical_pred$V1_Sex=='F']='Female'
clinical_pred$V1_Sex[clinical_pred$V1_Sex=='M']='Male'
clinical_pred$V1_PreviousantiTNF[clinical_pred$V1_PreviousantiTNF=='Yes']=1
clinical_pred$V1_PreviousantiTNF[clinical_pred$V1_PreviousantiTNF=='No']=0
clinical_pred$V1_PreviousantiTNF=as.numeric(as.character(clinical_pred$V1_PreviousantiTNF))
clinical_pred$V1_SerumCRP[clinical_pred$V1_SerumCRP=='<0.3'] = 0
clinical_pred$V1_SerumCRP=as.numeric(as.character(clinical_pred$V1_SerumCRP))

clinical_pred$V1_SerumCRP[clinical_pred$V1_SerumCRP=='<0.3'] = 0
#my_phenos_no_na <- drop_na(clinical_pred)


ratio_pred$V2_disease_activity[ratio_pred$V2_disease_activity=='No 
activity']=0
ratio_pred$V2_disease_activity[ratio_pred$V2_disease_activity=='Mild']=1
ratio_pred$V2_disease_activity[ratio_pred$V2_disease_activity=='Moderate']=2
ratio_pred$V2_disease_activity[ratio_pred$V2_disease_activity=='Severe']=3
ratio_pred$V2_disease_activity=as.numeric(as.character(ratio_pred$V2_disease_activity))

ratio_pred$V1_Sex[ratio_pred$V1_Sex=='female']='Female'
ratio_pred$V1_Sex[ratio_pred$V1_Sex=='Female ']='Female'
ratio_pred$V1_Sex[ratio_pred$V1_Sex=='male']='Male'
ratio_pred$V1_Sex[ratio_pred$V1_Sex=='F']='Female'
ratio_pred$V1_Sex[ratio_pred$V1_Sex=='M']='Male'
ratio_pred$V1_PreviousantiTNF[ratio_pred$V1_PreviousantiTNF=='Yes']=1
ratio_pred$V1_PreviousantiTNF[ratio_pred$V1_PreviousantiTNF=='No']=0
ratio_pred$V1_PreviousantiTNF=as.numeric(as.character(ratio_pred$V1_PreviousantiTNF))

ratio_pred$V1_SerumCRP[ratio_pred$V1_SerumCRP=='<0.3'] = 0
ratio_pred$V1_SerumCRP=as.numeric(as.character(ratio_pred$V1_SerumCRP))

ratio_pred$V1_SerumCRP[ratio_pred$V1_SerumCRP=='<0.3'] = 0
#my_phenos_no_na <- drop_na(ratio_pred)


# predict utrecht based on clinical phenotypes
partial_IBD_AUC <- pROC::auc(pROC::roc(y_pred, predict(partial_IBD, 
clinical_pred), quiet=T))


# predict utrecht based on ratio
ratio_model_AUC <- pROC::auc(pROC::roc(y_pred, predict(ratio_model, 
ratio_pred), quiet=T))


improvement_aucs <- c(improvement_aucs, ratio_model_AUC - partial_IBD_AUC)
partial_model <- c(partial_model, partial_IBD_AUC)
full_model <- c(full_model, ratio_model_AUC)


#pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/AUC_utrecht_antitnf.pdf")

obj.rpart1=pROC::roc(y_pred, predict(partial_IBD, clinical_pred), quiet=T, 
smooth = T)
obj.rpart2=pROC::roc(y_pred, predict(ratio_model, ratio_pred), quiet=T, 
smooth = T)
pROC::plot.roc(obj.rpart1, xlim=c(1,0), ylim=c(0,1), print.auc=TRUE, 
print.auc.x = 0.7, print.auc.y = 0.18, print.auc.pattern='Clinical model 
AUC=0.655')
pROC::plot.roc(obj.rpart2, xlim=c(1,0), ylim=c(0,1), print.auc=TRUE, 
add=TRUE, print.auc.x = 0.7, print.auc.y = 0.1, print.auc.pattern='Ratio 
model AUC=0.639')

#dev.off()



lrtest(partial_IBD, ratio_model)

```





### compare model qualities now:

## first Vedo
```{r  Vedo 6 months}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```




```{r  Vedo 2 year}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))


```


```{r  Vedo HBI SCCAI}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]

# add smoking and disease duration instead of age at fs 

my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


```{r  Vedo 6 months metabolites}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#
 

my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears", 
"V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

        #"V1_Currentsmoker",
        
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```




```{r  Vedo 2 year metabolites}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))


```


```{r  Vedo HBI SCCAI metabolites}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))


```


### add both ratios together

```{r  Vedo 6 months}


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
  
        my_phenos_no_na$ratio_microbe <- NULL
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbe <- vedo_6_month_microbe_ratio
        
        
        
        
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```




```{r  Vedo 2 year}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio_microbe <- NULL
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbe <- vedo_6_month_microbe_ratio
        
        
        
        
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))


```


```{r  Vedo HBI SCCAI}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=vedo_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio_microbe <- NULL
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbe <- vedo_6_month_microbe_ratio
        
        
        
        
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Vedolizumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(vedo_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        #trainIndex <- sample(1:nrow(my_phenos), 0.75 * nrow(my_phenos))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```









## ustek 
```{r  Ustek 6 months}


# ustek_all_new_ID_raw_filt.   

my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```




```{r  Ustek 2 year}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


```{r  Ustek HBI SCCAI}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```





```{r  Ustek 6 months metabolites}


# ustek_all_new_ID_raw_filt.   

my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```




```{r  Ustek 2 year metabolites}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


```{r  Ustek HBI SCCAI metabolites}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


```{r  Ustek 6 months}


# ustek_all_new_ID_raw_filt.   

my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio_microbiome <- NULL
        my_phenos_no_na$ratio_metabolites <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbiome <- vedo_6_month_microbe_ratio
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio
        
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```




```{r  Ustek 2 year}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio_microbiome <- NULL
        my_phenos_no_na$ratio_metabolites <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbiome <- vedo_6_month_microbe_ratio
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


```{r  Ustek HBI SCCAI}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=Ustek_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio_microbiome <- NULL
        my_phenos_no_na$ratio_metabolites <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Ustek_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbiome <- vedo_6_month_microbe_ratio
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 
'Ustekinumab',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(ustek_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```












## combined cohort:
```{r  Combined 6 months}
# ustek_all_new_ID_raw_filt.   

my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```




```{r  Combined 2 year}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


```{r  Combined HBI SCCAI}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```





```{r  Combined 6 months metabolites}


# ustek_all_new_ID_raw_filt.   

my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```




```{r  Combined 2 year metabolites}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '2 year',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


```{r  Combined HBI SCCAI metabolites}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```






```{r  Combined 6 months}
# ustek_all_new_ID_raw_filt.   

my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()
stored_rocs <- c()
for (i in 1:1000){
        my_phenos_no_na$ratio_microbiome <- NULL
        my_phenos_no_na$ratio_metabolites <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbiome <- vedo_6_month_microbe_ratio
        
        
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio
        
        
        
        stacked_df_pathways_filtered <- 
stacked_df_pathways[stacked_df_pathways$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_pathways_filtered[stacked_df_pathways_filtered$Medication == 
'Combined cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(Combined_pathways_codacore_input[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Combined_pathways_codacore_input[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_pathways <- vedo_6_month_microbe_ratio
        
        
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        #stored_rocs <- c(stored_rocs, 
pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```



```{r}
for (i in 1:1000){
  stored_rocs
}



```


```{r  Combined 2 year}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_2years", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_2years[my_phenos$Response_2years=="yes"]=1
my_phenos$Response_2years[my_phenos$Response_2years=="no"]=0
my_phenos$Response_2years=as.numeric(as.character(my_phenos$Response_2years))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio_microbiome <- NULL
        my_phenos_no_na$ratio_metabolites <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbiome <- vedo_6_month_microbe_ratio
        
        
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_2years
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_2years ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


```{r  Combined HBI SCCAI}

#vedo_clinical_data
#vedo_metabolites_only 
#vedo_microbiome_only
#


my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","Response_HBI_SCCAI", 
"V1_PreviousantiTNF", "V2_disease_activity")]


my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="yes"]=1
my_phenos$Response_HBI_SCCAI[my_phenos$Response_HBI_SCCAI=="no"]=0
my_phenos$Response_HBI_SCCAI=as.numeric(as.character(my_phenos$Response_HBI_SCCAI))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio_microbiome <- NULL
        my_phenos_no_na$ratio_metabolites <- NULL
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        # calculate ratios for vedo samples
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbiome <- vedo_6_month_microbe_ratio
        
        
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio
        
        
        
        
        
        
        stacked_df_pathways_filtered <- 
stacked_df_pathways[stacked_df_pathways$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_pathways_filtered[stacked_df_pathways_filtered$Medication == 
'Combined cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == 'HBI SCCAI',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$Response_HBI_SCCAI
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(Response_HBI_SCCAI ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

```


#### rankos human filtering
```{r }
#Ranko's functions
prepCleanHumann <- function(inPath, dropUnintegrated = T, dropUnmapped = 
T, dropTaxonSpecific = T,
                            presenceFilter = -1,minRelativeAbundance = -1, 
rescaleTaxa = T, novogeneIdsClean = T) {
  inDF <- read.table(inPath,sep='\t',header=T,quote ='',comment.char = '')
  #fix pathway ID (these tend to be weird coming out of humann)
  rownames(inDF) <- inDF$X..Pathway
  inDF$X..Pathway <- NULL
  #drop "junk" from humann (unintegrated/unmapped data & taxon-specific 
pathways)
  if (dropTaxonSpecific) {
    inDF <- inDF[grep('\\|',rownames(inDF),invert = T),]
  }
  if (dropUnintegrated) {
    inDF <- inDF[grep('UNINTEGRATED',rownames(inDF),invert = T),]
  }
  if (dropUnmapped) {
    inDF <- inDF[grep('UNMAPPED',rownames(inDF),invert = T),]
  }
  rownames(inDF)[grep('PWY',rownames(inDF),invert = T)] <- 
paste0('PWY_',rownames(inDF)[grep('PWY',rownames(inDF),invert = T)])
  inDF <- as.data.frame(t.data.frame(inDF))
  # fix sample IDs, remove duplicates
  inDF$ID <- rownames(inDF)
  if (novogeneIdsClean) {
    print ('NOTE: doing cleaning of Novogene IDs (keeping format 
<AB>_<CD>_<EFG...>)')
    for (i in c(1:nrow(inDF))) {
      ss <- strsplit(inDF$ID[i],'_')[[1]]
      sss <- paste0(ss[1],'_',ss[2],'_',ss[3])
      inDF$ID[i] <- sss
    }
  }
  if (sum(duplicated(inDF$ID) > 0)) {
    print(paste('WARNING: found ',sum(duplicated(inDF$ID) > 
0),'duplicates, dropping them!'))
  }
  inDF <- inDF[!duplicated(inDF$ID),]
  rownames(inDF) <- inDF$ID
  inDF$ID <- NULL
  # make sure columns are actually numbers (otherwise filter dies; NOTE: 
this should not be necessary, but... )
  for (c in colnames(inDF)) {inDF[[c]] <- as.numeric(inDF[[c]])}
  # clean, rescale and save
  inDFt2 <- filterHumannDF(inDF,presPerc = presenceFilter,minMRelAb = 
minRelativeAbundance,minMedRelAb = -1,rescale = T,minSum = 1,verbose = T)
  inDFt2 <- inDFt2[,colSums(inDFt2)!=0]
  inDFt2$ID <- row.names(inDFt2)
  print(paste('Done, returning ',nrow(inDFt2),'samples'))
  inDFt2
}


filterHumannDF <- function(inDF,presPerc = -1 ,minMRelAb = -1,minMedRelAb= 
-1, minSum = -1, rescale=T,verbose=T,type='MetaCyc') {
  
  nonPWYpwys <- c("ARG+POLYAMINE-SYN: superpathway of arginine and 
polyamine biosynthesis",
                  "CHLOROPHYLL-SYN: chlorophyllide a biosynthesis I 
(aerobic, light-dependent)",
                  "GLYCOLYSIS-E-D: superpathway of glycolysis and 
Entner-Doudoroff",
                  "GLYCOLYSIS-TCA-GLYOX-BYPASS: superpathway of 
glycolysis, pyruvate dehydrogenase, TCA, and glyoxylate bypass",
                  "GLYCOLYSIS: glycolysis I (from glucose 6-phosphate)",
                  "GLYOXYLATE-BYPASS: glyoxylate cycle",
                  "HEME-BIOSYNTHESIS-II: heme biosynthesis I (aerobic)",
                  "MANNOSYL-CHITO-DOLICHOL-BIOSYNTHESIS: protein 
N-glycosylation (eukaryotic, high mannose)",
                  "NAD-BIOSYNTHESIS-II: NAD salvage pathway II",                  
                  "REDCITCYC: TCA cycle VIII (helicobacter)",
                  "TCA-GLYOX-BYPASS: superpathway of glyoxylate bypass and 
TCA",
                  "TCA: TCA cycle I (prokaryotic)")
  
  colnames(inDF)[colnames(inDF) %in% nonPWYpwys] <- 
paste0('PWY_',colnames(inDF)[colnames(inDF) %in% nonPWYpwys])
  
  if (type=='MetaCyc') {
    nonPWYdf <- as.data.frame(inDF[,-grep('PWY',colnames(inDF))])
    cnsNonPWYdf <- 
colnames(inDF[colnames(inDF)[-grep('PWY',colnames(inDF))] ])
  } else if (type=='EC') {
    nonPWYdf <- as.data.frame(inDF[,-grep('^EC_',colnames(inDF))])
    cnsNonPWYdf <- 
colnames(inDF[colnames(inDF)[-grep('^EC_',colnames(inDF))] ])
  } else if (type=='RXN') {
    nonPWYdf <- as.data.frame(inDF[,-grep('RXN',colnames(inDF))])
    cnsNonPWYdf <- 
colnames(inDF[colnames(inDF)[-grep('RXN',colnames(inDF))] ])
  } else if (type=='PFAM') {
    nonPWYdf <- as.data.frame(inDF[,-grep('^PF[01]',colnames(inDF))])
    cnsNonPWYdf <- 
colnames(inDF[colnames(inDF)[-grep('^PF[01]',colnames(inDF))] ])
  } else if (type=='GO') {
    nonPWYdf <- as.data.frame(inDF[,-grep('^GO',colnames(inDF))])
    cnsNonPWYdf <- 
colnames(inDF[colnames(inDF)[-grep('^GO',colnames(inDF))] ])
  } else if (type=='KEGG') {
    nonPWYdf <- as.data.frame(inDF[,-grep('^K[012]',colnames(inDF))])
    cnsNonPWYdf <- 
colnames(inDF[colnames(inDF)[-grep('^K[012]',colnames(inDF))] ])
  }
  colnames(nonPWYdf) <- cnsNonPWYdf
  if (type=='MetaCyc') {
    yesPWYdf <- as.data.frame(inDF[,grep('PWY',colnames(inDF))])
    cnsYesPWYdf <- 
colnames(inDF[colnames(inDF)[grep('PWY',colnames(inDF))] ])
  } else if (type=='EC') {
    yesPWYdf <- as.data.frame(inDF[,grep('^EC_',colnames(inDF))])
    cnsYesPWYdf <- 
colnames(inDF[colnames(inDF)[grep('^EC_',colnames(inDF))] ])
  } else if (type=='RXN') {
    yesPWYdf <- as.data.frame(inDF[,grep('RXN',colnames(inDF))])
    cnsYesPWYdf <- 
colnames(inDF[colnames(inDF)[grep('RXN',colnames(inDF))] ])
  } else if (type=='PFAM') {
    yesPWYdf <- as.data.frame(inDF[,grep('^PF[01]',colnames(inDF))])
    cnsYesPWYdf <- 
colnames(inDF[colnames(inDF)[grep('^PF[01]',colnames(inDF))] ])
  } else if (type=='GO') {
    yesPWYdf <- as.data.frame(inDF[,grep('^GO',colnames(inDF))])
    cnsYesPWYdf <- 
colnames(inDF[colnames(inDF)[grep('^GO',colnames(inDF))] ])
  } else if (type=='KEGG') {
    yesPWYdf <- as.data.frame(inDF[,grep('^K[012]',colnames(inDF))])
    cnsYesPWYdf <- 
colnames(inDF[colnames(inDF)[grep('^K[012]',colnames(inDF))] ])
  }
  
  # replaces NAs with 0s
  for (c in colnames(yesPWYdf)) {
    yesPWYdf[,c][is.na(yesPWYdf[,c])] <- 0.0
  }
  # rescale to rel ab (if rescale = T)
  if (rescale==T) {
    if (verbose) {print ('  >> rescaling')}
    rsums <- rowSums(yesPWYdf)
    rsums[rsums==0] <- 1.0
    yesPWYdf <- yesPWYdf/rsums
  }
  
  # filter for presence
  # -----------------
  nrRemoved = 0
  toRemove = c()
  for (c in colnames(yesPWYdf)) {
    nrnZ = as.numeric(sum(yesPWYdf[,c]!=0.0))
    if (nrnZ/as.numeric(nrow(yesPWYdf)) < presPerc) {
      nrRemoved = nrRemoved + 1
      toRemove <- c(toRemove,c)
    }
  }
  if (length(toRemove) > 0) {
    yesPWYdf <- yesPWYdf[,!(colnames(yesPWYdf) %in% toRemove)]
  }
  if (verbose) {print (paste(' > presence filter: 
Removed',nrRemoved,'pathways!, ',length(colnames(yesPWYdf)),'pathways 
left!')); }
  
  # filter for abundance (mean)
  # ---------------------------
  nrRemoved = 0
  toRemove = c()
  for (c in colnames(yesPWYdf)) {
    mn = mean(yesPWYdf[,c])
    if ( mn < minMRelAb) {
      nrRemoved = nrRemoved + 1
      toRemove <- c(toRemove,c)
    }
  }
  if (length(toRemove) > 0) {
    yesPWYdf <- yesPWYdf[,!(colnames(yesPWYdf) %in% toRemove)]
  }
  if (verbose) {print (paste(' > mean abundance filter: 
Removed',nrRemoved,'pathways!, ',length(colnames(yesPWYdf)),'pathways 
left!')); }
  
  # filter for abundance (median)
  # -----------------------------
  nrRemoved = 0
  toRemove = c()
  for (c in colnames(yesPWYdf)) {
    mn = median(yesPWYdf[,c])
    if ( mn < minMedRelAb) {
      nrRemoved = nrRemoved + 1
      toRemove <- c(toRemove,c)
    }
  }
  if (length(toRemove) > 0) {
    yesPWYdf <- yesPWYdf[,!(colnames(yesPWYdf) %in% toRemove)]
  }
  if (verbose) {print (paste(' > median abundance filter: 
Removed',nrRemoved,'pathways!, ',length(colnames(yesPWYdf)),'pathways 
left!')); }
  
  # do final rescale
  if (rescale==T) {
    if (verbose) {print ('  >> rescaling')}
    rsums <- rowSums(yesPWYdf)
    rsums[rsums==0] <- 1.0
    yesPWYdf <- yesPWYdf/rsums
  }
  inDF <- cbind.data.frame(nonPWYdf,yesPWYdf)
  if (verbose) {print ('> DONE')}
  inDF
}







```








### pathways

```{r }
library(vegan)

# pathways dataprep

#df_pathways_everything <- 
read_csv('/Users/iwan/Research/drug_response/metadata/3_RF_mergedcohort_mergeddata_scores.csv')
#df_pathways_everything$...1 <- NULL
#df_pathways_clinical <-  df_pathways_everything[,1:15]
#df_pathways_microbes <-  df_pathways_everything[,16:70]
#df_pathways_metabolites <-  df_pathways_everything[,1:70] 
#df_pathways_pathways <-  df_pathways_everything[,1:70]


# pathways file  # read count file # merge with 


df_pathways <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/human_pathways_merged.txt')
Meta_vedo <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/RF_vedo_clinial_features.csv')
Meta_ustek <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/RF_ustek_clinial_features.csv')
Meta_merged <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/RF_mergedcohort_clinial_features.csv')
#rownames(df_pathways) <- df_pathways$Pathway

# rankos function


PWYs <- 
prepCleanHumann("/Users/iwan/Research/drug_response/metadata/lastest_metadata/human_pathways_merged.txt",
                        dropUnintegrated = T,dropUnmapped = T, 
dropTaxonSpecific = T,
                        presenceFilter = -1, minRelativeAbundance = -1, 
rescaleTaxa = T, novogeneIdsClean = T) #keeps 504 pathways

#### Selecting pathways for cohorts ----
#Select pathways for vedo
PWYs_vedo <- PWYs[grepl("^Vedo", rownames(PWYs)), ]
PWYs_vedo$ID <- NULL
PWYs_vedo <- PWYs_vedo[,colSums(PWYs_vedo)>0] #keep only columns that are 
not only zeros (from 504 to 469)
Vedo_IDs <- unique(na.omit(c(Meta_vedo$Fecal_sample_ID_1, 
Meta_vedo$Fecal_sample_ID_2))) #to only keep samples of included patients
PWYs_vedo <- PWYs_vedo[rownames(PWYs_vedo) %in% Vedo_IDs, ]
PWYs_vedo <- PWYs_vedo[rownames(PWYs_vedo) != "Vedop1_F12_94", ] # less 
than 100.000 reads
PWYs_vedo_baseline <- PWYs_vedo[rownames(PWYs_vedo) %in% 
Meta_vedo$Fecal_sample_ID_1, ] #samples baseline (42)

#Select pathways for ustek
PWYs_ustek <- PWYs[grepl("^Ust", rownames(PWYs)), ]
PWYs_ustek$ID <- NULL
PWYs_ustek <- PWYs_ustek[,colSums(PWYs_ustek)>0] #keep only columns that 
are not only zeros (from 504 to 452)
Ustek_IDs <- unique(na.omit(c(Meta_ustek$Fecal_sample_ID_1, 
Meta_ustek$Fecal_sample_ID_2))) #to only keep samples of included patients
PWYs_ustek <- PWYs_ustek[rownames(PWYs_ustek) %in% Ustek_IDs, ]
PWYs_ustek <- PWYs_ustek[rownames(PWYs_ustek) != "Ust45_1_199", ] #remove 
this sample <1 million reads
PWYs_ustek <- PWYs_ustek[rownames(PWYs_ustek) != "Ust31_1_133", ]  #remove 
this sample <1 million reads
PWYs_ustek_baseline <- PWYs_ustek[rownames(PWYs_ustek) %in% 
Meta_ustek$Fecal_sample_ID_1, ] #samples baseline

#Select pathways for merged cohort
PWYs_merged <- PWYs[grepl("^Vedo|^Ust", rownames(PWYs)), ]
PWYs_merged$ID <- NULL
PWYs_merged <- PWYs_merged[,colSums(PWYs_merged)>0] #keep only columns 
that are not only zeros (from 504 to 476)
Merged_IDs <- unique(na.omit(c(Meta_merged$Fecal_sample_ID_1, 
Meta_merged$Fecal_sample_ID_2))) #to only keep samples of included 
patients
PWYs_merged <- PWYs_merged[rownames(PWYs_merged) %in% Merged_IDs, ]
PWYs_merged <- PWYs_merged[rownames(PWYs_merged) != "Ust45_1_199", ] 
#remove this sample <1 million reads
PWYs_merged <- PWYs_merged[rownames(PWYs_merged) != "Ust31_1_133", ]  
#remove this sample <1 million reads
PWYs_merged <- PWYs_merged[rownames(PWYs_merged) != "Vedop1_F12_94", ] # 
less than 100.000 reads
PWYs_merged_baseline <- PWYs_merged[rownames(PWYs_merged) %in% 
Meta_merged$Fecal_sample_ID_1, ] #samples baseline

#Filtering pathways and CLR transformation
#vedo
PWYs_vedo_baseline_filt <- filterHumannDF(PWYs_vedo_baseline,presPerc = 
0.15,minMRelAb = 0.001,minMedRelAb = -1,rescale = T,minSum = 1,verbose = 
T) #197 left
PWYs_vedo_baseline_filt_clr <- decostand(PWYs_vedo_baseline_filt,method = 
'clr',pseudocount = min(PWYs_vedo_baseline_filt[PWYs_vedo_baseline_filt > 
0])/2)

#ustek
PWYs_ustek_baseline_filt <- filterHumannDF(PWYs_ustek_baseline,presPerc = 
0.15,minMRelAb = 0.001,minMedRelAb = -1,rescale = T,minSum = 1,verbose = 
T) #184 left
PWYs_ustek_baseline_filt_clr <- decostand(PWYs_ustek_baseline_filt,method 
= 'clr',pseudocount = 
min(PWYs_ustek_baseline_filt[PWYs_ustek_baseline_filt > 0])/2)

#merged
PWYs_merged_baseline_filt <- filterHumannDF(PWYs_merged_baseline,presPerc 
= 0.15,minMRelAb = 0.001,minMedRelAb = -1,rescale = T,minSum = 1,verbose = 
T) #193 left
PWYs_merged_baseline_filt_clr <- 
decostand(PWYs_merged_baseline_filt,method = 'clr',pseudocount = 
min(PWYs_merged_baseline_filt[PWYs_merged_baseline_filt > 0])/2)


# check for pseudocounts or counts or something
###### VEDO COUNTS
Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/Vedo_read_counts.txt', 
sep ='\t', header=TRUE)

#Microbiome_counts$Sample
merge_index_list <- c()
for (i in 1:length(Microbiome_counts$Sample)){
  merge_index_list <- c(merge_index_list, 
paste(str_split(Microbiome_counts$Sample, '_')[[i]][1:3],collapse='_'))

}

Microbiome_counts$merge_index <- merge_index_list

pathways_vedo_with_readcounts <- merge(PWYs_vedo_baseline_filt, 
Microbiome_counts, by.x = 0, by.y = 'merge_index')


pathways_vedo_with_readcounts$merge_index <- NULL

pathways_vedo_with_readcounts$Sample <- NULL
pathways_vedo_with_readcounts$Raw.reads..p1. <- NULL
pathways_vedo_with_readcounts$Raw.reads..p2. <- NULL
pathways_vedo_with_readcounts$Final..p2. <- NULL

pathways_vedo_with_readcounts$sample_id_right <- NULL
rownames(pathways_vedo_with_readcounts) <- 
pathways_vedo_with_readcounts$Row.names
pathways_vedo_with_readcounts$Row.names <- NULL

pathways_vedo_with_readcounts <- pathways_vedo_with_readcounts %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))

pathways_vedo_with_readcounts$Final..p1. <- NULL


pathways_vedo_with_readcounts[pathways_vedo_with_readcounts == 0] <- NA

pathways_vedo_with_readcounts[sapply(pathways_vedo_with_readcounts, 
is.numeric)] <- 
lapply(pathways_vedo_with_readcounts[sapply(pathways_vedo_with_readcounts, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))

Vedo_pathways_codacore_input <- pathways_vedo_with_readcounts

########### ustek counts
# check for pseudocounts or counts or something



# correct names
ustek_Microbiome_counts <- 
read.csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_updated_read_counts.csv', 
sep =';', header=TRUE)
ustek_Microbiome_counts <- ustek_Microbiome_counts %>%
  filter(!grepl("_2$", Sample))




PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)



pathways_vedo_with_readcounts <- merge(PWYs_ustek_baseline_filt, 
ustek_Microbiome_counts, by.x = 'merging', by.y = 'Sample')


pathways_vedo_with_readcounts$merge_index <- NULL

pathways_vedo_with_readcounts$Sample <- NULL
pathways_vedo_with_readcounts$Raw.reads..p1. <- NULL
pathways_vedo_with_readcounts$Raw.reads..p2. <- NULL
pathways_vedo_with_readcounts$Final..p2. <- NULL

pathways_vedo_with_readcounts$sample_id_right <- NULL
rownames(pathways_vedo_with_readcounts) <- 
pathways_vedo_with_readcounts$merging
pathways_vedo_with_readcounts$merging <- NULL

pathways_vedo_with_readcounts <- pathways_vedo_with_readcounts %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))

pathways_vedo_with_readcounts$Final..p1. <- NULL


pathways_vedo_with_readcounts[pathways_vedo_with_readcounts == 0] <- NA

pathways_vedo_with_readcounts[sapply(pathways_vedo_with_readcounts, 
is.numeric)] <- 
lapply(pathways_vedo_with_readcounts[sapply(pathways_vedo_with_readcounts, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))

Ustek_pathways_codacore_input <- pathways_vedo_with_readcounts











Microbiome_counts$Sample <- Microbiome_counts$merge_index
Microbiome_counts$Raw.reads..p1. <- NULL
Microbiome_counts$Raw.reads..p2. <- NULL
Microbiome_counts$Final..p2. <- NULL
Microbiome_counts$merge_index <- NULL


ustek_Microbiome_counts$Raw.reads..p1. <- NULL
ustek_Microbiome_counts$Raw.reads..p2. <- NULL
ustek_Microbiome_counts$Final..p2. <- NULL
#ustek_Microbiome_counts$merge_index <- NULL

#Microbiome_counts$merge_index <- NULL
#Microbiome_counts$merge_index <- NULL

all_read_counts <- rbind(ustek_Microbiome_counts, Microbiome_counts)





rownames(PWYs_merged_baseline_filt)   <-   
c(str_split_i(rownames(PWYs_merged_baseline_filt)[1:37], '_', 1), 
rownames(PWYs_merged_baseline_filt)[38:79])


#rownames(PWYs_ustek_baseline_filt) <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)

#merged_pathways <- rbind(PWYs_vedo_baseline_filt, 
PWYs_ustek_baseline_filt)
all_read_counts$Sample[1:43]<-  str_split_i(all_read_counts$Sample[1:43], 
'_', 1)

pathways_vedo_with_readcounts <- merge(PWYs_merged_baseline_filt, 
all_read_counts, by.x = 0, by.y = 'Sample')


#pathways_vedo_with_readcounts$merge_index <- NULL

pathways_vedo_with_readcounts$Sample <- NULL
#pathways_vedo_with_readcounts$Raw.reads..p1. <- NULL
##pathways_vedo_with_readcounts$Raw.reads..p2. <- NULL
#pathways_vedo_with_readcounts$Final..p2. <- NULL

#pathways_vedo_with_readcounts$sample_id_right <- NULL
rownames(pathways_vedo_with_readcounts) <- 
pathways_vedo_with_readcounts$Row.names
pathways_vedo_with_readcounts$Row.names <- NULL

pathways_vedo_with_readcounts <- pathways_vedo_with_readcounts %>%
  mutate(across(!c(Final..p1.), ~ . * Final..p1.))

pathways_vedo_with_readcounts$Final..p1. <- NULL


pathways_vedo_with_readcounts[pathways_vedo_with_readcounts == 0] <- NA

pathways_vedo_with_readcounts[sapply(pathways_vedo_with_readcounts, 
is.numeric)] <- 
lapply(pathways_vedo_with_readcounts[sapply(pathways_vedo_with_readcounts, 
is.numeric)], function(a) ifelse(is.na(a), min(a, na.rm = TRUE)/2, a))

Combined_pathways_codacore_input <- pathways_vedo_with_readcounts



# pathways codacore









# pathways vs clinical features



```


# codacore pathways Ustek 
```{r }

Meta_ustek$merging <- str_split_i(Meta_ustek$Fecal_sample_ID_1, '_', 1)
my_phenos=Meta_ustek
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)




#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]



taxa_cc=merge(my_phenos,Ustek_pathways_codacore_input,by.x 
="merging",by.y=0 )
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_6_month_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_6_month_pathways.csv')


```

# codacore pathways Ustek 2 year
```{r }

Meta_ustek$merging <- str_split_i(Meta_ustek$Fecal_sample_ID_1, '_', 1)
my_phenos=Meta_ustek
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)



Vedo_pathways_codacore_input
#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]



taxa_cc=merge(my_phenos,Ustek_pathways_codacore_input,by.x 
="merging",by.y=0 )
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]

predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_2_year_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_2_year_pathways.csv')


```



# codacore pathways Ustek clinical response
```{r }

Meta_ustek$merging <- str_split_i(Meta_ustek$Fecal_sample_ID_1, '_', 1)
my_phenos=Meta_ustek
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)




#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]



taxa_cc=merge(my_phenos,Ustek_pathways_codacore_input,by.x 
="merging",by.y=0 )
taxa_cc <- na.omit(taxa_cc)
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_hbi_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_hbi_pathways.csv')


```






# codacore pathways Vedo clinical response
```{r }
# change to vedo
#Meta_vedo$merging <- str_split_i(Meta_vedo$Fecal_sample_ID_1, '_', 1)
my_phenos=Meta_vedo
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)




#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

# vedo pathways codacore

taxa_cc=merge(my_phenos,Vedo_pathways_codacore_input,by.x 
="Fecal_sample_ID_1",by.y=0 )
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_hbi_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_hbi_pathways.csv')


```







# codacore pathways Vedo 6 month response
```{r }
# change to vedo
#Meta_vedo$merging <- str_split_i(Meta_vedo$Fecal_sample_ID_1, '_', 1)
my_phenos=Meta_vedo
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)




#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

# vedo pathways codacore

taxa_cc=merge(my_phenos,Vedo_pathways_codacore_input,by.x 
="Fecal_sample_ID_1",by.y=0 )
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_6_month_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_6_month_pathways.csv')


```







# codacore pathways Vedo 2 year response
```{r }
# change to vedo
#Meta_vedo$merging <- str_split_i(Meta_vedo$Fecal_sample_ID_1, '_', 1)
my_phenos=Meta_vedo
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)




#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

# vedo pathways codacore

taxa_cc=merge(my_phenos,Vedo_pathways_codacore_input,by.x 
="Fecal_sample_ID_1",by.y=0 )
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (verhuisd)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (overleden)", ]
predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_2_year_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_2_year_pathways.csv')


```




# codacore pathways combined clinical response
```{r }
# change to vedo
#Meta_vedo$merging <- str_split_i(Meta_vedo$Fecal_sample_ID_1, '_', 1)
Meta_merged$Fecal_sample_ID_1[1:45] <- 
str_split_i(Meta_merged$Fecal_sample_ID_1[1:45], '_', 1)
my_phenos=Meta_merged
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)




#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

# vedo pathways codacore

taxa_cc=merge(my_phenos,Combined_pathways_codacore_input,by.x 
="Fecal_sample_ID_1",by.y=0 )
taxa_cc <- na.omit(taxa_cc)
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_HBI_SCCAI != "", ]

predict_IBD=mtb_cc2[,"Response_HBI_SCCAI"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_nums_hbi_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_denums_hbi_pathways.csv')


```







# codacore pathways combined 6 month response
```{r }
# change to vedo
#Meta_vedo$merging <- str_split_i(Meta_vedo$Fecal_sample_ID_1, '_', 1)
Meta_merged$Fecal_sample_ID_1[1:45] <- 
str_split_i(Meta_merged$Fecal_sample_ID_1[1:45], '_', 1)

my_phenos=Meta_merged
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)




#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

# vedo pathways codacore

taxa_cc=merge(my_phenos,Combined_pathways_codacore_input,by.x 
="Fecal_sample_ID_1",by.y=0 )
taxa_cc <- na.omit(taxa_cc)
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$ResponseScoreYesNo != "", ]

predict_IBD=mtb_cc2[,"ResponseScoreYesNo"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_nums_6_month_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_denums_6_month_pathways.csv')


```







# codacore pathways combined 2 year response
```{r }
# change to vedo
#Meta_vedo$merging <- str_split_i(Meta_vedo$Fecal_sample_ID_1, '_', 1)
Meta_merged$Fecal_sample_ID_1[1:45] <- 
str_split_i(Meta_merged$Fecal_sample_ID_1[1:45], '_', 1)

my_phenos=Meta_merged
#PWYs_ustek_baseline_filt$merging <- 
str_split_i(rownames(PWYs_ustek_baseline_filt), '_', 1)


saved_pathways_combined <- 
merge(my_phenos,Combined_pathways_codacore_input,by.x 
="Fecal_sample_ID_1",by.y=0 )

#my_phenos=Ustek_clinical_data[,c("ResponseScoreYesNo","V1_FecalCalprotectine", 
"V1_BMI","V1_Sex","V1_CurrentIBDDiagnosis","V1_AgeatFecalSampling", 
"Response_2years", "Response_HBI_SCCAI", "Response_biochemical", 
"Response_biochemical_OR", "Response_FCP")]

# vedo pathways codacore

taxa_cc=merge(my_phenos,Combined_pathways_codacore_input,by.x 
="Fecal_sample_ID_1",by.y=0 )
taxa_cc <- na.omit(taxa_cc)
row.names(taxa_cc)=taxa_cc$Row.names
taxa_cc$Row.names=NULL
acc_list <- c()
num_list <- c()
denum_list <- c()
AUC_list <- c()
test_auc_list <- c()
mtb_cc2=taxa_cc

mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (verhuisd)", ]
mtb_cc2 = mtb_cc2[mtb_cc2$Response_2years != "NA (overleden)", ]
predict_IBD=mtb_cc2[,"Response_2years"]

metabolites=mtb_cc2[,17:ncol(mtb_cc2)]
counter <- 0
while (counter < 99){
  trainIndex <- sample(1:nrow(metabolites), 0.75 * nrow(metabolites))
  
  train_IBD_pheno <- predict_IBD[trainIndex]
  train_IBD_metabolite <- metabolites[trainIndex,]
  
  if(length(unique(train_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }
  
  test_IBD_pheno=predict_IBD[-trainIndex]
  test_IBD_metabolite=metabolites[-trainIndex,]
  if(length(unique(test_IBD_pheno)) == 1) {
    next  # Skip to the next iteration
  }

  possibleError <- tryCatch(
      thing(),
      error=function(e) e
  )

  
  possibleError <- tryCatch(
    {
    IBD_model=codacore(train_IBD_metabolite, train_IBD_pheno, logRatioType 
= 'amalgamations', lambda = 2)
  
    IBD_Hat <- predict(IBD_model, test_IBD_metabolite, logits = F, 
numLogRatios = 1)
    IBD_Hat2=IBD_Hat
    failure <- IBD_Hat2 < 0.5
    success <- IBD_Hat2 >= 0.5
    IBD_Hat2[failure] <- unique(predict_IBD)[1]
    IBD_Hat2[success] <- unique(predict_IBD)[2]
    

    test_auc <- pROC::auc(pROC::roc(test_IBD_pheno, IBD_Hat, quiet=T))
    nums <- colnames(metabolites[IBD_model$ensemble[[1]]$hard$numerator])
    denums <- 
colnames(metabolites[IBD_model$ensemble[[1]]$hard$denominator])
    AUC <- split(IBD_model$ensemble[[1]]$AUC, ' ')[[1]]
    acc <- round(mean(IBD_Hat2 == test_IBD_pheno), 2)
    num_list <- c(num_list, (nums))
    denum_list <- c(denum_list, (denums))
    test_auc_list <- c(test_auc_list, test_auc)
    AUC_list <- c(AUC_list, AUC)
    acc_list <- c(acc_list, acc)
    counter <- counter ++ 1
    },
    error=function(cond){
      #print(cond)
    }
  )
  if(inherits(possibleError, "error")) next
}

ustek_2_year_microbes_values <- c(mean(AUC_list), sd(AUC_list), 
mean(test_auc_list), sd(test_auc_list))
cat(mean(AUC_list))
cat(sd(AUC_list))
cat(mean(test_auc_list))
cat(sd(test_auc_list))





ustek_nums_2_year_microbes<- as.data.frame(table(num_list))
ustek_denums_2_year_microbes  <- as.data.frame(table(denum_list))


ustek_nums_2_year_microbes$values <- ustek_nums_2_year_microbes$Freq/100
ustek_denums_2_year_microbes$values <- 
ustek_denums_2_year_microbes$Freq/100




write.csv(ustek_nums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_nums_2_year_pathways.csv')
write.csv(ustek_denums_2_year_microbes, 
'/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_denums_2_year_pathways.csv')


```












## pathways plot dataprep
```{r }




# ustek again
path6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_6_month_pathways.csv')
path6_num$type <- 'Numerator'
path6_num$feature <- 'Pathway'
path6_num$response <- '6 month'
path6_num$medication <- 'Ustekinumab'
path6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_6_month_pathways.csv')
path6_denum$type <- 'Denominator'
path6_denum$feature <- 'Pathway'
path6_denum$response <- '6 month'
path6_denum$medication <- 'Ustekinumab'
path2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_2_year_pathways.csv')
path2_num$type <- 'Numerator'
path2_num$feature <- 'Pathway'
path2_num$response <- '2 year'
path2_num$medication <- 'Ustekinumab'
path2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_2_year_pathways.csv')
path2_denum$type <- 'Denominator'
path2_denum$feature <- 'Pathway'
path2_denum$response <- '2 year'
path2_denum$medication <- 'Ustekinumab'
pathhbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_hbi_pathways.csv')
pathhbi_num$type <- 'Numerator'
pathhbi_num$feature <- 'Pathway'
pathhbi_num$response <- 'HBI SCCAI'
pathhbi_num$medication <- 'Ustekinumab'
pathhbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_denums_hbi_pathways.csv')
pathhbi_denum$type <- 'Denominator'
pathhbi_denum$feature <- 'Pathway'
pathhbi_denum$response <- 'HBI SCCAI'
pathhbi_denum$medication <- 'Ustekinumab'



# vedo again 
vedo_path6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_6_month_pathways.csv')
vedo_path6_num$type <- 'Numerator'
vedo_path6_num$feature <- 'Pathway'
vedo_path6_num$response <- '6 month'
vedo_path6_num$medication <- 'Vedolizumab'
vedo_path6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_6_month_pathways.csv')
vedo_path6_denum$type <- 'Denominator'
vedo_path6_denum$feature <- 'Pathway'
vedo_path6_denum$response <- '6 month'
vedo_path6_denum$medication <- 'Vedolizumab'
vedo_path2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_2_year_pathways.csv')
vedo_path2_num$type <- 'Numerator'
vedo_path2_num$feature <- 'Pathway'
vedo_path2_num$response <- '2 year'
vedo_path2_num$medication <- 'Vedolizumab'
vedo_path2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_2_year_pathways.csv')
vedo_path2_denum$type <- 'Denominator'
vedo_path2_denum$feature <- 'Pathway'
vedo_path2_denum$response <- '2 year'
vedo_path2_denum$medication <- 'Vedolizumab'
vedo_pathhbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_nums_hbi_pathways.csv')
vedo_pathhbi_num$type <- 'Numerator'
vedo_pathhbi_num$feature <- 'Pathway'
vedo_pathhbi_num$response <- 'HBI SCCAI'
vedo_pathhbi_num$medication <- 'Vedolizumab'
vedo_pathhbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/vedo_denums_hbi_pathways.csv')
vedo_pathhbi_denum$type <- 'Denominator'
vedo_pathhbi_denum$feature <- 'Pathway'
vedo_pathhbi_denum$response <- 'HBI SCCAI'
vedo_pathhbi_denum$medication <- 'Vedolizumab'


# combined again 
all_path6_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_nums_6_month_pathways.csv')
all_path6_num$type <- 'Numerator'
all_path6_num$feature <- 'Pathway'
all_path6_num$response <- '6 month'
all_path6_num$medication <- 'Combined cohort'
all_path6_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_denums_6_month_pathways.csv')
all_path6_denum$type <- 'Denominator'
all_path6_denum$feature <- 'Pathway'
all_path6_denum$response <- '6 month'
all_path6_denum$medication <- 'Combined cohort'
all_path2_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_nums_2_year_pathways.csv')
all_path2_num$type <- 'Numerator'
all_path2_num$feature <- 'Pathway'
all_path2_num$response <- '2 year'
all_path2_num$medication <- 'Combined cohort'
all_path2_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_denums_2_year_pathways.csv')
all_path2_denum$type <- 'Denominator'
all_path2_denum$feature <- 'Pathway'
all_path2_denum$response <- '2 year'
all_path2_denum$medication <- 'Combined cohort'
all_pathhbi_num <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_nums_hbi_pathways.csv')
all_pathhbi_num$type <- 'Numerator'
all_pathhbi_num$feature <- 'Pathway'
all_pathhbi_num$response <- 'HBI SCCAI'
all_pathhbi_num$medication <- 'Combined cohort'
all_pathhbi_denum <- 
read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/combined_denums_hbi_pathways.csv')
all_pathhbi_denum$type <- 'Denominator'
all_pathhbi_denum$feature <- 'Pathway'
all_pathhbi_denum$response <- 'HBI SCCAI'
all_pathhbi_denum$medication <- 'Combined cohort'






#read_csv('/Users/iwan/Research/drug_response/metadata/lastest_metadata/ustek_nums_6_month_microbes.csv')
colnames(path6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(path6_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(path2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(path2_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(pathhbi_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(pathhbi_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_path6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_path6_denum) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")
colnames(vedo_path2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(vedo_path2_denum) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")
colnames(vedo_pathhbi_num) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")
colnames(vedo_pathhbi_denum) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")

colnames(all_path6_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_path6_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_path2_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_path2_denum) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_pathhbi_num) <- c("pass","predict", "freq", "values", "type", 
"feature", "response", "Medication")
colnames(all_pathhbi_denum) <- c("pass","predict", "freq", "values", 
"type", "feature", "response", "Medication")




stacked_df_pathways <- do.call(rbind, list(path6_num, path6_denum, 
path2_num,path2_denum,pathhbi_num, pathhbi_denum,
                                        vedo_path6_num, vedo_path6_denum, 
vedo_path2_num,vedo_path2_denum,vedo_pathhbi_num, vedo_pathhbi_denum,
                                        all_path6_num, all_path6_denum, 
all_path2_num,all_path2_denum,all_pathhbi_num, all_pathhbi_denum))





metabolite_names_df <- 
read.csv("/Users/iwan/Research/drug_response/metadata/lastest_metadata/metabolites_naming_key.txt", 
sep='\t', header = TRUE)
metabolite_names <- c()
for (i in stacked_df_metas$predict){
  metabolite_names <- c(metabolite_names, 
metabolite_names_df$CHEMICAL_NAME[metabolite_names_df$XCHEM_ID == i])
}


shortnames <- c()
for (i in stacked_df_bugs$predict){
  shortnames <- c(shortnames, str_split_i(i, '__',8))
}

stacked_df_bugs$new_predict <- shortnames
stacked_df_metas$new_predict <- metabolite_names
stacked_df_pathways$new_predict <- stacked_df_pathways$predict

stacked_df <- rbind(stacked_df_bugs, stacked_df_metas, 
stacked_df_pathways)

```



```{r}

library(ggh4x)



stacked_df_pathways_combined_only <- 
stacked_df_pathways[stacked_df_pathways$Medication == 'Combined cohort',]
stacked_df_pathways_combined_only_6_month <-  
stacked_df_pathways_combined_only[stacked_df_pathways_combined_only$response 
== '6 month',]
#x=reorder(name,depth)
bp <- 
ggplot(stacked_df_pathways_combined_only_6_month[stacked_df_pathways_combined_only_6_month$values 
> 0.1,], aes(x = response, y = reorder(predict, values), fill = values)) +
  geom_tile() + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  facet_nested(( ~  factor(Medication, levels =c('Vedolizumab', 
'Ustekinumab', 'Combined cohort')) +factor(type, levels=c('Numerator', 
'Denominator')) ), scales = "free", space = "free") +
  ggtitle("Frequency of metabolites used in CoDaCoRe predictions") +
  xlab("Response definition")+
  labs(fill="Frequency")+
  ylab("Features") #, title = "Features used in Ustek response 
definition") 


bp

stacked_df_combined_only <- stacked_df[stacked_df$Medication == 'Combined 
cohort',]
stacked_df_combined_only_6_month <-  
stacked_df_combined_only[stacked_df_combined_only$response == '6 month',]




pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/all_predictors_combined.pdf")
bp <- 
ggplot(stacked_df_combined_only_6_month[stacked_df_combined_only_6_month$values 
> 0.1,], aes(x = response, y = reorder(new_predict, values), fill = 
values)) +
  geom_tile() + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  facet_nested(( factor(feature, levels=c('Microbe', 'Metabolite', 
'Pathway'))~  factor(Medication, levels =c('Vedolizumab', 'Ustekinumab', 
'Combined cohort')) +factor(type, levels=c('Numerator', 'Denominator')) ), 
scales = "free", space = "free") +
  ggtitle("Frequency of features use in CoDaCoRe predictions") +
  xlab("Response definition")+
  ylab("Features") #, title = "Features used in Ustek response 
definition") 


bp

dev.off()
```



# plot the box/violin of the individual pathways 6 months combined
```{r}
stacked_df_bugs_filtered <- stacked_df_pathways[stacked_df_pathways$values 
> 0.1,]

stacked_df_bugs_filtered_combined <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]



stacked_df_bugs_filtered_combined_6month <- 
stacked_df_bugs_filtered_combined[stacked_df_bugs_filtered_combined$response 
== '6 month',]
stacked_df_bugs_filtered_combined_6month_num <- 
stacked_df_bugs_filtered_combined_6month[stacked_df_bugs_filtered_combined_6month$type 
== 'Numerator',]
stacked_df_bugs_filtered_combined_6month_denum<- 
stacked_df_bugs_filtered_combined_6month[stacked_df_bugs_filtered_combined_6month$type 
== 'Denominator',]



combined_responders_pathways_num <- 
saved_pathways_combined[,stacked_df_bugs_filtered_combined_6month_num$predict][saved_pathways_combined$ResponseScoreYesNo 
== 'yes',] 

combined_non_responders_pathways_num <- 
saved_pathways_combined[,stacked_df_bugs_filtered_combined_6month_num$predict][saved_pathways_combined$ResponseScoreYesNo 
== 'no',] 

combined_responders_pathways_denum <- 
saved_pathways_combined[,stacked_df_bugs_filtered_combined_6month_denum$predict][saved_pathways_combined$ResponseScoreYesNo 
== 'yes',]

combined_non_responders_pathways_denum <- 
saved_pathways_combined[,stacked_df_bugs_filtered_combined_6month_denum$predict][saved_pathways_combined$ResponseScoreYesNo 
== 'no',]


responder_ratio <- rowSums(combined_responders_pathways_num) / 
rowSums(combined_responders_pathways_denum)

non_responder_ratio <- rowSums(combined_non_responders_pathways_num) / 
rowSums(combined_non_responders_pathways_denum)


# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(responder_ratio)), rep(c("Non 
Responder"), each = length(non_responder_ratio))),
  Ratio = c(log(responder_ratio), log(non_responder_ratio))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)
pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/6_month_pathways.pdf")
# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Pathways ratio overview 6 month response in the combined 
cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)


dev.off()
responder_ratio_pathways <- rowSums(combined_responders_pathways_num) / 
rowSums(combined_responders_pathways_denum)

non_responder_ratio_pathways <- 
rowSums(combined_non_responders_pathways_num) / 
rowSums(combined_non_responders_pathways_denum)



```




```{r }

combined_ratios_responder <- (responder_ratio_pathways + 
responder_ratio_metabolites + responder_ratio_microbes) / 3
combined_ratios_non_responder <- (non_responder_ratio_pathways + 
non_responder_ratio_metabolites + non_responder_ratio_microbes) / 3






# Combine data into a data frame
data <- data.frame(
  Group = c(rep(c("Responder"), each=length(combined_ratios_responder)), 
rep(c("Non Responder"), each = length(combined_ratios_non_responder))),
  Ratio = c(log(combined_ratios_responder), 
log(combined_ratios_non_responder))
)

# Load required libraries
library(ggplot2)
library(ggdist)
library(ggridges)
library(gghalves)

pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/combined_ratio.pdf")
# Create the ggplot
p <- ggplot(data, aes(x = factor(Group, levels=c('Responder', 'Non 
Responder')) , y = Ratio, fill=Group)) +
  
  # Boxplot with individual data points
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  geom_point( alpha = 0.5, fill = "lightblue", position = 
position_jitter(width=0.1)) +
  geom_half_violin(side='r', nudge=0.15, alpha =0.8, color ='white') +
  
  # Customize appearance
  theme_bw() +
  labs(title = "Combined ratio overview 6 month response in the combined 
cohort",
       x = "",
       y = "Ratio")  +
  scale_fill_manual(values=c("#FF6633", "#339900"))

# Print the plot 
# position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), 
size = 2
print(p)
dev.off()
```


```{r }





my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


improvement_aucs <- c()
partial_model <- c()
full_model <- c()
partial_model_spec <- c()
partial_model_sens <- c()
full_model_spec <- c()
full_model_sens <- c()

micro_model_spec <- c()
micro_model_sens <- c()

meta_model_spec <- c()
meta_model_sens <- c()
path_model_spec <- c()
path_model_sens <- c()

paths_model <- c()
meta_model <- c()
micro_model <- c()

for (i in 1:1000){
        my_phenos_no_na$ratio_microbiome <- NULL
        my_phenos_no_na$ratio_metabolites <- NULL
        my_phenos_no_na$ratio_pathways <- NULL
        
        trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        if(length(unique(train_IBD_pheno_y[trainIndex])) == 1) {
          next  # Skip to the next iteration
        }

        if(length(unique(train_IBD_pheno_y[-trainIndex])) == 1) {
           next  # Skip to the next iteration
        }
        #Regress out age, sex, bmi, calprotectin to IBD
        partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        #Re-calculate ratios after adjusting by covariates 
        partial_IBD_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(partial_IBD, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        
        evaluator <- c(pROC::roc(train_IBD_pheno_y[-trainIndex], 
predict(partial_IBD, my_phenos_no_na[-trainIndex,]), quiet=T))
        partial_model_spec <- c(partial_model_spec, 
evaluator$specificities)
        partial_model_sens <- c(partial_model_sens, 
evaluator$sensitivities)
        
        
        # calculate ratios for vedo samples
        ######### ratio bugs
        stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values 
> 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_microbiome <- vedo_6_month_microbe_ratio
        
        
        
        
        
        ############ ratio metabolites
        
        
        stacked_df_bugs_filtered <- 
stacked_df_metas[stacked_df_metas$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(combined_metabolites_only[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_metabolites_only[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio

        
        
        ########### ratio pathways
        
        stacked_df_bugs_filtered <- 
stacked_df_pathways[stacked_df_pathways$values > 0.1,]
        
        stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
        stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
        stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
        stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
        vedo_6_month_microbe_ratio <- 
rowSums(Combined_pathways_codacore_input[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Combined_pathways_codacore_input[,stacked_df_bugs_vedo_6month_denum$predict])
        
        my_phenos$ratio_pathways <- vedo_6_month_microbe_ratio
        
        
        #### here we add the prediction based on only the ratio see if 
that works : 
        my_phenos_no_na <- drop_na(my_phenos)
        
        my_phenos_no_na_microbiome <- 
my_phenos_no_na[,c('ResponseScoreYesNo', 'ratio_microbiome')]
        my_phenos_no_na_metabolites <- 
my_phenos_no_na[,c('ResponseScoreYesNo', 'ratio_metabolites')]
        my_phenos_no_na_pathways <- 
my_phenos_no_na[,c('ResponseScoreYesNo', 'ratio_pathways')]
        
        
        train_correct_IBD=my_phenos_no_na_microbiome[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        micro_ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na_microbiome[-trainIndex,]), quiet=T))
        evaluator <- c(pROC::roc(train_IBD_pheno_y[-trainIndex], 
predict(ratio_model, my_phenos_no_na_microbiome[-trainIndex,]), quiet=T))
        micro_model_spec <- c(micro_model_spec, evaluator$specificities)
        micro_model_sens <- c(micro_model_sens, evaluator$sensitivities)
        
        
        
        train_correct_IBD=my_phenos_no_na_metabolites[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        metas_ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na_metabolites[-trainIndex,]), quiet=T))
        evaluator <- c(pROC::roc(train_IBD_pheno_y[-trainIndex], 
predict(ratio_model, my_phenos_no_na_metabolites[-trainIndex,]), quiet=T))
        meta_model_spec <- c(meta_model_spec, evaluator$specificities)
        meta_model_sens <- c(meta_model_sens, evaluator$sensitivities)
        
        train_correct_IBD=my_phenos_no_na_pathways[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}

        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        pathways_ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na_pathways[-trainIndex,]), quiet=T))
        evaluator <- c(pROC::roc(train_IBD_pheno_y[-trainIndex], 
predict(ratio_model, my_phenos_no_na_pathways[-trainIndex,]), quiet=T))
        path_model_spec <- c(path_model_spec, evaluator$specificities)
        path_model_sens <- c(path_model_sens, evaluator$sensitivities)
        
        
        # add new training data with ratio microbe
        
        my_phenos_no_na <- drop_na(my_phenos)
        
        
        train_correct_IBD=my_phenos_no_na[trainIndex,]
        
        train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
        for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
        ratio_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        ratio_model_AUC <- 
pROC::auc(pROC::roc(train_IBD_pheno_y[-trainIndex], predict(ratio_model, 
my_phenos_no_na[-trainIndex,]), quiet=T))
        evaluator <- c(pROC::roc(train_IBD_pheno_y[-trainIndex], 
predict(ratio_model, my_phenos_no_na[-trainIndex,]), quiet=T))
        full_model_spec <- c(full_model_spec, evaluator$specificities)
        full_model_sens <- c(full_model_sens, evaluator$sensitivities)
        
        improvement_aucs <- c(improvement_aucs, ratio_model_AUC - 
partial_IBD_AUC)
        partial_model <- c(partial_model, partial_IBD_AUC)
        full_model <- c(full_model, ratio_model_AUC)
        micro_model <- c(micro_model, micro_ratio_model_AUC)
        meta_model <- c(meta_model, metas_ratio_model_AUC)
        paths_model <- c(paths_model, pathways_ratio_model_AUC)

}


cat("Average AUC partial model: ", mean(partial_model), "with sd: ", 
sd(partial_model), '\n')
cat("Average AUC model with ratio: ", mean(full_model), "with sd: ", 
sd(full_model), '\n')
cat("Average AUC improvement by including ratio: ", 
mean(improvement_aucs), "with sd: ", sd(improvement_aucs))

cat("Average AUC model with micro ratio: ", mean(micro_model), "with sd: 
", sd(micro_model), '\n')
cat("Average AUC model with paths ratio: ", mean(paths_model), "with sd: 
", sd(paths_model), '\n')
cat("Average AUC model with meta ratio: ", mean(meta_model), "with sd: ", 
sd(meta_model), '\n')


```

library(pROC)

ggroc
17 datapoints per curve


```{r }

# Define the number of points per line
points_per_line <- 17

num_lines <- length(full_model_spec[1:170]) %/% points_per_line

#num_lines = 10
# Create a data frame for ggplot
df <- data.frame(Specificity = unlist(lapply(full_model_spec[1:170], 
function(x) rep(x, num_lines))),
                 Sensitivity = unlist(lapply(full_model_sens[1:170], 
function(x) rep(x, num_lines))),
                 Line = rep(1:num_lines, each = 
length(full_model_spec[1:170])))

# Plot ROC curves using ggplot
ggplot(df, aes(x = 1 - Specificity, y = Sensitivity, group = Line, color = 
as.factor(Line))) +
  geom_line() +
  #geom_point() +
  labs(x = "1 - Specificity", y = "Sensitivity", title = "ROC Curves") +
  theme_minimal()+ theme(legend.position="none")

full_model_spec[1:17]
full_model_sens[1:17]

spec_counts_1 <- c()
spec_counts_2 <- c()
spec_counts_3 <- c()
spec_counts_4 <- c()
spec_counts_5 <- c()
spec_counts_6 <- c()
spec_counts_7 <- c()
spec_counts_8 <- c()
spec_counts_9 <- c()
spec_counts_10 <- c()
spec_counts_11 <- c()
spec_counts_12 <- c()
spec_counts_13 <- c()
spec_counts_14 <- c()
spec_counts_15 <- c()
spec_counts_16 <- c()
spec_counts_17 <- c()
for (i in seq(1, length(full_model_spec), 17)){
  spec_counts_1<- c(spec_counts_1, full_model_spec[i])
  spec_counts_2<- c(spec_counts_2, full_model_spec[i+1])
  spec_counts_3<- c(spec_counts_3, full_model_spec[i+2])
  spec_counts_4<- c(spec_counts_4, full_model_spec[i+3])
  spec_counts_5<- c(spec_counts_5, full_model_spec[i+4])
  spec_counts_6<- c(spec_counts_6, full_model_spec[i+5])
  spec_counts_7<- c(spec_counts_7, full_model_spec[i+6])
  spec_counts_8<- c(spec_counts_8, full_model_spec[i+7])
  spec_counts_9<- c(spec_counts_9, full_model_spec[i+8])
  spec_counts_10<- c(spec_counts_10, full_model_spec[i+9])
  spec_counts_11<- c(spec_counts_11, full_model_spec[i+10])
  spec_counts_12<- c(spec_counts_12, full_model_spec[i+11])
  spec_counts_13<- c(spec_counts_13, full_model_spec[i+12])
  spec_counts_14<- c(spec_counts_14, full_model_spec[i+13])
  spec_counts_15<- c(spec_counts_15, full_model_spec[i+14])
  spec_counts_16<- c(spec_counts_16, full_model_spec[i+15])
  spec_counts_17<- c(spec_counts_17, full_model_spec[i+16])
}
sd(spec_counts_11)






sens_counts_1 <- c()
sens_counts_2 <- c()
sens_counts_3 <- c()
sens_counts_4 <- c()
sens_counts_5 <- c()
sens_counts_6 <- c()
sens_counts_7 <- c()
sens_counts_8 <- c()
sens_counts_9 <- c()
sens_counts_10 <- c()
sens_counts_11 <- c()
sens_counts_12 <- c()
sens_counts_13 <- c()
sens_counts_14 <- c()
sens_counts_15 <- c()
sens_counts_16 <- c()
sens_counts_17 <- c()
for (i in seq(1, length(full_model_sens), 17)){
  sens_counts_1<- c(sens_counts_1, full_model_sens[i])
  sens_counts_2<- c(sens_counts_2, full_model_sens[i+1])
  sens_counts_3<- c(sens_counts_3, full_model_sens[i+2])
  sens_counts_4<- c(sens_counts_4, full_model_sens[i+3])
  sens_counts_5<- c(sens_counts_5, full_model_sens[i+4])
  sens_counts_6<- c(sens_counts_6, full_model_sens[i+5])
  sens_counts_7<- c(sens_counts_7, full_model_sens[i+6])
  sens_counts_8<- c(sens_counts_8, full_model_sens[i+7])
  sens_counts_9<- c(sens_counts_9, full_model_sens[i+8])
  sens_counts_10<- c(sens_counts_10, full_model_sens[i+9])
  sens_counts_11<- c(sens_counts_11, full_model_sens[i+10])
  sens_counts_12<- c(sens_counts_12, full_model_sens[i+11])
  sens_counts_13<- c(sens_counts_13, full_model_sens[i+12])
  sens_counts_14<- c(sens_counts_14, full_model_sens[i+13])
  sens_counts_15<- c(sens_counts_15, full_model_sens[i+14])
  sens_counts_16<- c(sens_counts_16, full_model_sens[i+15])
  sens_counts_17<- c(sens_counts_17, full_model_sens[i+16])
}


mean(sens_counts_5)
sd(spec_counts_2)



x_full_data <- c(1-mean(sens_counts_1), 1-mean(sens_counts_2), 
1-mean(sens_counts_3), 1-mean(sens_counts_4), 1-mean(sens_counts_5), 
            1-mean(sens_counts_6), 1-mean(sens_counts_7), 
1-mean(sens_counts_8), 1-mean(sens_counts_9), 1-mean(sens_counts_10), 
            1-mean(sens_counts_11), 1-mean(sens_counts_12),1- 
mean(sens_counts_13), 1-mean(sens_counts_14), 1-mean(sens_counts_15), 
            1-mean(sens_counts_16), 1-mean(sens_counts_17))
y_full_data <- c(mean(spec_counts_1), mean(spec_counts_2), 
mean(spec_counts_3), mean(spec_counts_4), mean(spec_counts_5), 
            mean(spec_counts_6), mean(spec_counts_7), mean(spec_counts_8), 
mean(spec_counts_9), mean(spec_counts_10), 
            mean(spec_counts_11), mean(spec_counts_12), 
mean(spec_counts_13), mean(spec_counts_14), mean(spec_counts_15), 
            mean(spec_counts_16), mean(spec_counts_17))

x_full_sd <- c(sd(sens_counts_1), sd(sens_counts_2), sd(sens_counts_3), 
sd(sens_counts_4), sd(sens_counts_5), 
          sd(sens_counts_6), sd(sens_counts_7), sd(sens_counts_8), 
sd(sens_counts_9), sd(sens_counts_10), 
          sd(sens_counts_11), sd(sens_counts_12), sd(sens_counts_13), 
sd(sens_counts_14), sd(sens_counts_15), 
          sd(sens_counts_16), sd(sens_counts_17))
y_full_sd <- c(sd(spec_counts_1), sd(spec_counts_2), sd(spec_counts_3), 
sd(spec_counts_4), sd(spec_counts_5), 
          sd(spec_counts_6), sd(spec_counts_7), sd(spec_counts_8), 
sd(spec_counts_9), sd(spec_counts_10), 
          sd(spec_counts_11), sd(spec_counts_12), sd(spec_counts_13), 
sd(spec_counts_14), sd(spec_counts_15), 
          sd(spec_counts_16), sd(spec_counts_17))


spec_counts_1 <- c()
spec_counts_2 <- c()
spec_counts_3 <- c()
spec_counts_4 <- c()
spec_counts_5 <- c()
spec_counts_6 <- c()
spec_counts_7 <- c()
spec_counts_8 <- c()
spec_counts_9 <- c()
spec_counts_10 <- c()
spec_counts_11 <- c()
spec_counts_12 <- c()
spec_counts_13 <- c()
spec_counts_14 <- c()
spec_counts_15 <- c()
spec_counts_16 <- c()
spec_counts_17 <- c()
for (i in seq(1, length(partial_model_spec), 17)){
  spec_counts_1<- c(spec_counts_1, partial_model_spec[i])
  spec_counts_2<- c(spec_counts_2, partial_model_spec[i+1])
  spec_counts_3<- c(spec_counts_3, partial_model_spec[i+2])
  spec_counts_4<- c(spec_counts_4, partial_model_spec[i+3])
  spec_counts_5<- c(spec_counts_5, partial_model_spec[i+4])
  spec_counts_6<- c(spec_counts_6, partial_model_spec[i+5])
  spec_counts_7<- c(spec_counts_7, partial_model_spec[i+6])
  spec_counts_8<- c(spec_counts_8, partial_model_spec[i+7])
  spec_counts_9<- c(spec_counts_9, partial_model_spec[i+8])
  spec_counts_10<- c(spec_counts_10, partial_model_spec[i+9])
  spec_counts_11<- c(spec_counts_11, partial_model_spec[i+10])
  spec_counts_12<- c(spec_counts_12, partial_model_spec[i+11])
  spec_counts_13<- c(spec_counts_13, partial_model_spec[i+12])
  spec_counts_14<- c(spec_counts_14, partial_model_spec[i+13])
  spec_counts_15<- c(spec_counts_15, partial_model_spec[i+14])
  spec_counts_16<- c(spec_counts_16, partial_model_spec[i+15])
  spec_counts_17<- c(spec_counts_17, partial_model_spec[i+16])
}
sd(na.omit(spec_counts_11))






sens_counts_1 <- c()
sens_counts_2 <- c()
sens_counts_3 <- c()
sens_counts_4 <- c()
sens_counts_5 <- c()
sens_counts_6 <- c()
sens_counts_7 <- c()
sens_counts_8 <- c()
sens_counts_9 <- c()
sens_counts_10 <- c()
sens_counts_11 <- c()
sens_counts_12 <- c()
sens_counts_13 <- c()
sens_counts_14 <- c()
sens_counts_15 <- c()
sens_counts_16 <- c()
sens_counts_17 <- c()
for (i in seq(1, length(partial_model_sens), 17)){
  sens_counts_1<- c(sens_counts_1, partial_model_sens[i])
  sens_counts_2<- c(sens_counts_2, partial_model_sens[i+1])
  sens_counts_3<- c(sens_counts_3, partial_model_sens[i+2])
  sens_counts_4<- c(sens_counts_4, partial_model_sens[i+3])
  sens_counts_5<- c(sens_counts_5, partial_model_sens[i+4])
  sens_counts_6<- c(sens_counts_6, partial_model_sens[i+5])
  sens_counts_7<- c(sens_counts_7, partial_model_sens[i+6])
  sens_counts_8<- c(sens_counts_8, partial_model_sens[i+7])
  sens_counts_9<- c(sens_counts_9, partial_model_sens[i+8])
  sens_counts_10<- c(sens_counts_10, partial_model_sens[i+9])
  sens_counts_11<- c(sens_counts_11, partial_model_sens[i+10])
  sens_counts_12<- c(sens_counts_12, partial_model_sens[i+11])
  sens_counts_13<- c(sens_counts_13, partial_model_sens[i+12])
  sens_counts_14<- c(sens_counts_14, partial_model_sens[i+13])
  sens_counts_15<- c(sens_counts_15, partial_model_sens[i+14])
  sens_counts_16<- c(sens_counts_16, partial_model_sens[i+15])
  sens_counts_17<- c(sens_counts_17, partial_model_sens[i+16])
}





x_partial_data <- c(1-mean(na.omit(sens_counts_1)), 
1-mean(na.omit(sens_counts_2)), 1-mean(na.omit(sens_counts_3)), 
1-mean(na.omit(sens_counts_4)), 1-mean(na.omit(sens_counts_5)), 
            1-mean(na.omit(sens_counts_6)), 
1-mean(na.omit(sens_counts_7)), 1-mean(na.omit(sens_counts_8)), 
1-mean(na.omit(sens_counts_9)), 1-mean(na.omit(sens_counts_10)), 
            1-mean(na.omit(sens_counts_11)), 
1-mean(na.omit(sens_counts_12)),1- mean(na.omit(sens_counts_13)), 
1-mean(na.omit(sens_counts_14)), 1-mean(na.omit(sens_counts_15)), 
            1-mean(na.omit(sens_counts_16)), 
1-mean(na.omit(sens_counts_17)))
y_partial_data <- c(mean(na.omit(spec_counts_1)), 
mean(na.omit(spec_counts_2)), mean(na.omit(spec_counts_3)), 
mean(na.omit(spec_counts_4)), mean(na.omit(spec_counts_5)), 
            mean(na.omit(spec_counts_6)), mean(na.omit(spec_counts_7)), 
mean(na.omit(spec_counts_8)), mean(na.omit(spec_counts_9)), 
mean(na.omit(spec_counts_10)), 
            mean(na.omit(spec_counts_11)), mean(na.omit(spec_counts_12)), 
mean(na.omit(spec_counts_13)), mean(na.omit(spec_counts_14)), 
mean(na.omit(spec_counts_15)), 
            mean(na.omit(spec_counts_16)), mean(na.omit(spec_counts_17)))

x_partial_sd <- c(sd(na.omit(sens_counts_1)), sd(na.omit(sens_counts_2)), 
sd(na.omit(sens_counts_3)), sd(na.omit(sens_counts_4)), 
sd(na.omit(sens_counts_5)), 
          sd(na.omit(sens_counts_6)), sd(na.omit(sens_counts_7)), 
sd(na.omit(sens_counts_8)), sd(na.omit(sens_counts_9)), 
sd(na.omit(sens_counts_10)), 
          sd(na.omit(sens_counts_11)), sd(na.omit(sens_counts_12)), 
sd(na.omit(sens_counts_13)), sd(na.omit(sens_counts_14)), 
sd(na.omit(sens_counts_15)), 
          sd(na.omit(sens_counts_16)), sd(na.omit(sens_counts_17)))
y_partial_sd <- c(sd(na.omit(spec_counts_1)), sd(na.omit(spec_counts_2)), 
sd(na.omit(spec_counts_3)), sd(na.omit(spec_counts_4)), 
sd(na.omit(spec_counts_5)), 
          sd(na.omit(spec_counts_6)), sd(na.omit(spec_counts_7)), 
sd(na.omit(spec_counts_8)), sd(na.omit(spec_counts_9)), 
sd(na.omit(spec_counts_10)), 
          sd(na.omit(spec_counts_11)), sd(na.omit(spec_counts_12)), 
sd(na.omit(spec_counts_13)), sd(na.omit(spec_counts_14)), 
sd(na.omit(spec_counts_15)), 
          sd(na.omit(spec_counts_16)), sd(na.omit(spec_counts_17)))


new_x_partial_data <- c(x_partial_data[2], 
x_partial_data[3],x_partial_data[4],x_partial_data[5],x_partial_data[6],x_partial_data[7],x_partial_data[8],x_partial_data[9],x_partial_data[10],x_partial_data[11],x_partial_data[12],x_partial_data[13],x_partial_data[14],x_partial_data[15],x_partial_data[16],x_partial_data[17],x_partial_data[1])
new_y_partial_data <- c(y_partial_data[2], 
y_partial_data[3],y_partial_data[4],y_partial_data[5],y_partial_data[6],y_partial_data[7],y_partial_data[8],y_partial_data[9],y_partial_data[10],y_partial_data[11],y_partial_data[12],y_partial_data[13],y_partial_data[14],y_partial_data[15],y_partial_data[16],y_partial_data[17],y_partial_data[1])
new_y_partial_sd <- c(y_partial_sd[2], 
y_partial_sd[3],y_partial_sd[4],y_partial_sd[5],y_partial_sd[6],y_partial_sd[7],y_partial_sd[8],y_partial_sd[9],y_partial_sd[10],y_partial_sd[11],y_partial_sd[12],y_partial_sd[13],y_partial_sd[14],y_partial_sd[15],y_partial_sd[16],y_partial_sd[17],y_partial_sd[1])


plot(x_full_data, y_full_data, type='l')
plot(new_x_partial_data, new_y_partial_data, type='l')
```


### here we prep the lines for the roc auc plot
```{r }



spec_df <- micro_model_spec

spec_counts_1 <- c()
spec_counts_2 <- c()
spec_counts_3 <- c()
spec_counts_4 <- c()
spec_counts_5 <- c()
spec_counts_6 <- c()
spec_counts_7 <- c()
spec_counts_8 <- c()
spec_counts_9 <- c()
spec_counts_10 <- c()
spec_counts_11 <- c()
spec_counts_12 <- c()
spec_counts_13 <- c()
spec_counts_14 <- c()
spec_counts_15 <- c()
spec_counts_16 <- c()
spec_counts_17 <- c()
for (i in seq(1, length(spec_df), 17)){
  spec_counts_1<- c(spec_counts_1, spec_df[i])
  spec_counts_2<- c(spec_counts_2, spec_df[i+1])
  spec_counts_3<- c(spec_counts_3, spec_df[i+2])
  spec_counts_4<- c(spec_counts_4, spec_df[i+3])
  spec_counts_5<- c(spec_counts_5, spec_df[i+4])
  spec_counts_6<- c(spec_counts_6, spec_df[i+5])
  spec_counts_7<- c(spec_counts_7, spec_df[i+6])
  spec_counts_8<- c(spec_counts_8, spec_df[i+7])
  spec_counts_9<- c(spec_counts_9, spec_df[i+8])
  spec_counts_10<- c(spec_counts_10, spec_df[i+9])
  spec_counts_11<- c(spec_counts_11, spec_df[i+10])
  spec_counts_12<- c(spec_counts_12, spec_df[i+11])
  spec_counts_13<- c(spec_counts_13, spec_df[i+12])
  spec_counts_14<- c(spec_counts_14, spec_df[i+13])
  spec_counts_15<- c(spec_counts_15, spec_df[i+14])
  spec_counts_16<- c(spec_counts_16, spec_df[i+15])
  spec_counts_17<- c(spec_counts_17, spec_df[i+16])
}




sens_df <- micro_model_sens

sens_counts_1 <- c()
sens_counts_2 <- c()
sens_counts_3 <- c()
sens_counts_4 <- c()
sens_counts_5 <- c()
sens_counts_6 <- c()
sens_counts_7 <- c()
sens_counts_8 <- c()
sens_counts_9 <- c()
sens_counts_10 <- c()
sens_counts_11 <- c()
sens_counts_12 <- c()
sens_counts_13 <- c()
sens_counts_14 <- c()
sens_counts_15 <- c()
sens_counts_16 <- c()
sens_counts_17 <- c()
for (i in seq(1, length(sens_df), 17)){
  sens_counts_1<- c(sens_counts_1, sens_df[i])
  sens_counts_2<- c(sens_counts_2, sens_df[i+1])
  sens_counts_3<- c(sens_counts_3, sens_df[i+2])
  sens_counts_4<- c(sens_counts_4, sens_df[i+3])
  sens_counts_5<- c(sens_counts_5, sens_df[i+4])
  sens_counts_6<- c(sens_counts_6, sens_df[i+5])
  sens_counts_7<- c(sens_counts_7, sens_df[i+6])
  sens_counts_8<- c(sens_counts_8, sens_df[i+7])
  sens_counts_9<- c(sens_counts_9, sens_df[i+8])
  sens_counts_10<- c(sens_counts_10, sens_df[i+9])
  sens_counts_11<- c(sens_counts_11, sens_df[i+10])
  sens_counts_12<- c(sens_counts_12, sens_df[i+11])
  sens_counts_13<- c(sens_counts_13, sens_df[i+12])
  sens_counts_14<- c(sens_counts_14, sens_df[i+13])
  sens_counts_15<- c(sens_counts_15, sens_df[i+14])
  sens_counts_16<- c(sens_counts_16, sens_df[i+15])
  sens_counts_17<- c(sens_counts_17, sens_df[i+16])
}


x_micro_data <- c(1-mean(sens_counts_1, na.rm = TRUE), 
1-mean(sens_counts_2, na.rm = TRUE), 1-mean(sens_counts_3, na.rm = TRUE), 
1-mean(sens_counts_4, na.rm = TRUE), 1-mean(sens_counts_5, na.rm = TRUE), 
            1-mean(sens_counts_6, na.rm = TRUE), 1-mean(sens_counts_7, 
na.rm = TRUE), 1-mean(sens_counts_8, na.rm = TRUE), 1-mean(sens_counts_9, 
na.rm = TRUE), 1-mean(sens_counts_10, na.rm = TRUE), 
            1-mean(sens_counts_11, na.rm = TRUE), 1-mean(sens_counts_12, 
na.rm = TRUE),1- mean(sens_counts_13, na.rm = TRUE), 
1-mean(sens_counts_14, na.rm = TRUE), 1-mean(sens_counts_15, na.rm = 
TRUE), 
            1-mean(sens_counts_16, na.rm = TRUE), 1-mean(sens_counts_17, 
na.rm = TRUE))
y_micro_data <- c(mean(spec_counts_1, na.rm = TRUE), mean(spec_counts_2, 
na.rm = TRUE), mean(spec_counts_3, na.rm = TRUE), mean(spec_counts_4, 
na.rm = TRUE), mean(spec_counts_5, na.rm = TRUE), 
            mean(spec_counts_6, na.rm = TRUE), mean(spec_counts_7, na.rm = 
TRUE), mean(spec_counts_8, na.rm = TRUE), mean(spec_counts_9, na.rm = 
TRUE), mean(spec_counts_10, na.rm = TRUE), 
            mean(spec_counts_11, na.rm = TRUE), mean(spec_counts_12, na.rm 
= TRUE), mean(spec_counts_13, na.rm = TRUE), mean(spec_counts_14, na.rm = 
TRUE), mean(spec_counts_15, na.rm = TRUE), 
            mean(spec_counts_16, na.rm = TRUE), mean(spec_counts_17, na.rm 
= TRUE))

x_micro_sd <- c(sd(sens_counts_1, na.rm = TRUE), sd(sens_counts_2, na.rm = 
TRUE), sd(sens_counts_3, na.rm = TRUE), sd(sens_counts_4, na.rm = TRUE), 
sd(sens_counts_5, na.rm = TRUE), 
          sd(sens_counts_6, na.rm = TRUE), sd(sens_counts_7, na.rm = 
TRUE), sd(sens_counts_8, na.rm = TRUE), sd(sens_counts_9, na.rm = TRUE), 
sd(sens_counts_10, na.rm = TRUE), 
          sd(sens_counts_11, na.rm = TRUE), sd(sens_counts_12, na.rm = 
TRUE), sd(sens_counts_13, na.rm = TRUE), sd(sens_counts_14, na.rm = TRUE), 
sd(sens_counts_15, na.rm = TRUE), 
          sd(sens_counts_16, na.rm = TRUE), sd(sens_counts_17, na.rm = 
TRUE))
y_micro_sd <- c(sd(spec_counts_1, na.rm = TRUE), sd(spec_counts_2, na.rm = 
TRUE), sd(spec_counts_3, na.rm = TRUE), sd(spec_counts_4, na.rm = TRUE), 
sd(spec_counts_5, na.rm = TRUE), 
          sd(spec_counts_6, na.rm = TRUE), sd(spec_counts_7, na.rm = 
TRUE), sd(spec_counts_8, na.rm = TRUE), sd(spec_counts_9, na.rm = TRUE), 
sd(spec_counts_10, na.rm = TRUE), 
          sd(spec_counts_11, na.rm = TRUE), sd(spec_counts_12, na.rm = 
TRUE), sd(spec_counts_13, na.rm = TRUE), sd(spec_counts_14, na.rm = TRUE), 
sd(spec_counts_15, na.rm = TRUE), 
          sd(spec_counts_16, na.rm = TRUE), sd(spec_counts_17, na.rm = 
TRUE))

spec_df <- meta_model_spec

spec_counts_1 <- c()
spec_counts_2 <- c()
spec_counts_3 <- c()
spec_counts_4 <- c()
spec_counts_5 <- c()
spec_counts_6 <- c()
spec_counts_7 <- c()
spec_counts_8 <- c()
spec_counts_9 <- c()
spec_counts_10 <- c()
spec_counts_11 <- c()
spec_counts_12 <- c()
spec_counts_13 <- c()
spec_counts_14 <- c()
spec_counts_15 <- c()
spec_counts_16 <- c()
spec_counts_17 <- c()
for (i in seq(1, length(spec_df), 17)){
  spec_counts_1<- c(spec_counts_1, spec_df[i])
  spec_counts_2<- c(spec_counts_2, spec_df[i+1])
  spec_counts_3<- c(spec_counts_3, spec_df[i+2])
  spec_counts_4<- c(spec_counts_4, spec_df[i+3])
  spec_counts_5<- c(spec_counts_5, spec_df[i+4])
  spec_counts_6<- c(spec_counts_6, spec_df[i+5])
  spec_counts_7<- c(spec_counts_7, spec_df[i+6])
  spec_counts_8<- c(spec_counts_8, spec_df[i+7])
  spec_counts_9<- c(spec_counts_9, spec_df[i+8])
  spec_counts_10<- c(spec_counts_10, spec_df[i+9])
  spec_counts_11<- c(spec_counts_11, spec_df[i+10])
  spec_counts_12<- c(spec_counts_12, spec_df[i+11])
  spec_counts_13<- c(spec_counts_13, spec_df[i+12])
  spec_counts_14<- c(spec_counts_14, spec_df[i+13])
  spec_counts_15<- c(spec_counts_15, spec_df[i+14])
  spec_counts_16<- c(spec_counts_16, spec_df[i+15])
  spec_counts_17<- c(spec_counts_17, spec_df[i+16])
}




sens_df <- meta_model_sens

sens_counts_1 <- c()
sens_counts_2 <- c()
sens_counts_3 <- c()
sens_counts_4 <- c()
sens_counts_5 <- c()
sens_counts_6 <- c()
sens_counts_7 <- c()
sens_counts_8 <- c()
sens_counts_9 <- c()
sens_counts_10 <- c()
sens_counts_11 <- c()
sens_counts_12 <- c()
sens_counts_13 <- c()
sens_counts_14 <- c()
sens_counts_15 <- c()
sens_counts_16 <- c()
sens_counts_17 <- c()
for (i in seq(1, length(sens_df), 17)){
  sens_counts_1<- c(sens_counts_1, sens_df[i])
  sens_counts_2<- c(sens_counts_2, sens_df[i+1])
  sens_counts_3<- c(sens_counts_3, sens_df[i+2])
  sens_counts_4<- c(sens_counts_4, sens_df[i+3])
  sens_counts_5<- c(sens_counts_5, sens_df[i+4])
  sens_counts_6<- c(sens_counts_6, sens_df[i+5])
  sens_counts_7<- c(sens_counts_7, sens_df[i+6])
  sens_counts_8<- c(sens_counts_8, sens_df[i+7])
  sens_counts_9<- c(sens_counts_9, sens_df[i+8])
  sens_counts_10<- c(sens_counts_10, sens_df[i+9])
  sens_counts_11<- c(sens_counts_11, sens_df[i+10])
  sens_counts_12<- c(sens_counts_12, sens_df[i+11])
  sens_counts_13<- c(sens_counts_13, sens_df[i+12])
  sens_counts_14<- c(sens_counts_14, sens_df[i+13])
  sens_counts_15<- c(sens_counts_15, sens_df[i+14])
  sens_counts_16<- c(sens_counts_16, sens_df[i+15])
  sens_counts_17<- c(sens_counts_17, sens_df[i+16])
}


x_meta_data <- c(1-mean(sens_counts_1, na.rm = TRUE), 
1-mean(sens_counts_2, na.rm = TRUE), 1-mean(sens_counts_3, na.rm = TRUE), 
1-mean(sens_counts_4, na.rm = TRUE), 1-mean(sens_counts_5, na.rm = TRUE), 
            1-mean(sens_counts_6, na.rm = TRUE), 1-mean(sens_counts_7, 
na.rm = TRUE), 1-mean(sens_counts_8, na.rm = TRUE), 1-mean(sens_counts_9, 
na.rm = TRUE), 1-mean(sens_counts_10, na.rm = TRUE), 
            1-mean(sens_counts_11, na.rm = TRUE), 1-mean(sens_counts_12, 
na.rm = TRUE),1- mean(sens_counts_13, na.rm = TRUE), 
1-mean(sens_counts_14, na.rm = TRUE), 1-mean(sens_counts_15, na.rm = 
TRUE), 
            1-mean(sens_counts_16, na.rm = TRUE), 1-mean(sens_counts_17, 
na.rm = TRUE))
y_meta_data <- c(mean(spec_counts_1, na.rm = TRUE), mean(spec_counts_2, 
na.rm = TRUE), mean(spec_counts_3, na.rm = TRUE), mean(spec_counts_4, 
na.rm = TRUE), mean(spec_counts_5, na.rm = TRUE), 
            mean(spec_counts_6, na.rm = TRUE), mean(spec_counts_7, na.rm = 
TRUE), mean(spec_counts_8, na.rm = TRUE), mean(spec_counts_9, na.rm = 
TRUE), mean(spec_counts_10, na.rm = TRUE), 
            mean(spec_counts_11, na.rm = TRUE), mean(spec_counts_12, na.rm 
= TRUE), mean(spec_counts_13, na.rm = TRUE), mean(spec_counts_14, na.rm = 
TRUE), mean(spec_counts_15, na.rm = TRUE), 
            mean(spec_counts_16, na.rm = TRUE), mean(spec_counts_17, na.rm 
= TRUE))

x_meta_sd <- c(sd(sens_counts_1, na.rm = TRUE), sd(sens_counts_2, na.rm = 
TRUE), sd(sens_counts_3, na.rm = TRUE), sd(sens_counts_4, na.rm = TRUE), 
sd(sens_counts_5, na.rm = TRUE), 
          sd(sens_counts_6, na.rm = TRUE), sd(sens_counts_7, na.rm = 
TRUE), sd(sens_counts_8, na.rm = TRUE), sd(sens_counts_9, na.rm = TRUE), 
sd(sens_counts_10, na.rm = TRUE), 
          sd(sens_counts_11, na.rm = TRUE), sd(sens_counts_12, na.rm = 
TRUE), sd(sens_counts_13, na.rm = TRUE), sd(sens_counts_14, na.rm = TRUE), 
sd(sens_counts_15, na.rm = TRUE), 
          sd(sens_counts_16, na.rm = TRUE), sd(sens_counts_17, na.rm = 
TRUE))
y_meta_sd <- c(sd(spec_counts_1, na.rm = TRUE), sd(spec_counts_2, na.rm = 
TRUE), sd(spec_counts_3, na.rm = TRUE), sd(spec_counts_4, na.rm = TRUE), 
sd(spec_counts_5, na.rm = TRUE), 
          sd(spec_counts_6, na.rm = TRUE), sd(spec_counts_7, na.rm = 
TRUE), sd(spec_counts_8, na.rm = TRUE), sd(spec_counts_9, na.rm = TRUE), 
sd(spec_counts_10, na.rm = TRUE), 
          sd(spec_counts_11, na.rm = TRUE), sd(spec_counts_12, na.rm = 
TRUE), sd(spec_counts_13, na.rm = TRUE), sd(spec_counts_14, na.rm = TRUE), 
sd(spec_counts_15, na.rm = TRUE), 
          sd(spec_counts_16, na.rm = TRUE), sd(spec_counts_17, na.rm = 
TRUE))




spec_df <- path_model_spec

spec_counts_1 <- c()
spec_counts_2 <- c()
spec_counts_3 <- c()
spec_counts_4 <- c()
spec_counts_5 <- c()
spec_counts_6 <- c()
spec_counts_7 <- c()
spec_counts_8 <- c()
spec_counts_9 <- c()
spec_counts_10 <- c()
spec_counts_11 <- c()
spec_counts_12 <- c()
spec_counts_13 <- c()
spec_counts_14 <- c()
spec_counts_15 <- c()
spec_counts_16 <- c()
spec_counts_17 <- c()
for (i in seq(1, length(spec_df), 17)){
  spec_counts_1<- c(spec_counts_1, spec_df[i])
  spec_counts_2<- c(spec_counts_2, spec_df[i+1])
  spec_counts_3<- c(spec_counts_3, spec_df[i+2])
  spec_counts_4<- c(spec_counts_4, spec_df[i+3])
  spec_counts_5<- c(spec_counts_5, spec_df[i+4])
  spec_counts_6<- c(spec_counts_6, spec_df[i+5])
  spec_counts_7<- c(spec_counts_7, spec_df[i+6])
  spec_counts_8<- c(spec_counts_8, spec_df[i+7])
  spec_counts_9<- c(spec_counts_9, spec_df[i+8])
  spec_counts_10<- c(spec_counts_10, spec_df[i+9])
  spec_counts_11<- c(spec_counts_11, spec_df[i+10])
  spec_counts_12<- c(spec_counts_12, spec_df[i+11])
  spec_counts_13<- c(spec_counts_13, spec_df[i+12])
  spec_counts_14<- c(spec_counts_14, spec_df[i+13])
  spec_counts_15<- c(spec_counts_15, spec_df[i+14])
  spec_counts_16<- c(spec_counts_16, spec_df[i+15])
  spec_counts_17<- c(spec_counts_17, spec_df[i+16])
}




sens_df <- path_model_sens

sens_counts_1 <- c()
sens_counts_2 <- c()
sens_counts_3 <- c()
sens_counts_4 <- c()
sens_counts_5 <- c()
sens_counts_6 <- c()
sens_counts_7 <- c()
sens_counts_8 <- c()
sens_counts_9 <- c()
sens_counts_10 <- c()
sens_counts_11 <- c()
sens_counts_12 <- c()
sens_counts_13 <- c()
sens_counts_14 <- c()
sens_counts_15 <- c()
sens_counts_16 <- c()
sens_counts_17 <- c()
for (i in seq(1, length(sens_df), 17)){
  sens_counts_1<- c(sens_counts_1, sens_df[i])
  sens_counts_2<- c(sens_counts_2, sens_df[i+1])
  sens_counts_3<- c(sens_counts_3, sens_df[i+2])
  sens_counts_4<- c(sens_counts_4, sens_df[i+3])
  sens_counts_5<- c(sens_counts_5, sens_df[i+4])
  sens_counts_6<- c(sens_counts_6, sens_df[i+5])
  sens_counts_7<- c(sens_counts_7, sens_df[i+6])
  sens_counts_8<- c(sens_counts_8, sens_df[i+7])
  sens_counts_9<- c(sens_counts_9, sens_df[i+8])
  sens_counts_10<- c(sens_counts_10, sens_df[i+9])
  sens_counts_11<- c(sens_counts_11, sens_df[i+10])
  sens_counts_12<- c(sens_counts_12, sens_df[i+11])
  sens_counts_13<- c(sens_counts_13, sens_df[i+12])
  sens_counts_14<- c(sens_counts_14, sens_df[i+13])
  sens_counts_15<- c(sens_counts_15, sens_df[i+14])
  sens_counts_16<- c(sens_counts_16, sens_df[i+15])
  sens_counts_17<- c(sens_counts_17, sens_df[i+16])
}


x_path_data <- c(1-mean(sens_counts_1, na.rm = TRUE), 
1-mean(sens_counts_2, na.rm = TRUE), 1-mean(sens_counts_3, na.rm = TRUE), 
1-mean(sens_counts_4, na.rm = TRUE), 1-mean(sens_counts_5, na.rm = TRUE), 
            1-mean(sens_counts_6, na.rm = TRUE), 1-mean(sens_counts_7, 
na.rm = TRUE), 1-mean(sens_counts_8, na.rm = TRUE), 1-mean(sens_counts_9, 
na.rm = TRUE), 1-mean(sens_counts_10, na.rm = TRUE), 
            1-mean(sens_counts_11, na.rm = TRUE), 1-mean(sens_counts_12, 
na.rm = TRUE),1- mean(sens_counts_13, na.rm = TRUE), 
1-mean(sens_counts_14, na.rm = TRUE), 1-mean(sens_counts_15, na.rm = 
TRUE), 
            1-mean(sens_counts_16, na.rm = TRUE), 1-mean(sens_counts_17, 
na.rm = TRUE))
y_path_data <- c(mean(spec_counts_1, na.rm = TRUE), mean(spec_counts_2, 
na.rm = TRUE), mean(spec_counts_3, na.rm = TRUE), mean(spec_counts_4, 
na.rm = TRUE), mean(spec_counts_5, na.rm = TRUE), 
            mean(spec_counts_6, na.rm = TRUE), mean(spec_counts_7, na.rm = 
TRUE), mean(spec_counts_8, na.rm = TRUE), mean(spec_counts_9, na.rm = 
TRUE), mean(spec_counts_10, na.rm = TRUE), 
            mean(spec_counts_11, na.rm = TRUE), mean(spec_counts_12, na.rm 
= TRUE), mean(spec_counts_13, na.rm = TRUE), mean(spec_counts_14, na.rm = 
TRUE), mean(spec_counts_15, na.rm = TRUE), 
            mean(spec_counts_16, na.rm = TRUE), mean(spec_counts_17, na.rm 
= TRUE))

x_path_sd <- c(sd(sens_counts_1, na.rm = TRUE), sd(sens_counts_2, na.rm = 
TRUE), sd(sens_counts_3, na.rm = TRUE), sd(sens_counts_4, na.rm = TRUE), 
sd(sens_counts_5, na.rm = TRUE), 
          sd(sens_counts_6, na.rm = TRUE), sd(sens_counts_7, na.rm = 
TRUE), sd(sens_counts_8, na.rm = TRUE), sd(sens_counts_9, na.rm = TRUE), 
sd(sens_counts_10, na.rm = TRUE), 
          sd(sens_counts_11, na.rm = TRUE), sd(sens_counts_12, na.rm = 
TRUE), sd(sens_counts_13, na.rm = TRUE), sd(sens_counts_14, na.rm = TRUE), 
sd(sens_counts_15, na.rm = TRUE), 
          sd(sens_counts_16, na.rm = TRUE), sd(sens_counts_17, na.rm = 
TRUE))
y_path_sd <- c(sd(spec_counts_1, na.rm = TRUE), sd(spec_counts_2, na.rm = 
TRUE), sd(spec_counts_3, na.rm = TRUE), sd(spec_counts_4, na.rm = TRUE), 
sd(spec_counts_5, na.rm = TRUE), 
          sd(spec_counts_6, na.rm = TRUE), sd(spec_counts_7, na.rm = 
TRUE), sd(spec_counts_8, na.rm = TRUE), sd(spec_counts_9, na.rm = TRUE), 
sd(spec_counts_10, na.rm = TRUE), 
          sd(spec_counts_11, na.rm = TRUE), sd(spec_counts_12, na.rm = 
TRUE), sd(spec_counts_13, na.rm = TRUE), sd(spec_counts_14, na.rm = TRUE), 
sd(spec_counts_15, na.rm = TRUE), 
          sd(spec_counts_16, na.rm = TRUE), sd(spec_counts_17, na.rm = 
TRUE))






```





```{r }
x <- new_x_partial_data
y <- new_y_partial_data

upper_bound <- y + new_y_partial_sd
lower_bound <- y - new_y_partial_sd

# Create a data frame
data1 <- data.frame(x = x, y = y, upper = upper_bound, lower = 
lower_bound)


x2 <- x_full_data
y2 <- y_full_data
upper_bound2 <- y2 + y_full_sd
lower_bound2 <- y2 - y_full_sd

xmicro <- x_micro_data
ymicro <- y_micro_data
upper_bound_micro <- ymicro + y_micro_sd
lower_bound_micro <- ymicro - y_micro_sd


xmeta <- x_meta_data
ymeta <- y_meta_data
upper_bound_meta <- ymeta + y_meta_sd
lower_bound_meta <- ymeta - y_meta_sd


xpaths <- x_path_data
ypaths <- y_path_data
upper_bound_paths <- ypaths + y_path_sd
lower_bound_paths <- ypaths - y_path_sd


data2 <- data.frame(x = x2, y = y2, upper = upper_bound2, lower = 
lower_bound2)
data3 <- data.frame(x=c(0,1), y=c(0,1))
data_micro <- data.frame(x = xmicro, y = ymicro, upper = 
upper_bound_micro, lower = lower_bound_micro)
data_meta <- data.frame(x = xmeta, y = ymeta, upper = upper_bound_meta, 
lower = lower_bound_meta)
data_paths <- data.frame(x = xpaths, y = ypaths, upper = 
upper_bound_paths, lower = lower_bound_paths)
# Plot
ggplot() +
  
  #geom_point(data = data1, aes(x = x, y = y)) +
  geom_ribbon(data = data2, aes(x = x, ymin = lower, ymax = upper), fill = 
"black", color = NA, alpha=0.2 ) +
  geom_line(data = data2, aes(x = x, y = y), color = "black", alpha=0.7, 
linewidth=2) +
  geom_ribbon(data = data1, aes(x = x, ymin = lower, ymax = upper), fill = 
"darkgreen", color = NA, alpha=0.1) +
  geom_line(data = data1, aes(x = x, y = y), color='darkgreen', alpha=1, 
linetype=5) +
  geom_ribbon(data = data_micro, aes(x = x, ymin = lower, ymax = upper), 
fill = "red", color = NA, alpha=0.1 ) +
  geom_line(data = data_micro, aes(x = x, y = y), color = "red", 
alpha=0.7, linetype=5) +
  geom_ribbon(data = data_meta, aes(x = x, ymin = lower, ymax = upper), 
fill = "purple", color = NA, alpha=0.1 ) +
  geom_line(data = data_meta, aes(x = x, y = y), color = "purple", 
alpha=0.7, linetype=5) +
  geom_ribbon(data = data_paths, aes(x = x, ymin = lower, ymax = upper), 
fill = "blue4", color = NA, alpha=0.1 ) +
  geom_line(data = data_paths, aes(x = x, y = y), color = "blue", 
alpha=0.7, linetype=5) +
  geom_line(data=data3, aes(x = x, y = y), color = "black", alpha=0.7, 
linetype=2) +
  #geom_point(data = data2, aes(x = x, y = y), color = "red") +
  labs(x = "1 - specificity", y = "Sensitivity", title='ROC-AUC plot 
Clinical features vs Microbial, Pathway and Metabolite ratios') +
  theme_bw()



pdf("/Users/iwan/Research/images/drug_Response/drug_response_pdfs/allAUCs.pdf")
ggplot() +
  geom_ribbon(data = data2, aes(x = x, ymin = lower, ymax = upper), fill = 
"black", color = NA, alpha=0.3, show.legend = FALSE) +
  
  geom_ribbon(data = data1, aes(x = x, ymin = lower, ymax = upper), fill = 
"grey70", color = NA, alpha=0.2, show.legend = FALSE) +
  
  geom_ribbon(data = data_micro, aes(x = x, ymin = lower, ymax = upper), 
fill = "grey70", color = NA, alpha=0.2, show.legend = FALSE) +
  
  geom_ribbon(data = data_meta, aes(x = x, ymin = lower, ymax = upper), 
fill = "grey70", color = NA, alpha=0.2, show.legend = FALSE) +
  
  geom_ribbon(data = data_paths, aes(x = x, ymin = lower, ymax = upper), 
fill = "grey70", color = NA, alpha=0.2, show.legend = FALSE) +
  geom_line(data = data1, aes(x = x, y = y, color = "2Green dashed line"), 
alpha=1, linetype=5) +
  geom_line(data = data_micro, aes(x = x, y = y, color = "3Red dashed 
line"), alpha=0.7, linetype=5) +
  geom_line(data = data_meta, aes(x = x, y = y, color = "4Purple dashed 
line"), alpha=0.7, linetype=5) +
  geom_line(data = data2, aes(x = x, y = y, color = "1Black line"), 
alpha=0.7, linewidth=2) +
  geom_line(data = data_paths, aes(x = x, y = y, color = "5Blue dashed 
line"), alpha=0.7, linetype=5) +
  geom_line(data=data3, aes(x = x, y = y, color = "6Black dashed line"), 
alpha=0.7, linetype=2) +
  labs(x = "1 - specificity", y = "Sensitivity", title='ROC-AUC plot 
Clinical features vs Microbial, Pathway and Metabolite ratios') +
  scale_color_manual(values = c("1Black line" = "black", 
                                 "2Green dashed line" = "darkgreen", 
                                 "3Red dashed line" = "red", 
                                 "4Purple dashed line" = "purple", 
                                 "5Blue dashed line" = "blue", 
                                 "6Black dashed line" = "black"),
                     labels = c("All features combined, AUC=0.73±0.12", 
"Clinical features only, AUC=0.71±0.13", "Microbes only, AUC=0.71±0.13", 
"Metabolites only, AUC=0.70±0.14", "Pathways only, AUC=0.61±0.11", 
"Baseline")) +
  theme_bw()
dev.off()
``` 


# mimenet dataprep
```{r }

rownames(combined_all_new_ID_raw_filt) <- All_clinical_data$Row.names
rownames(combined_microbiome_sub_counts) <-  All_clinical_data$Row.names


microbiome_values <- as.data.frame(t(combined_microbiome_sub_counts))
metabolite_values <- as.data.frame(t(combined_all_new_ID_raw_filt))

write.csv(microbiome_values, 
"/Users/iwan/Research/mimenet/MiMeNet/data/drug_response/microbiome.csv")
write.csv(metabolite_values, 
"/Users/iwan/Research/mimenet/MiMeNet/data/drug_response/metabolome.csv")
write.csv(All_clinical_data[, c('Row.names', 'ResponseScoreYesNo')], 
"/Users/iwan/Research/mimenet/MiMeNet/data/drug_response/diagnosis_response.csv", 
row.names = FALSE)




interaction_output <- 
read.csv('/Users/iwan/Research/mimenet/MiMeNet/results/drug_response/CV/interaction_score_matrix.csv',  
row.names='X')
microbe_output <- 
read.csv('/Users/iwan/Research/mimenet/MiMeNet/results/drug_response/CV/microbe_cluster_matrix.csv',  
row.names='X')
metabolites_output <- 
read.csv('/Users/iwan/Research/mimenet/MiMeNet/results/drug_response/CV/metabolite_cluster_matrix.csv',  
row.names='X')
heatmap(as.matrix(microbe_output))
```









## plotting network graph
```{r }

library(igraph)

create_combined_matrix <- function(corr_matrix, metabolites, microbes) {
  n_metabolites <- length(metabolites)
  n_microbes <- length(microbes)
  combined_matrix <- matrix(NA, nrow = n_metabolites + n_microbes, ncol = 
n_metabolites + n_microbes,
                            dimnames = list(c(metabolites, microbes), 
c(metabolites, microbes)))
  combined_matrix[1:n_metabolites, (n_metabolites + 1):(n_metabolites + 
n_microbes)] <- corr_matrix
  combined_matrix[(n_metabolites + 1):(n_metabolites + n_microbes), 
1:n_metabolites] <- t(corr_matrix)
  return(combined_matrix)
}

# Replace "metabolites" and "microbes" with your actual vectors of names
# Replace "corr_matrix" with the name of your correlation matrix variable
combined_matrix <- create_combined_matrix(interaction_output, 
rownames(interaction_output),colnames(interaction_output))

# Print or use the combined matrix as needed
print(combined_matrix)

 
# Threshold value for considering correlations
threshold <- 0.13

# Function to convert correlation matrix to adjacency matrix
convert_to_adjacency <- function(corr_matrix, threshold) {
  adjacency_matrix <- ifelse(abs(corr_matrix) > threshold, 1, 0)
  diag(adjacency_matrix) <- 0 # Set diagonal elements to 0
  return(adjacency_matrix)
}

# Convert correlation matrix to adjacency matrix
adjacency_matrix <- convert_to_adjacency(interaction_output, threshold)

# Print or use the adjacency matrix as needed
#print(adjacency_matrix)

network <- graph_from_adjacency_matrix(adjacency_matrix[1:50, 1:50] , 
mode='undirected', diag=F )
plot(network, layout=layout.fruchterman.reingold, 
main="fruchterman.reingold")
plot(network, layout=layout.circle, main="circle")
```

Every node is a metabolite and or microbe
plot correlations between them 
with the size of the association,

determine cutoff 

double the matrix, 
then its also an adjacency matrix ?






```{r }


interaction_output 
microbe_output 
metabolites_output 
microbe_clusters_named <- 
read.csv('/Users/iwan/Research/mimenet/MiMeNet/results/drug_response/microbe_clusters.csv')
metabolite_clusters_named <- 
read.csv('/Users/iwan/Research/mimenet/MiMeNet/results/drug_response/metabolite_clusters.csv')
devtools::install_github("briatte/ggnet")
library(ggnet)
library(network)
library(sna)
library(ggplot2)


net = rgraph(10, mode = "graph", tprob = 0.5)
net = network(net, directed = FALSE)

# vertex names
network.vertex.names(net) = letters[1:10]
ggnet2(net)


metabolite_clusters <- c(0,1,2,3,4,5,6,7,8,9)
microbe_clusters <- c(0,1,2,3,4,5,6)


zero_zero <- 
mean(rowMeans(interaction_output[microbe_clusters_named[microbe_clusters_named$Cluster 
==0 ,]$X, 
metabolite_clusters_named[metabolite_clusters_named$Cluster==0,]$X]))
cluster_plot_matrix <- matrix(0, 10,7)
for (i in 0:9){
  for (j in 0:6){
    cluster_plot_matrix[i+1,j+1] <- 
mean(rowMeans(interaction_output[microbe_clusters_named[microbe_clusters_named$Cluster 
==j ,]$X, 
metabolite_clusters_named[metabolite_clusters_named$Cluster==i,]$X]))
  }
}

cluster_plot_matrix
network(cluster_plot_matrix, directed = FALSE)

cluster_plot_matrix_v2 <- matrix(0,17,17)

for (i in 0:9){
  for (j in 0:6){
    cluster_plot_matrix_v2[i+1,j+1] <- 
mean(rowMeans(interaction_output[microbe_clusters_named[microbe_clusters_named$Cluster 
==j ,]$X, 
metabolite_clusters_named[metabolite_clusters_named$Cluster==i,]$X]))
  }
}
cluster_plot_matrix_v2

g <- graph_from_adjacency_matrix(cluster_plot_matrix_v2, mode = 
"undirected", weighted = TRUE)





plot(g, edge.arrow.size = 0.5, edge.label = round(E(g)$weight, 2))

# Function to plot edges above a certain threshold
plot_edges_above_threshold <- function(threshold) {
  E(g)$color <- ifelse(E(g)$weight > threshold, "blue", "grey")
  plot(g, edge.arrow.size = 0.5, edge.label = round(E(g)$weight, 2), 
edge.color = E(g)$color)
}

# Example: Plot edges above a threshold of 0.5
plot_edges_above_threshold(0.1)



cluster_plot_matrix_v2 <- matrix(0, 17,17)
cluster_plot_matrix
for (i in 1:7){
  for (j in 1:10){
    cluster_plot_matrix_v2[i,7+j] <- cluster_plot_matrix[j,i]
    cluster_plot_matrix_v2[7+j, i] <- cluster_plot_matrix[j,i]
  }

}

g <- graph_from_adjacency_matrix(cluster_plot_matrix_v2, mode = 
"undirected", weighted = TRUE)
plot(g, edge.arrow.size = 0.5, edge.label = round(E(g)$weight, 2))

# Function to plot edges above a certain threshold
plot_edges_above_threshold <- function(threshold) {
  E(g)$color <- ifelse(E(g)$weight > threshold, "blue", "grey")
  plot(g, edge.arrow.size = 0.5, edge.label = round(E(g)$weight, 2), 
edge.color = E(g)$color)
}

# Example: Plot edges above a threshold of 0.5
plot_edges_above_threshold(0.1)


cluster_plot_matrix_v3  <- cluster_plot_matrix_v2
for (i in 1:17){
  for (j in 1:17){
    if (abs(cluster_plot_matrix_v3[i,j]) < 0.02 ){
      cluster_plot_matrix_v3[i,j] <- 0
    }
     
  }
}


#plot(g, edge.arrow.size = 0.5, edge.label = round(E(g)$weight, 2))
node_scalar <- 3

g <- graph_from_adjacency_matrix(cluster_plot_matrix_v3, mode = 
"undirected", weighted = TRUE)

plot_edges_color_by_weight <- function() {
  #node_degrees <- degree(g)
  E(g)$color <- ifelse(E(g)$weight > 0, "black", "grey")
  E(g)$weight <- abs(E(g)$weight) 
  layout <- layout_in_circle(g)
  #node_degrees <- degree(g)
  V(g)$color <- ifelse(V(g) %in% 1:7, "#eb7e9f", "#6aadee")
  V(g)$name <- c('M1','M2','M3','M4','M5','M6', 'M7', 
'm1','m2','m3','m4','m5','m6','m7','m8','m9','m10')
  # Scale the node sizes based on their degrees
  #node_sizes <- 10 + 15 * (node_degrees - min(node_degrees)) / 
(max(node_degrees) - min(node_degrees))
  
  plot(g, edge.arrow.size = 1.5, edge.color = E(g)$color, vertex.color = 
V(g)$color,vertex.frame.color="white", 
       vertex.label.color='black', vertex.label.font=0, layout = layout, 
main="Network graph metabolite and microbe cluster interactions"
       ) 
}

# Plot the graph with colored edges. ,  vertex.size = node_sizes,    
plot_edges_color_by_weight()

#, vertex.label.dist = 4,


#layout_with_graphopt(g, charge=8)
#'Bifidobacterium_bifidum\nCluster size=5', 
'Escherichia_coli\nRuminococcus_gnavus\nBacteroides_fragilis\nCluster 
size=13','Clostridiales_bacterium\nCluster 
size=9','Faecalibacterium_prausnitzii\nCluster size=6',
#                 'Alistipes_finegoldii\nBacteroides_caccae\nCluster 
size=21','Bacteroides_eggerthii\nRoseburia_faecis\nCluster 
size=6','Prevotella_copri_clade_A\nCluster #size=3',
#                 'Serotonin\nCluster 
size=158','Cholate\nGulonate\nSucrose\nCluster size=86',
#                 'Beta-alanine\nCluster size=50','Lactate\nCluster 
size=55','Ethanolamides\nCluster size=16','Glycerol\nCytosine\nCluster 
size=34','Maltose\nCluster #size=10','Salicylate\nCluster 
size=5','Urate\nCluster size=20','Glycylisoleucine\nCluster size=4'





metabolite_names_df <- 
read.csv("/Users/iwan/Research/drug_response/metadata/lastest_metadata/metabolites_naming_key.txt", 
sep='\t', header = TRUE)
metabolite_names <- c()
for (i in metabolite_clusters_named$X){
  metabolite_names <- c(metabolite_names, 
metabolite_names_df$CHEMICAL_NAME[metabolite_names_df$XCHEM_ID == i])
}
metabolite_clusters_named$names <- metabolite_names

shortnames <- c()
for (i in microbe_clusters_named$X){
  shortnames <- c(shortnames, str_split_i(i, '__',8))
}

microbe_clusters_named$names <- shortnames


for (i in 0:9){
  
print(metabolite_clusters_named[metabolite_clusters_named$Cluster==i,]$names)
} 

for (i in 0:6){
  print(microbe_clusters_named[microbe_clusters_named$Cluster==i,]$names)
}


print(metabolite_clusters_named[metabolite_clusters_named$Cluster==4,]$names)
```



### Here we compare some models
```{r }

# assign models and train
library(lmtest)


my_phenos=All_clinical_data[,c("V1_DiseaseDurationYears","V1_BMI","V1_Sex", 
"V1_FecalCalprotectine","V1_SerumCRP","ResponseScoreYesNo", 
"V1_PreviousantiTNF", "V2_disease_activity")]

my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="yes"]=1
my_phenos$ResponseScoreYesNo[my_phenos$ResponseScoreYesNo=="no"]=0
my_phenos$ResponseScoreYesNo=as.numeric(as.character(my_phenos$ResponseScoreYesNo))

my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='No 
activity']=0
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Mild']=1
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Moderate']=2
my_phenos$V2_disease_activity[my_phenos$V2_disease_activity=='Severe']=3
my_phenos$V2_disease_activity=as.numeric(as.character(my_phenos$V2_disease_activity))

my_phenos$V1_Sex[my_phenos$V1_Sex=='female']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='Female ']='Female'
my_phenos$V1_Sex[my_phenos$V1_Sex=='male']='Male'
my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos$V1_SerumCRP=as.numeric(as.character(my_phenos$V1_SerumCRP))

my_phenos$V1_SerumCRP[my_phenos$V1_SerumCRP=='<0.3'] = 0
my_phenos_no_na <- drop_na(my_phenos)


my_phenos_no_na$ratio_microbiome <- NULL
my_phenos_no_na$ratio_metabolites <- NULL
my_phenos_no_na$ratio_pathways <- NULL
        
#trainIndex <- sample(1:nrow(my_phenos_no_na), 0.75 * 
nrow(my_phenos_no_na))
        
#my_phenos_no_na <- drop_na(my_phenos)
  
train_correct_IBD=my_phenos_no_na
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo

        #Regress out age, sex, bmi, calprotectin to IBD
partial_IBD <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')

stacked_df_bugs_filtered <- stacked_df_bugs[stacked_df_bugs$values > 0.1,]
        
stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
vedo_6_month_microbe_ratio <- 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_microbiome_sub_counts[,stacked_df_bugs_vedo_6month_denum$predict])
        
my_phenos$ratio_microbiome <- vedo_6_month_microbe_ratio
        
        

stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]
        
stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio

stacked_df_bugs_filtered <- stacked_df_pathways[stacked_df_pathways$values 
> 0.1,]
        
stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
vedo_6_month_microbe_ratio <- 
rowSums(Combined_pathways_codacore_input[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(Combined_pathways_codacore_input[,stacked_df_bugs_vedo_6month_denum$predict])
        
my_phenos$ratio_pathways <- vedo_6_month_microbe_ratio
        
        
        #### here we add the prediction based on only the ratio see if 
that works : 
my_phenos_no_na <- drop_na(my_phenos)
        
my_phenos_no_na_microbiome <- my_phenos_no_na[,c('ResponseScoreYesNo', 
'ratio_microbiome')]
my_phenos_no_na_metabolites <- my_phenos_no_na[,c('ResponseScoreYesNo', 
'ratio_metabolites')]
my_phenos_no_na_pathways <- my_phenos_no_na[,c('ResponseScoreYesNo', 
'ratio_pathways')]
        
        
train_correct_IBD=my_phenos_no_na_microbiome
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
ratio_micro_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
train_correct_IBD=my_phenos_no_na_metabolites
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
ratio_meta_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
        
        
train_correct_IBD=my_phenos_no_na_pathways
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}

ratio_paths_model <- glm(ResponseScoreYesNo ~ ., data=train_correct_IBD, 
family='binomial')
        
  
my_phenos_no_na <- drop_na(my_phenos)
        
        
train_correct_IBD=my_phenos_no_na
        
train_IBD_pheno_y <- my_phenos_no_na$ResponseScoreYesNo
for (cc in colnames(train_correct_IBD)) {if 
(sum(is.na(train_correct_IBD[[cc]])) > 0) 
{train_correct_IBD[[cc]][is.na(train_correct_IBD[[cc]])] <- 
median(train_correct_IBD[[cc]],na.rm = T)}}
        
ratio_everything_model <- glm(ResponseScoreYesNo ~ ., 
data=train_correct_IBD, family='binomial')
        




# compare
lrtest(partial_IBD, ratio_everything_model)



```

## order the metabolite and microbiome clusters for the paper
```{r }
metabolite_clusters <- 
read.csv('/Users/iwan/Research/mimenet/MiMeNet/results/drug_response/metabolite_clusters.csv')
microbiome_clusters <- 
read.csv('/Users/iwan/Research/mimenet/MiMeNet/results/drug_response/microbe_clusters.csv')

# name the clusters
metabolite_clusters


metabolite_names_df <- 
read.csv("/Users/iwan/Research/drug_response/metadata/lastest_metadata/metabolites_naming_key.txt", 
sep='\t', header = TRUE)
metabolite_names <- c()
for (i in metabolite_clusters$X){
  metabolite_names <- c(metabolite_names, 
metabolite_names_df$CHEMICAL_NAME[metabolite_names_df$XCHEM_ID == i])
}
metabolite_clusters$names <- metabolite_names



for (i in microbiome_clusters[microbiome_clusters$Cluster == 7,'X']){
  print(i)
}
```


### check if the metabolites cluster together

```{r }


stacked_df_bugs_filtered <- stacked_df_metas[stacked_df_metas$values > 
0.1,]
        
stacked_df_bugs_vedo <- 
stacked_df_bugs_filtered[stacked_df_bugs_filtered$Medication == 'Combined 
cohort',]
        
stacked_df_bugs_vedo_6month <- 
stacked_df_bugs_vedo[stacked_df_bugs_vedo$response == '6 month',]
stacked_df_bugs_vedo_6month_num <- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Numerator',]
stacked_df_bugs_vedo_6month_denum<- 
stacked_df_bugs_vedo_6month[stacked_df_bugs_vedo_6month$type == 
'Denominator',]
        
vedo_6_month_microbe_ratio <- 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_num$predict]) 
/ 
rowSums(combined_all_new_ID_raw_filt[,stacked_df_bugs_vedo_6month_denum$predict])
        
my_phenos$ratio_metabolites <- vedo_6_month_microbe_ratio


corrplot(cor(combined_all_new_ID_raw_filt[,c(stacked_df_bugs_vedo_6month_num$predict, 
stacked_df_bugs_vedo_6month_denum$predict)]), method="circle")

```






